{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a7ea89e2-c057-46da-a7b3-a60962c13169",
   "metadata": {},
   "source": [
    "# mouse model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34f5af65-b574-48ba-b7db-45ca44212d94",
   "metadata": {},
   "source": [
    "## get positive and negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af91d44c-f0d4-4625-8cc6-dd828a5415f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b81b6728-d9e3-4947-8688-5f36ecf386f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/ruiyanhou/nfs_share2/conda_env/envs/Py39/lib/python3.9/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001343744.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>90838871</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868201</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>90839174</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868202</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90838871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868203</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90838871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868204 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+-------------+-----------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature     | Start     | End       | +20   |\n",
       "| (category)   | (object)   | (object)    | (int64)   | (int64)   | ...   |\n",
       "|--------------+------------+-------------+-----------+-----------+-------|\n",
       "| chr1         | HAVANA     | gene        | 3073252   | 3074322   | ...   |\n",
       "| chr1         | HAVANA     | transcript  | 3073252   | 3074322   | ...   |\n",
       "| chr1         | HAVANA     | exon        | 3073252   | 3074322   | ...   |\n",
       "| chr1         | ENSEMBL    | gene        | 3102015   | 3102125   | ...   |\n",
       "| ...          | ...        | ...         | ...       | ...       | ...   |\n",
       "| chrY         | ENSEMBL    | CDS         | 90838871  | 90839177  | ...   |\n",
       "| chrY         | ENSEMBL    | start_codon | 90839174  | 90839177  | ...   |\n",
       "| chrY         | ENSEMBL    | stop_codon  | 90838868  | 90838871  | ...   |\n",
       "| chrY         | ENSEMBL    | UTR         | 90838868  | 90838871  | ...   |\n",
       "+--------------+------------+-------------+-----------+-----------+-------+\n",
       "Stranded PyRanges object has 1,868,204 rows and 25 columns from 22 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "20 hidden columns: Score, Strand, Frame, gene_id, gene_type, gene_name, level, ... (+ 13 more.)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genepr=pr.read_gtf('/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode_mm10/gencode.vM23.annotation.gtf')\n",
    "genepr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "15751219-345c-452c-901e-a1a0790b3d3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001343744.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868199</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868200</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>90838871</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868201</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>90839174</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868202</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90838871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868203</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>UTR</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90838871</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Gm21748-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001062768.1</td>\n",
       "      <td>ENSMUSP00000137361.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868204 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source      Feature     Start       End Score Strand  \\\n",
       "0             chr1   HAVANA         gene   3073252   3074322     .      +   \n",
       "1             chr1   HAVANA   transcript   3073252   3074322     .      +   \n",
       "2             chr1   HAVANA         exon   3073252   3074322     .      +   \n",
       "3             chr1  ENSEMBL         gene   3102015   3102125     .      +   \n",
       "4             chr1  ENSEMBL   transcript   3102015   3102125     .      +   \n",
       "...            ...      ...          ...       ...       ...   ...    ...   \n",
       "1868199       chrY  ENSEMBL         exon  90838868  90839177     .      -   \n",
       "1868200       chrY  ENSEMBL          CDS  90838871  90839177     .      -   \n",
       "1868201       chrY  ENSEMBL  start_codon  90839174  90839177     .      -   \n",
       "1868202       chrY  ENSEMBL   stop_codon  90838868  90838871     .      -   \n",
       "1868203       chrY  ENSEMBL          UTR  90838868  90838871     .      -   \n",
       "\n",
       "        Frame               gene_id       gene_type  ... transcript_type  \\\n",
       "0           .  ENSMUSG00000102693.1             TEC  ...             NaN   \n",
       "1           .  ENSMUSG00000102693.1             TEC  ...             TEC   \n",
       "2           .  ENSMUSG00000102693.1             TEC  ...             TEC   \n",
       "3           .  ENSMUSG00000064842.1           snRNA  ...             NaN   \n",
       "4           .  ENSMUSG00000064842.1           snRNA  ...           snRNA   \n",
       "...       ...                   ...             ...  ...             ...   \n",
       "1868199     .  ENSMUSG00000096850.1  protein_coding  ...  protein_coding   \n",
       "1868200     0  ENSMUSG00000096850.1  protein_coding  ...  protein_coding   \n",
       "1868201     0  ENSMUSG00000096850.1  protein_coding  ...  protein_coding   \n",
       "1868202     0  ENSMUSG00000096850.1  protein_coding  ...  protein_coding   \n",
       "1868203     .  ENSMUSG00000096850.1  protein_coding  ...  protein_coding   \n",
       "\n",
       "           transcript_name transcript_support_level                 tag  \\\n",
       "0                      NaN                      NaN                 NaN   \n",
       "1        4933401J01Rik-201                       NA               basic   \n",
       "2        4933401J01Rik-201                       NA               basic   \n",
       "3                      NaN                      NaN                 NaN   \n",
       "4              Gm26206-201                       NA               basic   \n",
       "...                    ...                      ...                 ...   \n",
       "1868199        Gm21748-201                       NA  appris_principal_1   \n",
       "1868200        Gm21748-201                       NA  appris_principal_1   \n",
       "1868201        Gm21748-201                       NA  appris_principal_1   \n",
       "1868202        Gm21748-201                       NA  appris_principal_1   \n",
       "1868203        Gm21748-201                       NA  appris_principal_1   \n",
       "\n",
       "            havana_transcript exon_number               exon_id  \\\n",
       "0                         NaN         NaN                   NaN   \n",
       "1        OTTMUST00000127109.1         NaN                   NaN   \n",
       "2        OTTMUST00000127109.1           1  ENSMUSE00001343744.1   \n",
       "3                         NaN         NaN                   NaN   \n",
       "4                         NaN         NaN                   NaN   \n",
       "...                       ...         ...                   ...   \n",
       "1868199                   NaN           1  ENSMUSE00001062768.1   \n",
       "1868200                   NaN           1  ENSMUSE00001062768.1   \n",
       "1868201                   NaN           1  ENSMUSE00001062768.1   \n",
       "1868202                   NaN           1  ENSMUSE00001062768.1   \n",
       "1868203                   NaN           1  ENSMUSE00001062768.1   \n",
       "\n",
       "                   protein_id ccdsid  ont  \n",
       "0                         NaN    NaN  NaN  \n",
       "1                         NaN    NaN  NaN  \n",
       "2                         NaN    NaN  NaN  \n",
       "3                         NaN    NaN  NaN  \n",
       "4                         NaN    NaN  NaN  \n",
       "...                       ...    ...  ...  \n",
       "1868199  ENSMUSP00000137361.1    NaN  NaN  \n",
       "1868200  ENSMUSP00000137361.1    NaN  NaN  \n",
       "1868201  ENSMUSP00000137361.1    NaN  NaN  \n",
       "1868202  ENSMUSP00000137361.1    NaN  NaN  \n",
       "1868203  ENSMUSP00000137361.1    NaN  NaN  \n",
       "\n",
       "[1868204 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genedf=genepr.df\n",
    "genedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa578b0e-2fe2-4314-9769-42253c5097b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3252756</td>\n",
       "      <td>3253236</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102851.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudo_consens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3466586</td>\n",
       "      <td>3513553</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000089699.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3531794</td>\n",
       "      <td>3532720</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000103147.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudo_consens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868158</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90499046</td>\n",
       "      <td>90531645</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100964.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868170</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90603500</td>\n",
       "      <td>90605864</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099619.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868180</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90665345</td>\n",
       "      <td>90667625</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099399.6</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868190</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90752426</td>\n",
       "      <td>90755467</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095366.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55335 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0             chr1   HAVANA    gene   3073252   3074322     .      +     .   \n",
       "3             chr1  ENSEMBL    gene   3102015   3102125     .      +     .   \n",
       "6             chr1   HAVANA    gene   3252756   3253236     .      +     .   \n",
       "9             chr1   HAVANA    gene   3466586   3513553     .      +     .   \n",
       "13            chr1   HAVANA    gene   3531794   3532720     .      +     .   \n",
       "...            ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "1868158       chrY   HAVANA    gene  90499046  90531645     .      -     .   \n",
       "1868170       chrY   HAVANA    gene  90603500  90605864     .      -     .   \n",
       "1868180       chrY   HAVANA    gene  90665345  90667625     .      -     .   \n",
       "1868190       chrY   HAVANA    gene  90752426  90755467     .      -     .   \n",
       "1868197       chrY  ENSEMBL    gene  90838868  90839177     .      -     .   \n",
       "\n",
       "                      gene_id               gene_type  ... transcript_type  \\\n",
       "0        ENSMUSG00000102693.1                     TEC  ...             NaN   \n",
       "3        ENSMUSG00000064842.1                   snRNA  ...             NaN   \n",
       "6        ENSMUSG00000102851.1    processed_pseudogene  ...             NaN   \n",
       "9        ENSMUSG00000089699.1                  lncRNA  ...             NaN   \n",
       "13       ENSMUSG00000103147.1    processed_pseudogene  ...             NaN   \n",
       "...                       ...                     ...  ...             ...   \n",
       "1868158  ENSMUSG00000100964.1  unprocessed_pseudogene  ...             NaN   \n",
       "1868170  ENSMUSG00000099619.6                  lncRNA  ...             NaN   \n",
       "1868180  ENSMUSG00000099399.6                  lncRNA  ...             NaN   \n",
       "1868190  ENSMUSG00000095366.2                  lncRNA  ...             NaN   \n",
       "1868197  ENSMUSG00000096850.1          protein_coding  ...             NaN   \n",
       "\n",
       "        transcript_name transcript_support_level             tag  \\\n",
       "0                   NaN                      NaN             NaN   \n",
       "3                   NaN                      NaN             NaN   \n",
       "6                   NaN                      NaN  pseudo_consens   \n",
       "9                   NaN                      NaN             NaN   \n",
       "13                  NaN                      NaN  pseudo_consens   \n",
       "...                 ...                      ...             ...   \n",
       "1868158             NaN                      NaN             NaN   \n",
       "1868170             NaN                      NaN             NaN   \n",
       "1868180             NaN                      NaN             NaN   \n",
       "1868190             NaN                      NaN             NaN   \n",
       "1868197             NaN                      NaN             NaN   \n",
       "\n",
       "        havana_transcript exon_number exon_id protein_id ccdsid  ont  \n",
       "0                     NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "3                     NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "6                     NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "9                     NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "13                    NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "...                   ...         ...     ...        ...    ...  ...  \n",
       "1868158               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868170               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868180               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868190               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868197               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "\n",
       "[55335 rows x 25 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptdf=genedf[genedf['Feature']=='gene']\n",
    "transcriptdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f8c0ea1c-adb1-4e14-885b-a1dac4538c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886770</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897909</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606131</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151122</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530908</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745532</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21849 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "42            chr1   HAVANA    gene   4807787   4848410     .      +     .   \n",
       "191           chr1   HAVANA    gene   4807891   4886770     .      +     .   \n",
       "217           chr1   HAVANA    gene   4857813   4897909     .      +     .   \n",
       "276           chr1   HAVANA    gene   5070017   5162529     .      +     .   \n",
       "424           chr1   HAVANA    gene   5588465   5606131     .      +     .   \n",
       "...            ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "1867618       chrY   HAVANA    gene  79148788  79151122     .      -     .   \n",
       "1867857       chrY   HAVANA    gene  84562571  84564907     .      -     .   \n",
       "1867918       chrY   HAVANA    gene  85528516  85530908     .      -     .   \n",
       "1868127       chrY   HAVANA    gene  89713423  89745532     .      -     .   \n",
       "1868197       chrY  ENSEMBL    gene  90838868  90839177     .      -     .   \n",
       "\n",
       "                       gene_id       gene_type  ... transcript_type  \\\n",
       "42       ENSMUSG00000025903.14  protein_coding  ...             NaN   \n",
       "191       ENSMUSG00000104217.1  protein_coding  ...             NaN   \n",
       "217      ENSMUSG00000033813.15  protein_coding  ...             NaN   \n",
       "276      ENSMUSG00000033793.12  protein_coding  ...             NaN   \n",
       "424      ENSMUSG00000025905.14  protein_coding  ...             NaN   \n",
       "...                        ...             ...  ...             ...   \n",
       "1867618   ENSMUSG00000095867.2  protein_coding  ...             NaN   \n",
       "1867857   ENSMUSG00000094660.2  protein_coding  ...             NaN   \n",
       "1867918   ENSMUSG00000095650.2  protein_coding  ...             NaN   \n",
       "1868127   ENSMUSG00000100608.1  protein_coding  ...             NaN   \n",
       "1868197   ENSMUSG00000096850.1  protein_coding  ...             NaN   \n",
       "\n",
       "        transcript_name transcript_support_level                tag  \\\n",
       "42                  NaN                      NaN  overlapping_locus   \n",
       "191                 NaN                      NaN  overlapping_locus   \n",
       "217                 NaN                      NaN  overlapping_locus   \n",
       "276                 NaN                      NaN                NaN   \n",
       "424                 NaN                      NaN                NaN   \n",
       "...                 ...                      ...                ...   \n",
       "1867618             NaN                      NaN                NaN   \n",
       "1867857             NaN                      NaN                NaN   \n",
       "1867918             NaN                      NaN                NaN   \n",
       "1868127             NaN                      NaN                NaN   \n",
       "1868197             NaN                      NaN                NaN   \n",
       "\n",
       "        havana_transcript exon_number exon_id protein_id ccdsid  ont  \n",
       "42                    NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "191                   NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "217                   NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "276                   NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "424                   NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "...                   ...         ...     ...        ...    ...  ...  \n",
       "1867618               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1867857               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1867918               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868127               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "1868197               NaN         NaN     NaN        NaN    NaN  NaN  \n",
       "\n",
       "[21849 rows x 25 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_codingdf=transcriptdf[transcriptdf['gene_type']=='protein_coding']\n",
    "protein_codingdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0f503ac5-32a0-4688-a404-3c090782ac10",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/2348304993.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  protein_codingdf['PAS']=np.where(protein_codingdf['Strand']=='+',protein_codingdf['End'],protein_codingdf['Start'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807787</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807891</td>\n",
       "      <td>4886770</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104217.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4857813</td>\n",
       "      <td>4897909</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033813.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5070017</td>\n",
       "      <td>5162529</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000033793.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>5588465</td>\n",
       "      <td>5606131</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000025905.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>79148788</td>\n",
       "      <td>79151122</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095867.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>84562571</td>\n",
       "      <td>84564907</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000094660.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>85528516</td>\n",
       "      <td>85530908</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000095650.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713423</td>\n",
       "      <td>89745532</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000100608.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000096850.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21849 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "42            chr1   HAVANA    gene   4807787   4848410     .      +     .   \n",
       "191           chr1   HAVANA    gene   4807891   4886770     .      +     .   \n",
       "217           chr1   HAVANA    gene   4857813   4897909     .      +     .   \n",
       "276           chr1   HAVANA    gene   5070017   5162529     .      +     .   \n",
       "424           chr1   HAVANA    gene   5588465   5606131     .      +     .   \n",
       "...            ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "1867618       chrY   HAVANA    gene  79148788  79151122     .      -     .   \n",
       "1867857       chrY   HAVANA    gene  84562571  84564907     .      -     .   \n",
       "1867918       chrY   HAVANA    gene  85528516  85530908     .      -     .   \n",
       "1868127       chrY   HAVANA    gene  89713423  89745532     .      -     .   \n",
       "1868197       chrY  ENSEMBL    gene  90838868  90839177     .      -     .   \n",
       "\n",
       "                       gene_id       gene_type  ... transcript_name  \\\n",
       "42       ENSMUSG00000025903.14  protein_coding  ...             NaN   \n",
       "191       ENSMUSG00000104217.1  protein_coding  ...             NaN   \n",
       "217      ENSMUSG00000033813.15  protein_coding  ...             NaN   \n",
       "276      ENSMUSG00000033793.12  protein_coding  ...             NaN   \n",
       "424      ENSMUSG00000025905.14  protein_coding  ...             NaN   \n",
       "...                        ...             ...  ...             ...   \n",
       "1867618   ENSMUSG00000095867.2  protein_coding  ...             NaN   \n",
       "1867857   ENSMUSG00000094660.2  protein_coding  ...             NaN   \n",
       "1867918   ENSMUSG00000095650.2  protein_coding  ...             NaN   \n",
       "1868127   ENSMUSG00000100608.1  protein_coding  ...             NaN   \n",
       "1868197   ENSMUSG00000096850.1  protein_coding  ...             NaN   \n",
       "\n",
       "        transcript_support_level                tag havana_transcript  \\\n",
       "42                           NaN  overlapping_locus               NaN   \n",
       "191                          NaN  overlapping_locus               NaN   \n",
       "217                          NaN  overlapping_locus               NaN   \n",
       "276                          NaN                NaN               NaN   \n",
       "424                          NaN                NaN               NaN   \n",
       "...                          ...                ...               ...   \n",
       "1867618                      NaN                NaN               NaN   \n",
       "1867857                      NaN                NaN               NaN   \n",
       "1867918                      NaN                NaN               NaN   \n",
       "1868127                      NaN                NaN               NaN   \n",
       "1868197                      NaN                NaN               NaN   \n",
       "\n",
       "        exon_number exon_id protein_id ccdsid  ont       PAS  \n",
       "42              NaN     NaN        NaN    NaN  NaN   4848410  \n",
       "191             NaN     NaN        NaN    NaN  NaN   4886770  \n",
       "217             NaN     NaN        NaN    NaN  NaN   4897909  \n",
       "276             NaN     NaN        NaN    NaN  NaN   5162529  \n",
       "424             NaN     NaN        NaN    NaN  NaN   5606131  \n",
       "...             ...     ...        ...    ...  ...       ...  \n",
       "1867618         NaN     NaN        NaN    NaN  NaN  79148788  \n",
       "1867857         NaN     NaN        NaN    NaN  NaN  84562571  \n",
       "1867918         NaN     NaN        NaN    NaN  NaN  85528516  \n",
       "1868127         NaN     NaN        NaN    NaN  NaN  89713423  \n",
       "1868197         NaN     NaN        NaN    NaN  NaN  90838868  \n",
       "\n",
       "[21849 rows x 26 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_codingdf['PAS']=np.where(protein_codingdf['Strand']=='+',protein_codingdf['End'],protein_codingdf['Start'])\n",
    "protein_codingdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "528ba605-4110-410b-aac6-7b8b1de9bee4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21849 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome Strand       PAS\n",
       "42            chr1      +   4848410\n",
       "191           chr1      +   4886770\n",
       "217           chr1      +   4897909\n",
       "276           chr1      +   5162529\n",
       "424           chr1      +   5606131\n",
       "...            ...    ...       ...\n",
       "1867618       chrY      -  79148788\n",
       "1867857       chrY      -  84562571\n",
       "1867918       chrY      -  85528516\n",
       "1868127       chrY      -  89713423\n",
       "1868197       chrY      -  90838868\n",
       "\n",
       "[21849 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_codingdf=protein_codingdf[['Chromosome','Strand','PAS']]\n",
    "protein_codingdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "744d465c-2282-47f3-bd1f-6f0efc7dc31a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/745903821.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  protein_codingdf['name']=protein_codingdf['Chromosome'].astype('str')+'_'+protein_codingdf['Strand'].astype('str')+'_'+protein_codingdf['PAS'].astype('str')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>79148788</td>\n",
       "      <td>chrY_-_79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>84562571</td>\n",
       "      <td>chrY_-_84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>85528516</td>\n",
       "      <td>chrY_-_85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>89713423</td>\n",
       "      <td>chrY_-_89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>90838868</td>\n",
       "      <td>chrY_-_90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21849 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome Strand       PAS             name\n",
       "42            chr1      +   4848410   chr1_+_4848410\n",
       "191           chr1      +   4886770   chr1_+_4886770\n",
       "217           chr1      +   4897909   chr1_+_4897909\n",
       "276           chr1      +   5162529   chr1_+_5162529\n",
       "424           chr1      +   5606131   chr1_+_5606131\n",
       "...            ...    ...       ...              ...\n",
       "1867618       chrY      -  79148788  chrY_-_79148788\n",
       "1867857       chrY      -  84562571  chrY_-_84562571\n",
       "1867918       chrY      -  85528516  chrY_-_85528516\n",
       "1868127       chrY      -  89713423  chrY_-_89713423\n",
       "1868197       chrY      -  90838868  chrY_-_90838868\n",
       "\n",
       "[21849 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_codingdf['name']=protein_codingdf['Chromosome'].astype('str')+'_'+protein_codingdf['Strand'].astype('str')+'_'+protein_codingdf['PAS'].astype('str')\n",
    "protein_codingdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bdca9e32-44b5-42bf-a42e-2a17078a24bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>79148788</td>\n",
       "      <td>chrY_-_79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>84562571</td>\n",
       "      <td>chrY_-_84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>85528516</td>\n",
       "      <td>chrY_-_85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>89713423</td>\n",
       "      <td>chrY_-_89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>90838868</td>\n",
       "      <td>chrY_-_90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21786 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome Strand       PAS             name\n",
       "42            chr1      +   4848410   chr1_+_4848410\n",
       "191           chr1      +   4886770   chr1_+_4886770\n",
       "217           chr1      +   4897909   chr1_+_4897909\n",
       "276           chr1      +   5162529   chr1_+_5162529\n",
       "424           chr1      +   5606131   chr1_+_5606131\n",
       "...            ...    ...       ...              ...\n",
       "1867618       chrY      -  79148788  chrY_-_79148788\n",
       "1867857       chrY      -  84562571  chrY_-_84562571\n",
       "1867918       chrY      -  85528516  chrY_-_85528516\n",
       "1868127       chrY      -  89713423  chrY_-_89713423\n",
       "1868197       chrY      -  90838868  chrY_-_90838868\n",
       "\n",
       "[21786 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptbed=protein_codingdf.drop_duplicates('name')\n",
    "transcriptbed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a7699c5a-a48e-4b4a-b313-9f19668ea0d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>79148788</td>\n",
       "      <td>chrY_-_79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>84562571</td>\n",
       "      <td>chrY_-_84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>85528516</td>\n",
       "      <td>chrY_-_85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>89713423</td>\n",
       "      <td>chrY_-_89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>90838868</td>\n",
       "      <td>chrY_-_90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21786 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand       PAS             name\n",
       "42       chr1      +   4848410   chr1_+_4848410\n",
       "191      chr1      +   4886770   chr1_+_4886770\n",
       "217      chr1      +   4897909   chr1_+_4897909\n",
       "276      chr1      +   5162529   chr1_+_5162529\n",
       "424      chr1      +   5606131   chr1_+_5606131\n",
       "...       ...    ...       ...              ...\n",
       "1867618  chrY      -  79148788  chrY_-_79148788\n",
       "1867857  chrY      -  84562571  chrY_-_84562571\n",
       "1867918  chrY      -  85528516  chrY_-_85528516\n",
       "1868127  chrY      -  89713423  chrY_-_89713423\n",
       "1868197  chrY      -  90838868  chrY_-_90838868\n",
       "\n",
       "[21786 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptbed.columns=['chr','strand','PAS','name']\n",
    "transcriptbed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ef1921d-ecbd-4e9b-89a0-2513264f84f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#transcriptbed.to_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/redo/filtered_gencode.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deda48a2-8c97-432c-adab-c2dc8d0b0ed5",
   "metadata": {},
   "source": [
    "## ************"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c89d637e-60d9-4fff-8d2e-46ed093674c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "26ae82b2-ada1-4bdd-a370-4f0b2f677b47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867618</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>79148788</td>\n",
       "      <td>chrY_-_79148788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867857</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>84562571</td>\n",
       "      <td>chrY_-_84562571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867918</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>85528516</td>\n",
       "      <td>chrY_-_85528516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868127</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>89713423</td>\n",
       "      <td>chrY_-_89713423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868197</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>90838868</td>\n",
       "      <td>chrY_-_90838868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21786 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand       PAS             name\n",
       "42       chr1      +   4848410   chr1_+_4848410\n",
       "191      chr1      +   4886770   chr1_+_4886770\n",
       "217      chr1      +   4897909   chr1_+_4897909\n",
       "276      chr1      +   5162529   chr1_+_5162529\n",
       "424      chr1      +   5606131   chr1_+_5606131\n",
       "...       ...    ...       ...              ...\n",
       "1867618  chrY      -  79148788  chrY_-_79148788\n",
       "1867857  chrY      -  84562571  chrY_-_84562571\n",
       "1867918  chrY      -  85528516  chrY_-_85528516\n",
       "1868127  chrY      -  89713423  chrY_-_89713423\n",
       "1868197  chrY      -  90838868  chrY_-_90838868\n",
       "\n",
       "[21786 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencodedf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/redo/filtered_gencode.tsv',delimiter='\\t',index_col=0)\n",
    "gencodedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "004690b8-3290-4273-9419-839989bb06fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3121185</td>\n",
       "      <td>chr1_-_3121185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3121186</td>\n",
       "      <td>chr1_-_3121186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3206893</td>\n",
       "      <td>chr1_-_3206893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3206894</td>\n",
       "      <td>chr1_-_3206894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3206895</td>\n",
       "      <td>chr1_-_3206895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394337</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>1028592</td>\n",
       "      <td>chrY_+_1028592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394358</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260704</td>\n",
       "      <td>chrY_-_1260704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394359</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260715</td>\n",
       "      <td>chrY_-_1260715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394360</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260716</td>\n",
       "      <td>chrY_-_1260716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394361</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260755</td>\n",
       "      <td>chrY_-_1260755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85108 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr strand      PAS            name\n",
       "1       chr1      -  3121185  chr1_-_3121185\n",
       "3       chr1      -  3121186  chr1_-_3121186\n",
       "12      chr1      -  3206893  chr1_-_3206893\n",
       "13      chr1      -  3206894  chr1_-_3206894\n",
       "14      chr1      -  3206895  chr1_-_3206895\n",
       "...      ...    ...      ...             ...\n",
       "394337  chrY      +  1028592  chrY_+_1028592\n",
       "394358  chrY      -  1260704  chrY_-_1260704\n",
       "394359  chrY      -  1260715  chrY_-_1260715\n",
       "394360  chrY      -  1260716  chrY_-_1260716\n",
       "394361  chrY      -  1260755  chrY_-_1260755\n",
       "\n",
       "[85108 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threedatasetdf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/redo/three_dataset_merge.tsv',delimiter='\\t',index_col=0)\n",
    "threedatasetdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ed3e89f4-5063-481a-b59e-efe38d43aab3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394337</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>1028592</td>\n",
       "      <td>chrY_+_1028592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394358</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260704</td>\n",
       "      <td>chrY_-_1260704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394359</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260715</td>\n",
       "      <td>chrY_-_1260715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394360</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260716</td>\n",
       "      <td>chrY_-_1260716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394361</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260755</td>\n",
       "      <td>chrY_-_1260755</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106894 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr strand      PAS            name\n",
       "42      chr1      +  4848410  chr1_+_4848410\n",
       "191     chr1      +  4886770  chr1_+_4886770\n",
       "217     chr1      +  4897909  chr1_+_4897909\n",
       "276     chr1      +  5162529  chr1_+_5162529\n",
       "424     chr1      +  5606131  chr1_+_5606131\n",
       "...      ...    ...      ...             ...\n",
       "394337  chrY      +  1028592  chrY_+_1028592\n",
       "394358  chrY      -  1260704  chrY_-_1260704\n",
       "394359  chrY      -  1260715  chrY_-_1260715\n",
       "394360  chrY      -  1260716  chrY_-_1260716\n",
       "394361  chrY      -  1260755  chrY_-_1260755\n",
       "\n",
       "[106894 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf=pd.concat([gencodedf,threedatasetdf],axis=0)\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d029c148-581b-490a-96dd-1bbecbb7ebad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394337</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>1028592</td>\n",
       "      <td>chrY_+_1028592</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394358</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260704</td>\n",
       "      <td>chrY_-_1260704</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394359</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260715</td>\n",
       "      <td>chrY_-_1260715</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394360</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260716</td>\n",
       "      <td>chrY_-_1260716</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394361</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>1260755</td>\n",
       "      <td>chrY_-_1260755</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106894 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr strand      PAS            name  label\n",
       "42      chr1      +  4848410  chr1_+_4848410      1\n",
       "191     chr1      +  4886770  chr1_+_4886770      1\n",
       "217     chr1      +  4897909  chr1_+_4897909      1\n",
       "276     chr1      +  5162529  chr1_+_5162529      1\n",
       "424     chr1      +  5606131  chr1_+_5606131      1\n",
       "...      ...    ...      ...             ...    ...\n",
       "394337  chrY      +  1028592  chrY_+_1028592      1\n",
       "394358  chrY      -  1260704  chrY_-_1260704      1\n",
       "394359  chrY      -  1260715  chrY_-_1260715      1\n",
       "394360  chrY      -  1260716  chrY_-_1260716      1\n",
       "394361  chrY      -  1260755  chrY_-_1260755      1\n",
       "\n",
       "[106894 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf['label']=1\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f8f6ccd9-9ca5-4f12-a06f-46895dc1f7f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8889964</td>\n",
       "      <td>8890164</td>\n",
       "      <td>chr1_8889964_8890164</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5603521</td>\n",
       "      <td>5603721</td>\n",
       "      <td>chr1_5603521_5603721</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5069121</td>\n",
       "      <td>5069321</td>\n",
       "      <td>chr1_5069121_5069321</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4362832</td>\n",
       "      <td>4363032</td>\n",
       "      <td>chr1_4362832_4363032</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7063750</td>\n",
       "      <td>7063950</td>\n",
       "      <td>chr1_7063750_7063950</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3579661</td>\n",
       "      <td>3579861</td>\n",
       "      <td>chr1_3579661_3579861</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2942026</td>\n",
       "      <td>2942226</td>\n",
       "      <td>chr1_2942026_2942226</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6506153</td>\n",
       "      <td>6506353</td>\n",
       "      <td>chr1_6506153_6506353</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5215228</td>\n",
       "      <td>5215428</td>\n",
       "      <td>chr1_5215228_5215428</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2405301</td>\n",
       "      <td>2405501</td>\n",
       "      <td>chr1_2405301_2405501</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92164 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0        1        2                     3  4  5\n",
       "0      chr1  8889964  8890164  chr1_8889964_8890164  1  +\n",
       "1      chr1  5603521  5603721  chr1_5603521_5603721  1  -\n",
       "2      chr1  5069121  5069321  chr1_5069121_5069321  1  +\n",
       "3      chr1  4362832  4363032  chr1_4362832_4363032  1  -\n",
       "4      chr1  7063750  7063950  chr1_7063750_7063950  1  +\n",
       "...     ...      ...      ...                   ... .. ..\n",
       "92159  chr1  3579661  3579861  chr1_3579661_3579861  1  -\n",
       "92160  chr1  2942026  2942226  chr1_2942026_2942226  1  +\n",
       "92161  chr1  6506153  6506353  chr1_6506153_6506353  1  -\n",
       "92162  chr1  5215228  5215428  chr1_5215228_5215428  1  -\n",
       "92163  chr1  2405301  2405501  chr1_2405301_2405501  1  +\n",
       "\n",
       "[92164 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model_redo/filtered_negative.bed',delimiter='\\t',header=None)\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0995eef5-dcc1-47d1-8a28-e69e52c75059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8889964</td>\n",
       "      <td>8890164</td>\n",
       "      <td>chr1_8889964_8890164</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>8890064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5603521</td>\n",
       "      <td>5603721</td>\n",
       "      <td>chr1_5603521_5603721</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>5603621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5069121</td>\n",
       "      <td>5069321</td>\n",
       "      <td>chr1_5069121_5069321</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>5069221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4362832</td>\n",
       "      <td>4363032</td>\n",
       "      <td>chr1_4362832_4363032</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>4362932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7063750</td>\n",
       "      <td>7063950</td>\n",
       "      <td>chr1_7063750_7063950</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>7063850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3579661</td>\n",
       "      <td>3579861</td>\n",
       "      <td>chr1_3579661_3579861</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2942026</td>\n",
       "      <td>2942226</td>\n",
       "      <td>chr1_2942026_2942226</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6506153</td>\n",
       "      <td>6506353</td>\n",
       "      <td>chr1_6506153_6506353</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5215228</td>\n",
       "      <td>5215428</td>\n",
       "      <td>chr1_5215228_5215428</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2405301</td>\n",
       "      <td>2405501</td>\n",
       "      <td>chr1_2405301_2405501</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92164 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0        1        2                     3  4  5      PAS\n",
       "0      chr1  8889964  8890164  chr1_8889964_8890164  1  +  8890064\n",
       "1      chr1  5603521  5603721  chr1_5603521_5603721  1  -  5603621\n",
       "2      chr1  5069121  5069321  chr1_5069121_5069321  1  +  5069221\n",
       "3      chr1  4362832  4363032  chr1_4362832_4363032  1  -  4362932\n",
       "4      chr1  7063750  7063950  chr1_7063750_7063950  1  +  7063850\n",
       "...     ...      ...      ...                   ... .. ..      ...\n",
       "92159  chr1  3579661  3579861  chr1_3579661_3579861  1  -  3579761\n",
       "92160  chr1  2942026  2942226  chr1_2942026_2942226  1  +  2942126\n",
       "92161  chr1  6506153  6506353  chr1_6506153_6506353  1  -  6506253\n",
       "92162  chr1  5215228  5215428  chr1_5215228_5215428  1  -  5215328\n",
       "92163  chr1  2405301  2405501  chr1_2405301_2405501  1  +  2405401\n",
       "\n",
       "[92164 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf['PAS']=intergenicdf[1]+100\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9a032ba0-f1ef-4045-bafd-aba207bfc368",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>8890064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5603621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5069221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>4362932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>7063850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92164 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0  5      PAS\n",
       "0      chr1  +  8890064\n",
       "1      chr1  -  5603621\n",
       "2      chr1  +  5069221\n",
       "3      chr1  -  4362932\n",
       "4      chr1  +  7063850\n",
       "...     ... ..      ...\n",
       "92159  chr1  -  3579761\n",
       "92160  chr1  +  2942126\n",
       "92161  chr1  -  6506253\n",
       "92162  chr1  -  5215328\n",
       "92163  chr1  +  2405401\n",
       "\n",
       "[92164 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf=intergenicdf[[0,5,'PAS']]\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5437c407-7f25-4d61-a24f-327b2d8be50a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/733091880.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  intergenicdf['label']=0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>8890064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5603621</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5069221</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>4362932</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>7063850</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92164 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0  5      PAS  label\n",
       "0      chr1  +  8890064      0\n",
       "1      chr1  -  5603621      0\n",
       "2      chr1  +  5069221      0\n",
       "3      chr1  -  4362932      0\n",
       "4      chr1  +  7063850      0\n",
       "...     ... ..      ...    ...\n",
       "92159  chr1  -  3579761      0\n",
       "92160  chr1  +  2942126      0\n",
       "92161  chr1  -  6506253      0\n",
       "92162  chr1  -  5215328      0\n",
       "92163  chr1  +  2405401      0\n",
       "\n",
       "[92164 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf['label']=0\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f6be9a67-38e3-41e9-8bc6-f2f2a3290197",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>8890064</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5603621</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5069221</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>4362932</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>7063850</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92164 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand      PAS  label\n",
       "0      chr1      +  8890064      0\n",
       "1      chr1      -  5603621      0\n",
       "2      chr1      +  5069221      0\n",
       "3      chr1      -  4362932      0\n",
       "4      chr1      +  7063850      0\n",
       "...     ...    ...      ...    ...\n",
       "92159  chr1      -  3579761      0\n",
       "92160  chr1      +  2942126      0\n",
       "92161  chr1      -  6506253      0\n",
       "92162  chr1      -  5215328      0\n",
       "92163  chr1      +  2405401      0\n",
       "\n",
       "[92164 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf.columns=['chr','strand','PAS','label']\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "723d9391-0332-4a11-9a4f-8abc4495a214",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>3073252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>3102015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3252756</td>\n",
       "      <td>3253236</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>3252756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3466586</td>\n",
       "      <td>3513553</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>3466586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3531794</td>\n",
       "      <td>3532720</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>3531794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142346</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90665345</td>\n",
       "      <td>90667625</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>90667625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142347</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90665345</td>\n",
       "      <td>90667625</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>90667625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142348</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90752426</td>\n",
       "      <td>90755467</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>90755467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142349</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90754512</td>\n",
       "      <td>90754821</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>90754821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142350</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>90838868</td>\n",
       "      <td>90839177</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>90839177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142351 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Feature     Start       End Score Strand       TSS\n",
       "0            chr1  transcript   3073252   3074322     .      +   3073252\n",
       "1            chr1  transcript   3102015   3102125     .      +   3102015\n",
       "2            chr1  transcript   3252756   3253236     .      +   3252756\n",
       "3            chr1  transcript   3466586   3513553     .      +   3466586\n",
       "4            chr1  transcript   3531794   3532720     .      +   3531794\n",
       "...           ...         ...       ...       ...   ...    ...       ...\n",
       "142346       chrY  transcript  90665345  90667625     .      -  90667625\n",
       "142347       chrY  transcript  90665345  90667625     .      -  90667625\n",
       "142348       chrY  transcript  90752426  90755467     .      -  90755467\n",
       "142349       chrY  transcript  90754512  90754821     .      -  90754821\n",
       "142350       chrY  transcript  90838868  90839177     .      -  90839177\n",
       "\n",
       "[142351 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model_redo/TSS_site_negative.bed',delimiter='\\t')\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "61fd2f4c-8b37-484d-b254-bb64b4020298",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>transcript</td>\n",
       "      <td>48670261</td>\n",
       "      <td>48717762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>94443324</td>\n",
       "      <td>94448708</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>transcript</td>\n",
       "      <td>36392243</td>\n",
       "      <td>36440048</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>transcript</td>\n",
       "      <td>74586802</td>\n",
       "      <td>74592002</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>19371439</td>\n",
       "      <td>19383284</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22904</th>\n",
       "      <td>chr3</td>\n",
       "      <td>transcript</td>\n",
       "      <td>152346477</td>\n",
       "      <td>152393612</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>152346477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80118</th>\n",
       "      <td>chr10</td>\n",
       "      <td>transcript</td>\n",
       "      <td>81233175</td>\n",
       "      <td>81237577</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>81233175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127220</th>\n",
       "      <td>chr18</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12128773</td>\n",
       "      <td>12130834</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>12128773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>chr2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>74855705</td>\n",
       "      <td>74856311</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>74855705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95707</th>\n",
       "      <td>chr12</td>\n",
       "      <td>transcript</td>\n",
       "      <td>109643338</td>\n",
       "      <td>109646763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>109643338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14730 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Feature      Start        End Score Strand        TSS\n",
       "105734      chr14  transcript   48670261   48717762     .      +   48670261\n",
       "21343        chr3  transcript   94443324   94448708     .      +   94443324\n",
       "132062      chr19  transcript   36392243   36440048     .      +   36392243\n",
       "35878        chr5  transcript   74586802   74592002     .      +   74586802\n",
       "9022         chr2  transcript   19371439   19383284     .      +   19371439\n",
       "...           ...         ...        ...        ...   ...    ...        ...\n",
       "22904        chr3  transcript  152346477  152393612     .      +  152346477\n",
       "80118       chr10  transcript   81233175   81237577     .      +   81233175\n",
       "127220      chr18  transcript   12128773   12130834     .      +   12128773\n",
       "10772        chr2  transcript   74855705   74856311     .      +   74855705\n",
       "95707       chr12  transcript  109643338  109646763     .      +  109643338\n",
       "\n",
       "[14730 rows x 7 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=TSSdf.sample(n=14730,random_state=666)\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "28602ee1-bd9c-4121-a3fc-25fb31678c75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22904</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>152346477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80118</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>81233175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127220</th>\n",
       "      <td>chr18</td>\n",
       "      <td>+</td>\n",
       "      <td>12128773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>74855705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95707</th>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>109643338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14730 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Strand        TSS\n",
       "105734      chr14      +   48670261\n",
       "21343        chr3      +   94443324\n",
       "132062      chr19      +   36392243\n",
       "35878        chr5      +   74586802\n",
       "9022         chr2      +   19371439\n",
       "...           ...    ...        ...\n",
       "22904        chr3      +  152346477\n",
       "80118       chr10      +   81233175\n",
       "127220      chr18      +   12128773\n",
       "10772        chr2      +   74855705\n",
       "95707       chr12      +  109643338\n",
       "\n",
       "[14730 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=TSSdf[['Chromosome','Strand','TSS']]\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b7d187f3-b403-427b-bb17-6165c1b2f31c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22904</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>152346477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80118</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>81233175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127220</th>\n",
       "      <td>chr18</td>\n",
       "      <td>+</td>\n",
       "      <td>12128773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>74855705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95707</th>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>109643338</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14730 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand        PAS\n",
       "105734  chr14      +   48670261\n",
       "21343    chr3      +   94443324\n",
       "132062  chr19      +   36392243\n",
       "35878    chr5      +   74586802\n",
       "9022     chr2      +   19371439\n",
       "...       ...    ...        ...\n",
       "22904    chr3      +  152346477\n",
       "80118   chr10      +   81233175\n",
       "127220  chr18      +   12128773\n",
       "10772    chr2      +   74855705\n",
       "95707   chr12      +  109643338\n",
       "\n",
       "[14730 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf.columns=['chr','strand','PAS']\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cdc3097a-be42-40da-9d08-26ff4cff1bdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/1272850224.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  TSSdf['label']=0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22904</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>152346477</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80118</th>\n",
       "      <td>chr10</td>\n",
       "      <td>+</td>\n",
       "      <td>81233175</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127220</th>\n",
       "      <td>chr18</td>\n",
       "      <td>+</td>\n",
       "      <td>12128773</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10772</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>74855705</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95707</th>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>109643338</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14730 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand        PAS  label\n",
       "105734  chr14      +   48670261      0\n",
       "21343    chr3      +   94443324      0\n",
       "132062  chr19      +   36392243      0\n",
       "35878    chr5      +   74586802      0\n",
       "9022     chr2      +   19371439      0\n",
       "...       ...    ...        ...    ...\n",
       "22904    chr3      +  152346477      0\n",
       "80118   chr10      +   81233175      0\n",
       "127220  chr18      +   12128773      0\n",
       "10772    chr2      +   74855705      0\n",
       "95707   chr12      +  109643338      0\n",
       "\n",
       "[14730 rows x 4 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf['label']=0\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "58ec3646-dbaf-499f-8a8c-bb4da82bdb4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106894 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand       PAS  label\n",
       "105734  chr14      +  48670261      0\n",
       "21343    chr3      +  94443324      0\n",
       "132062  chr19      +  36392243      0\n",
       "35878    chr5      +  74586802      0\n",
       "9022     chr2      +  19371439      0\n",
       "...       ...    ...       ...    ...\n",
       "92159    chr1      -   3579761      0\n",
       "92160    chr1      +   2942126      0\n",
       "92161    chr1      -   6506253      0\n",
       "92162    chr1      -   5215328      0\n",
       "92163    chr1      +   2405401      0\n",
       "\n",
       "[106894 rows x 4 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negativedf=pd.concat([TSSdf,intergenicdf],axis=0)\n",
    "negativedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0ed92f1d-f842-40ab-8a78-d23b54033e53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105734</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>48670261</td>\n",
       "      <td>0</td>\n",
       "      <td>chr14_+_48670261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21343</th>\n",
       "      <td>chr3</td>\n",
       "      <td>+</td>\n",
       "      <td>94443324</td>\n",
       "      <td>0</td>\n",
       "      <td>chr3_+_94443324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132062</th>\n",
       "      <td>chr19</td>\n",
       "      <td>+</td>\n",
       "      <td>36392243</td>\n",
       "      <td>0</td>\n",
       "      <td>chr19_+_36392243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35878</th>\n",
       "      <td>chr5</td>\n",
       "      <td>+</td>\n",
       "      <td>74586802</td>\n",
       "      <td>0</td>\n",
       "      <td>chr5_+_74586802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9022</th>\n",
       "      <td>chr2</td>\n",
       "      <td>+</td>\n",
       "      <td>19371439</td>\n",
       "      <td>0</td>\n",
       "      <td>chr2_+_19371439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_-_3579761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_2942126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_-_6506253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_-_5215328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_2405401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106894 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand       PAS  label              name\n",
       "105734  chr14      +  48670261      0  chr14_+_48670261\n",
       "21343    chr3      +  94443324      0   chr3_+_94443324\n",
       "132062  chr19      +  36392243      0  chr19_+_36392243\n",
       "35878    chr5      +  74586802      0   chr5_+_74586802\n",
       "9022     chr2      +  19371439      0   chr2_+_19371439\n",
       "...       ...    ...       ...    ...               ...\n",
       "92159    chr1      -   3579761      0    chr1_-_3579761\n",
       "92160    chr1      +   2942126      0    chr1_+_2942126\n",
       "92161    chr1      -   6506253      0    chr1_-_6506253\n",
       "92162    chr1      -   5215328      0    chr1_-_5215328\n",
       "92163    chr1      +   2405401      0    chr1_+_2405401\n",
       "\n",
       "[106894 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negativedf['name']=negativedf['chr']+'_'+negativedf['strand']+'_'+negativedf['PAS'].astype('str')\n",
    "negativedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d7ddb796-74d4-4f3a-a0f9-e876001cf5d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>chr1_-_3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>chr1_+_2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>chr1_-_6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>chr1_-_5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>chr1_+_2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213788 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand      PAS            name  label\n",
       "42     chr1      +  4848410  chr1_+_4848410      1\n",
       "191    chr1      +  4886770  chr1_+_4886770      1\n",
       "217    chr1      +  4897909  chr1_+_4897909      1\n",
       "276    chr1      +  5162529  chr1_+_5162529      1\n",
       "424    chr1      +  5606131  chr1_+_5606131      1\n",
       "...     ...    ...      ...             ...    ...\n",
       "92159  chr1      -  3579761  chr1_-_3579761      0\n",
       "92160  chr1      +  2942126  chr1_+_2942126      0\n",
       "92161  chr1      -  6506253  chr1_-_6506253      0\n",
       "92162  chr1      -  5215328  chr1_-_5215328      0\n",
       "92163  chr1      +  2405401  chr1_+_2405401      0\n",
       "\n",
       "[213788 rows x 5 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf=pd.concat([positivedf,negativedf],axis=0)\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "13a30772-0897-4beb-a0ab-6c37785c807d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92159</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>chr1_-_3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92160</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>chr1_+_2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92161</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>chr1_-_6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92162</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>chr1_-_5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92163</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>chr1_+_2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213787 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand      PAS            name  label\n",
       "42     chr1      +  4848410  chr1_+_4848410      1\n",
       "191    chr1      +  4886770  chr1_+_4886770      1\n",
       "217    chr1      +  4897909  chr1_+_4897909      1\n",
       "276    chr1      +  5162529  chr1_+_5162529      1\n",
       "424    chr1      +  5606131  chr1_+_5606131      1\n",
       "...     ...    ...      ...             ...    ...\n",
       "92159  chr1      -  3579761  chr1_-_3579761      0\n",
       "92160  chr1      +  2942126  chr1_+_2942126      0\n",
       "92161  chr1      -  6506253  chr1_-_6506253      0\n",
       "92162  chr1      -  5215328  chr1_-_5215328      0\n",
       "92163  chr1      +  2405401  chr1_+_2405401      0\n",
       "\n",
       "[213787 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf=alldf[alldf['PAS']>100]\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e359aa50-6dcd-4ff6-b141-430719dc433d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213782</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>chr1_-_3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213783</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>chr1_+_2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213784</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>chr1_-_6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213785</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>chr1_-_5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213786</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>chr1_+_2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213787 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr strand      PAS            name  label\n",
       "0       chr1      +  4848410  chr1_+_4848410      1\n",
       "1       chr1      +  4886770  chr1_+_4886770      1\n",
       "2       chr1      +  4897909  chr1_+_4897909      1\n",
       "3       chr1      +  5162529  chr1_+_5162529      1\n",
       "4       chr1      +  5606131  chr1_+_5606131      1\n",
       "...      ...    ...      ...             ...    ...\n",
       "213782  chr1      -  3579761  chr1_-_3579761      0\n",
       "213783  chr1      +  2942126  chr1_+_2942126      0\n",
       "213784  chr1      -  6506253  chr1_-_6506253      0\n",
       "213785  chr1      -  5215328  chr1_-_5215328      0\n",
       "213786  chr1      +  2405401  chr1_+_2405401      0\n",
       "\n",
       "[213787 rows x 5 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf.reset_index(inplace=True,drop=True)\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48c25087-8b92-41f8-82a5-a22e2c65eea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#alldf.to_csv('/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model_redo/final_all_sample_negative_positive_balance.tsv',sep='\\t',index=None)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64c316c4-1278-4729-8a84-55644118d10c",
   "metadata": {},
   "source": [
    "# train mouse model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "567282bf-cad4-4edd-8431-ebe5b4e62c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "from kipoiseq.transforms.functional import one_hot\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "from sklearn.model_selection import KFold\n",
    "from torch.utils.data import DataLoader\n",
    "from torch.utils.data import Dataset\n",
    "from tensorboardX import SummaryWriter\n",
    "from scipy import stats\n",
    "import torch.optim as optim\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import torch.nn as nn\n",
    "import pickle\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import h5py\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ef408b16-a0d5-4b19-b42e-f5907cacb3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def setup_seed(seed):\n",
    "     torch.manual_seed(seed)\n",
    "     torch.cuda.manual_seed_all(seed)\n",
    "     np.random.seed(seed)\n",
    "     random.seed(seed)\n",
    "     torch.backends.cudnn.deterministic = True\n",
    "# 设置随机数种子\n",
    "\n",
    "setup_seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a2c9c5a4-1564-4e2b-8ce8-5a17b4b730b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "class scDataset(Dataset):\n",
    "    def __init__(self,ref_fastq,bed,idx,train_idx,col_name):\n",
    "        \n",
    "        #read bed file\n",
    "        genes=Fasta(ref_fastq)\n",
    "        bedfile=pd.read_csv(bed,delimiter='\\t')\n",
    "        bedfile['start']=bedfile['PAS']-100\n",
    "        bedfile['end']=bedfile['PAS']+100\n",
    "        bedfile=bedfile[bedfile['start']>1]\n",
    "        #print(bedfile)\n",
    "        #extract sequence and do one-hot encoding \n",
    "        onehotdict={}\n",
    "        for chrom, start, end, strand, genename in tqdm(bedfile[['chr','start','end','strand','name']].values,desc='Loading Loci'):\n",
    "            if strand=='+':\n",
    "                seq_=genes.get_seq(chrom,start+1,end).seq\n",
    "            else:\n",
    "                seq_=genes.get_seq(chrom,start+1,end,rc=True).seq\n",
    "\n",
    "            onehotnp=one_hot(seq_,['A','C','T','G'])\n",
    "            onehotdict[genename]=onehotnp\n",
    "            \n",
    "        \n",
    "        #get y value \n",
    "        self.y_value=np.array(bedfile[col_name][idx])\n",
    "        #print(self.y_value)\n",
    "        \n",
    "        \n",
    "        # get corresponding X value \n",
    "        self.gn=bedfile['name'][idx].tolist()\n",
    "        #print(self.gn)\n",
    "        \n",
    "        onehotls=[]\n",
    "        for i in self.gn:\n",
    "            onehotls.append(onehotdict[i])\n",
    "        #print(onehotls)\n",
    "            \n",
    "        self.combine_array=np.stack(onehotls)\n",
    "        #print(self.combine_array)\n",
    "                \n",
    "    def __len__(self):\n",
    "        return len(self.y_value)\n",
    "    \n",
    "    \n",
    "    def __getitem__(self,index):\n",
    "        #print(self.combine_array.shape)\n",
    "        #print(index)\n",
    "        seq=np.transpose(self.combine_array[index]).astype(np.float32)\n",
    "        y_value=self.y_value[index].astype(np.float32)\n",
    "        TSS_name=self.gn[index]\n",
    "        \n",
    "        #print(TSS_name)\n",
    "\n",
    "        sample = {\"seq\": seq, \"y_true_value\": y_value,'TSS_name':TSS_name}\n",
    "        return sample \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "629e856a-1264-4778-9020-6cedd42caec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "\n",
    "# Validation\n",
    "\n",
    "def evaluate(device,net,dataloader,criterion):\n",
    "    net.eval()\n",
    "    avg_loss=0\n",
    "    TSS_name_ls=[]\n",
    "    gt_labels=[]\n",
    "    pred_labels=[]\n",
    "    y_scorels=[]\n",
    "    with torch.no_grad():\n",
    "        for data in dataloader:\n",
    "            seq=data['seq'].to(device,dtype=torch.float)\n",
    "\n",
    "            label=data['y_true_value'].to(device,dtype=torch.int64)\n",
    "            output=net(seq)\n",
    "            current_loss = criterion(output, label)\n",
    "            avg_loss+=current_loss.item() / len(dataloader) \n",
    "\n",
    "\n",
    "            gt_labels.append(label.cpu().data.numpy())\n",
    "            TSS_name=data['TSS_name']\n",
    "            TSS_name_ls.append(TSS_name[0])\n",
    "\n",
    "            preds = torch.argmax(output, 1)\n",
    "            pred_labels.append(preds.cpu().data.numpy())\n",
    "            \n",
    "            y_scorels.append(output[:,1].cpu().data.numpy())\n",
    "\n",
    "\n",
    "    return avg_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0b5a5cca-b516-4974-aca3-c36de92eeb53",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6eb09d6d-828a-4f80-ba2a-fdc9aca98bc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train\n",
    "\n",
    "def train(device,train_loader,validation_loader,net,criterion,learning_rate,lr_sched,num_epochs,logs_dir,model_dir):\n",
    "    logger = SummaryWriter(logs_dir)\n",
    "    #optimizer=optim.Adam(net.parameters(),lr=learning_rate)\n",
    "    optimizer = optim.SGD(net.parameters(), lr=0.003,momentum=0.8)\n",
    "    scheduler=optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', \n",
    "                                                     factor=0.1, patience=5)\n",
    "    best_loss=100\n",
    "    \n",
    "    for epoch in range(0,num_epochs):\n",
    "        net.train()\n",
    "        avg_train_loss=0\n",
    "        for data in train_loader:\n",
    "            seq=data['seq'].to(device)\n",
    "            # print(\"Training seq:\")\n",
    "            # print(seq)\n",
    "            #hl=data['hl'].to(device,dtype=torch.float)\n",
    "            y_true_value=data['y_true_value'].to(device,torch.int64)\n",
    "            #print(y_true_value)\n",
    "            \n",
    "\n",
    "        \n",
    "            \n",
    "            with torch.set_grad_enabled(True):\n",
    "                optimizer.zero_grad()\n",
    "                \n",
    "                #Forward pass\n",
    "                y_pred_value=net(seq)\n",
    "                #print(y_pred_value)\n",
    "                \n",
    "                current_loss = criterion(y_pred_value,y_true_value)\n",
    "                current_loss.backward()\n",
    "                optimizer.step()\n",
    "                avg_train_loss+=current_loss.item()/len(train_loader)\n",
    "                logger.add_scalar('train_loss_epoch', avg_train_loss, epoch + 1)\n",
    "                \n",
    "                \n",
    "        print(\"The average train loss is %f\" %avg_train_loss)\n",
    "        \n",
    "        \n",
    "        #get validation loss \n",
    "        validation_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls=evaluate(device,net,validation_loader,criterion)\n",
    "        \n",
    "        # print(gt_labels)\n",
    "        # print(y_scorels)\n",
    "        \n",
    "        auc=roc_auc_score(gt_labels, y_scorels)\n",
    "        print('Validation auc: %f'%auc)\n",
    "        \n",
    "        print(\"The average validation loss is %f\" %validation_loss)\n",
    "\n",
    "        logger.add_scalar('valid_loss_epoch', validation_loss, epoch + 1)\n",
    "        \n",
    "        \n",
    "        # store best model\n",
    "        if validation_loss<best_loss:\n",
    "            best_loss=validation_loss\n",
    "            state={'epoch':epoch+1,'state_dict':net.state_dict(),'optimizer':optimizer.state_dict(),'loss':criterion,'scheduler':scheduler.state_dict()}\n",
    "            torch.save(state,model_dir+'/model_last.pth.tar')\n",
    "            \n",
    "            \n",
    "        # LR scheduler on plateau (based on validation loss)\n",
    "        if lr_sched:\n",
    "            scheduler.step(validation_loss)\n",
    "            \n",
    "        print(\"Finish training.\")\n",
    "            \n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c8d04e4c-9d96-4d2a-b73c-b2c7ac2dbac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test\n",
    "\n",
    "def test(testdataloader,device,net,model_dir,save_file,fold_no):\n",
    "    \n",
    "    #get model\n",
    "    check_point=torch.load(model_dir+'/model_last.pth.tar')\n",
    "    net.load_state_dict(check_point['state_dict'])\n",
    "    criterion = check_point['loss']\n",
    "    \n",
    "    avg_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls=evaluate(device,net,testdataloader,criterion)\n",
    "    #y_true_value=np.array([i[0]for i in y_true_value])\n",
    "    auc_value=roc_auc_score(gt_labels, y_scorels)\n",
    "    # print(\"True y value: \\n\")\n",
    "    # print(y_true_value)\n",
    "    \n",
    "    # print(\"Pred y value: \\n\")\n",
    "    # print(y_pred_value)\n",
    "    \n",
    "    # Save predictions\n",
    "    if save_file is not None:\n",
    "        pickle.dump({'y_true': gt_labels, 'y_pred': pred_labels,'y_score':y_scorels,'TSS':TSS_name_ls}, open(save_file, 'wb'))\n",
    "\n",
    "    print(\"The test auc is %f\" %auc_value)\n",
    "    # print(\"Test loss is %f\"%test_loss)\n",
    "    # print(corr_value)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c3408f88-417a-4e71-a0ac-0babd22df50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model\n",
    "\n",
    "class Net(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Net, self).__init__()\n",
    "        self.conv = nn.Sequential(\n",
    "            nn.Conv1d(4, 128, 8),\n",
    "             nn.ReLU(),\n",
    "            nn.Conv1d(128,64,4),\n",
    "            nn.ReLU(),\n",
    "            nn.Conv1d(64,32,2),\n",
    "             nn.BatchNorm1d(32),\n",
    "              nn.MaxPool2d(2),\n",
    "            nn.Dropout(0.4),\n",
    "             nn.Flatten(),\n",
    "            nn.Linear(1504, 32),\n",
    "            nn.ReLU(),\n",
    "        )\n",
    "        self.fc = nn.Sequential(\n",
    "            nn.Linear(32, 2),\n",
    "            nn.Sigmoid()\n",
    "        )\n",
    "        \n",
    "    def forward(self, x):\n",
    "        x = self.conv(x)\n",
    "        #print(x.shape)\n",
    "        #combinex=torch.cat((x,extrafeature),dim=1)\n",
    "        #x = x.view(-1, 3040)\n",
    "        x = self.fc(x)\n",
    "        # x = nn.functional.normalize(x)\n",
    "        return x\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "369dfe9c-39f8-4891-95de-f0bf11ae1cdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Main\n",
    "\n",
    "#device=torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')\n",
    "device=1\n",
    "\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f35ca92f-6e7c-4158-b0c7-2203a9f7d4e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KFold(n_splits=10, random_state=666, shuffle=True)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kf=KFold(n_splits=10,random_state=666,shuffle=True)\n",
    "kf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0b254e44-3a36-4b16-a94a-25107064d76a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_fastq='/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode_mm10/GRCm38.primary_assembly.genome.fa'\n",
    "bed='/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model_redo/final_all_sample_negative_positive_balance.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b4dad797-2fd4-431c-b7ff-d9cbdd0f59b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>name</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>4848410</td>\n",
       "      <td>chr1_+_4848410</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>4886770</td>\n",
       "      <td>chr1_+_4886770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>4897909</td>\n",
       "      <td>chr1_+_4897909</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>5162529</td>\n",
       "      <td>chr1_+_5162529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>5606131</td>\n",
       "      <td>chr1_+_5606131</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>3579761</td>\n",
       "      <td>chr1_-_3579761</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>2942126</td>\n",
       "      <td>chr1_+_2942126</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>6506253</td>\n",
       "      <td>chr1_-_6506253</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>5215328</td>\n",
       "      <td>chr1_-_5215328</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>2405401</td>\n",
       "      <td>chr1_+_2405401</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213787 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     strand      PAS            name  label\n",
       "chr                                        \n",
       "chr1      +  4848410  chr1_+_4848410      1\n",
       "chr1      +  4886770  chr1_+_4886770      1\n",
       "chr1      +  4897909  chr1_+_4897909      1\n",
       "chr1      +  5162529  chr1_+_5162529      1\n",
       "chr1      +  5606131  chr1_+_5606131      1\n",
       "...     ...      ...             ...    ...\n",
       "chr1      -  3579761  chr1_-_3579761      0\n",
       "chr1      +  2942126  chr1_+_2942126      0\n",
       "chr1      -  6506253  chr1_-_6506253      0\n",
       "chr1      -  5215328  chr1_-_5215328      0\n",
       "chr1      +  2405401  chr1_+_2405401      0\n",
       "\n",
       "[213787 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beddf=pd.read_csv(bed,delimiter='\\t',index_col=0)\n",
    "beddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cb5a6a2-7b1d-4ed3-b661-4f6d7b2e74e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fold_no, (train_index, test_index) in enumerate(kf.split(beddf)):\n",
    "    print(\"this is fold number: %i\" %fold_no)\n",
    "    \n",
    "    print(len(train_index))\n",
    "    print(len(test_index))\n",
    "    \n",
    "\n",
    "    np.random.seed(666)\n",
    "    \n",
    "    idx_rest = np.random.permutation(len(beddf)-len(test_index))\n",
    "    \n",
    "    idx_val = train_index[idx_rest[:int(0.2*len(idx_rest))]]\n",
    "    idx_train = train_index[idx_rest[int(0.2*len(idx_rest)):]]\n",
    "    \n",
    "    print(\"train_id\")\n",
    "    print(idx_train[:10])\n",
    "    \n",
    "    print(\"validation_id\")\n",
    "    print(idx_val[:10])\n",
    "    \n",
    "    print(\"test_id\")\n",
    "    print(test_index[:10])\n",
    "    \n",
    "\n",
    "    train_set=scDataset(ref_fastq,bed,idx_train,idx_train,'label')\n",
    "    validation_set=scDataset(ref_fastq,bed,idx_val,idx_train,'label')\n",
    "    test_set=scDataset(ref_fastq,bed,test_index,idx_train,'label')\n",
    "    \n",
    "    train_loader = DataLoader(train_set, batch_size=128, shuffle=True)\n",
    "    val_loader = DataLoader(validation_set, batch_size=1, shuffle=False)\n",
    "    test_loader = DataLoader(test_set, batch_size=1, shuffle=False)\n",
    "    \n",
    "\n",
    "    model_dir='/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/model_save/model_dir'\n",
    "    log_dir='/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/model_save/log_dir'\n",
    "    save_dir='/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model/model_save/PAS_true_pred_y_'+str(fold_no)+'.pkl'\n",
    "    \n",
    "    \n",
    "    criterion  = nn.CrossEntropyLoss().to(device)\n",
    "    net = Net().to(device)\n",
    "    init_lr =0.0005\n",
    "    lr_sched = True\n",
    "    num_epoch=100\n",
    "    \n",
    "    train(device,train_loader,val_loader,net,criterion,init_lr,lr_sched,num_epoch,log_dir,model_dir)\n",
    "    \n",
    "    test(test_loader,device,net,model_dir,save_dir,fold_no)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46abb033-9b61-46ce-be1f-566a46a775d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "914e70ef-2f97-44eb-bfbf-4b120358864b",
   "metadata": {},
   "source": [
    "# Human model "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149b0b5f-44a6-496f-82b9-24776fff8719",
   "metadata": {},
   "source": [
    "## get positive samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cafd3eb8-54ae-495d-a315-bbb2852151be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "861c5727-dc64-4cf6-a21f-535179f37869",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16341</td>\n",
       "      <td>16541</td>\n",
       "      <td>chr1:16441-16442</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16443</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16452</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141373</td>\n",
       "      <td>141573</td>\n",
       "      <td>chr1:141473-141475</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>262620</td>\n",
       "      <td>262820</td>\n",
       "      <td>chr1:262720-262721</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57130187</td>\n",
       "      <td>57130387</td>\n",
       "      <td>chrY:57130287-57130289</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57194881</td>\n",
       "      <td>57195081</td>\n",
       "      <td>chrY:57194981-57194983</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57197235</td>\n",
       "      <td>57197435</td>\n",
       "      <td>chrY:57197335-57197337</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201227</td>\n",
       "      <td>57201427</td>\n",
       "      <td>chrY:57201327-57201329</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201228</td>\n",
       "      <td>57201428</td>\n",
       "      <td>chrY:57201328-57201330</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2                       3  4  5\n",
       "0       chr1     16341     16541        chr1:16441-16442  0  -\n",
       "1       chr1     16342     16542        chr1:16442-16443  0  -\n",
       "2       chr1     16342     16542        chr1:16442-16452  0  -\n",
       "3       chr1    141373    141573      chr1:141473-141475  0  -\n",
       "4       chr1    262620    262820      chr1:262720-262721  0  -\n",
       "...      ...       ...       ...                     ... .. ..\n",
       "251067  chrY  57130187  57130387  chrY:57130287-57130289  0  +\n",
       "251068  chrY  57194881  57195081  chrY:57194981-57194983  0  +\n",
       "251069  chrY  57197235  57197435  chrY:57197335-57197337  0  +\n",
       "251070  chrY  57201227  57201427  chrY:57201327-57201329  0  -\n",
       "251071  chrY  57201228  57201428  chrY:57201328-57201330  0  -\n",
       "\n",
       "[251072 rows x 6 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/stringent_annotation_upstream100_down100.bed',header=None,delimiter='\\t')\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b83ab2f2-02bc-4578-bf34-1a36cc569855",
   "metadata": {},
   "source": [
    "## get negative sample"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d12c851-5773-408b-bef4-41d31530793c",
   "metadata": {},
   "source": [
    "### get intergenic region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e21fce4f-8aaf-44f9-a052-ad962d9858b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "63d8f582-c5ef-40f5-93d4-82d1424ec738",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/ruiyanhou/nfs_share2/conda_env/envs/Py39/lib/python3.9/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424184</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004015123.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424185</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015124.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424186</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015125.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424187</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00004015126.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424188</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00004015127.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3424189 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "0             chr1  HAVANA        gene     11868     14409     .      +     .   \n",
       "1             chr1  HAVANA  transcript     11868     14409     .      +     .   \n",
       "2             chr1  HAVANA        exon     11868     12227     .      +     .   \n",
       "3             chr1  HAVANA        exon     12612     12721     .      +     .   \n",
       "4             chr1  HAVANA        exon     13220     14409     .      +     .   \n",
       "...            ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "3424184       chrY  HAVANA        exon  57214349  57214397     .      -     .   \n",
       "3424185       chrY  HAVANA        exon  57213879  57213964     .      -     .   \n",
       "3424186       chrY  HAVANA        exon  57213525  57213602     .      -     .   \n",
       "3424187       chrY  HAVANA        exon  57213203  57213357     .      -     .   \n",
       "3424188       chrY  HAVANA        exon  57212183  57213125     .      -     .   \n",
       "\n",
       "                   gene_id               gene_type  ...  \\\n",
       "0        ENSG00000290825.1                  lncRNA  ...   \n",
       "1        ENSG00000290825.1                  lncRNA  ...   \n",
       "2        ENSG00000290825.1                  lncRNA  ...   \n",
       "3        ENSG00000290825.1                  lncRNA  ...   \n",
       "4        ENSG00000290825.1                  lncRNA  ...   \n",
       "...                    ...                     ...  ...   \n",
       "3424184  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424185  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424186  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424187  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424188  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "\n",
       "        transcript_support_level     havana_transcript exon_number  \\\n",
       "0                            NaN                   NaN         NaN   \n",
       "1                              1  OTTHUMT00000362751.1         NaN   \n",
       "2                              1  OTTHUMT00000362751.1           1   \n",
       "3                              1  OTTHUMT00000362751.1           2   \n",
       "4                              1  OTTHUMT00000362751.1           3   \n",
       "...                          ...                   ...         ...   \n",
       "3424184                       NA  OTTHUMT00000058841.1           1   \n",
       "3424185                       NA  OTTHUMT00000058841.1           2   \n",
       "3424186                       NA  OTTHUMT00000058841.1           3   \n",
       "3424187                       NA  OTTHUMT00000058841.1           4   \n",
       "3424188                       NA  OTTHUMT00000058841.1           5   \n",
       "\n",
       "                   exon_id     hgnc_id           havana_gene          ont  \\\n",
       "0                      NaN         NaN                   NaN          NaN   \n",
       "1                      NaN         NaN                   NaN          NaN   \n",
       "2        ENSE00002234944.1         NaN                   NaN          NaN   \n",
       "3        ENSE00003582793.1         NaN                   NaN          NaN   \n",
       "4        ENSE00002312635.1         NaN                   NaN          NaN   \n",
       "...                    ...         ...                   ...          ...   \n",
       "3424184  ENSE00004015123.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424185  ENSE00004015124.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424186  ENSE00004015125.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424187  ENSE00004015126.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424188  ENSE00004015127.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "\n",
       "        protein_id ccdsid artif_dupl  \n",
       "0              NaN    NaN        NaN  \n",
       "1              NaN    NaN        NaN  \n",
       "2              NaN    NaN        NaN  \n",
       "3              NaN    NaN        NaN  \n",
       "4              NaN    NaN        NaN  \n",
       "...            ...    ...        ...  \n",
       "3424184        NaN    NaN        NaN  \n",
       "3424185        NaN    NaN        NaN  \n",
       "3424186        NaN    NaN        NaN  \n",
       "3424187        NaN    NaN        NaN  \n",
       "3424188        NaN    NaN        NaN  \n",
       "\n",
       "[3424189 rows x 26 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr=pr.read_gtf('/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode/gencode.v44.annotation.gtf')\n",
    "\n",
    "grdf=gr.df\n",
    "\n",
    "grdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "60b749a7-5208-46c7-90bd-de3415dff45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424184</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424185</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424186</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424187</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424188</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3424189 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Start       End\n",
       "0             chr1     11868     14409\n",
       "1             chr1     11868     14409\n",
       "2             chr1     11868     12227\n",
       "3             chr1     12612     12721\n",
       "4             chr1     13220     14409\n",
       "...            ...       ...       ...\n",
       "3424184       chrY  57214349  57214397\n",
       "3424185       chrY  57213879  57213964\n",
       "3424186       chrY  57213525  57213602\n",
       "3424187       chrY  57213203  57213357\n",
       "3424188       chrY  57212183  57213125\n",
       "\n",
       "[3424189 rows x 3 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beddf=grdf[['Chromosome','Start','End']]\n",
    "beddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca9d6f51-ef92-4bd0-aab5-61f7c5d0c409",
   "metadata": {},
   "outputs": [],
   "source": [
    "#beddf.to_csv('/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode/get_intergenic_region/annotated_region.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68d7e8d-8eb2-4088-a765-3f774a07263f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5f01acb6-ef1a-4c15-bb24-8a7b16dcf649",
   "metadata": {},
   "source": [
    "## get the length of each chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de97218e-3b7e-469a-bb2f-89da64f1eb89",
   "metadata": {},
   "source": [
    "samtools faidx GRCh38.primary_assembly.genome.fa\n",
    "cut -f1,2 GRCh38.primary_assembly.genome.fa.fai > ./get_intergenic_region/human.genome\n",
    "sort -k1,1 human.genome > sort_human.genome\n",
    "sort -k1,1 -k2,2n -k3,3n annotated_region.bed > annotated_sort_region.bed\n",
    "bedtools complement -i annotated_sort_region.bed  -g sort_human.genome >intergenic_region_human.bed \n",
    "bedtools random -g intergenic_region_human.bed -n 350000 -l 1 > negative_sample.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eacf163b-ef9a-43e9-a3af-47a6e33d9d0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3699c3d2-4db6-43c4-a16f-84cec02d1956",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "adc9f666-0f48-45a3-831a-deb090a577a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865988</td>\n",
       "      <td>25865989</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915713</td>\n",
       "      <td>102915714</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185513</td>\n",
       "      <td>134185514</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91025</td>\n",
       "      <td>91026</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36054050</td>\n",
       "      <td>36054051</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479634</td>\n",
       "      <td>24479635</td>\n",
       "      <td>349996</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210873</td>\n",
       "      <td>7210874</td>\n",
       "      <td>349997</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67696001</td>\n",
       "      <td>67696002</td>\n",
       "      <td>349998</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502892</td>\n",
       "      <td>32502893</td>\n",
       "      <td>349999</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532467</td>\n",
       "      <td>12532468</td>\n",
       "      <td>350000</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1          2       3  4  5\n",
       "0        chr4   25865988   25865989       1  1  -\n",
       "1        chr4  102915713  102915714       2  1  +\n",
       "2        chr4  134185513  134185514       3  1  -\n",
       "3        chr1      91025      91026       4  1  -\n",
       "4        chr6   36054050   36054051       5  1  -\n",
       "...       ...        ...        ...     ... .. ..\n",
       "349995   chr4   24479634   24479635  349996  1  -\n",
       "349996   chr5    7210873    7210874  349997  1  +\n",
       "349997   chr6   67696001   67696002  349998  1  +\n",
       "349998  chr13   32502892   32502893  349999  1  -\n",
       "349999   chr4   12532467   12532468  350000  1  -\n",
       "\n",
       "[350000 rows x 6 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv('/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode/get_intergenic_region/negative_sample.bed',header=None,delimiter='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "98e90a00-003a-476a-b36b-913c1a77a9a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865988</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915713</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185513</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91025</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36054050</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479634</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210873</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67696001</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502892</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532467</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1  4  5\n",
       "0        chr4   25865988  1  -\n",
       "1        chr4  102915713  1  +\n",
       "2        chr4  134185513  1  -\n",
       "3        chr1      91025  1  -\n",
       "4        chr6   36054050  1  -\n",
       "...       ...        ... .. ..\n",
       "349995   chr4   24479634  1  -\n",
       "349996   chr5    7210873  1  +\n",
       "349997   chr6   67696001  1  +\n",
       "349998  chr13   32502892  1  -\n",
       "349999   chr4   12532467  1  -\n",
       "\n",
       "[350000 rows x 4 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=df[[0,1,4,5]]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d4b570e3-35c4-4081-9e67-6756787ec557",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/3160653307.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['start']=df[1]-100\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865988</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915713</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185513</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91025</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36054050</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479634</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>24479534</td>\n",
       "      <td>24479734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210873</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>7210773</td>\n",
       "      <td>7210973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67696001</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>67695901</td>\n",
       "      <td>67696101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502892</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>32502792</td>\n",
       "      <td>32502992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532467</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>12532367</td>\n",
       "      <td>12532567</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1  4  5      start        end\n",
       "0        chr4   25865988  1  -   25865888   25866088\n",
       "1        chr4  102915713  1  +  102915613  102915813\n",
       "2        chr4  134185513  1  -  134185413  134185613\n",
       "3        chr1      91025  1  -      90925      91125\n",
       "4        chr6   36054050  1  -   36053950   36054150\n",
       "...       ...        ... .. ..        ...        ...\n",
       "349995   chr4   24479634  1  -   24479534   24479734\n",
       "349996   chr5    7210873  1  +    7210773    7210973\n",
       "349997   chr6   67696001  1  +   67695901   67696101\n",
       "349998  chr13   32502892  1  -   32502792   32502992\n",
       "349999   chr4   12532467  1  -   12532367   12532567\n",
       "\n",
       "[350000 rows x 6 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['start']=df[1]-100\n",
    "df['end']=df[1]+100\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "cf3c5b43-e987-4b27-93ea-1de30ec1fef3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865988</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "      <td>chr4_25865888_25866088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915713</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "      <td>chr4_102915613_102915813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185513</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "      <td>chr4_134185413_134185613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91025</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "      <td>chr1_90925_91125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36054050</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "      <td>chr6_36053950_36054150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479634</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>24479534</td>\n",
       "      <td>24479734</td>\n",
       "      <td>chr4_24479534_24479734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210873</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>7210773</td>\n",
       "      <td>7210973</td>\n",
       "      <td>chr5_7210773_7210973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67696001</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>67695901</td>\n",
       "      <td>67696101</td>\n",
       "      <td>chr6_67695901_67696101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502892</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>32502792</td>\n",
       "      <td>32502992</td>\n",
       "      <td>chr13_32502792_32502992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532467</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>12532367</td>\n",
       "      <td>12532567</td>\n",
       "      <td>chr4_12532367_12532567</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350000 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1  4  5      start        end                      name\n",
       "0        chr4   25865988  1  -   25865888   25866088    chr4_25865888_25866088\n",
       "1        chr4  102915713  1  +  102915613  102915813  chr4_102915613_102915813\n",
       "2        chr4  134185513  1  -  134185413  134185613  chr4_134185413_134185613\n",
       "3        chr1      91025  1  -      90925      91125          chr1_90925_91125\n",
       "4        chr6   36054050  1  -   36053950   36054150    chr6_36053950_36054150\n",
       "...       ...        ... .. ..        ...        ...                       ...\n",
       "349995   chr4   24479634  1  -   24479534   24479734    chr4_24479534_24479734\n",
       "349996   chr5    7210873  1  +    7210773    7210973      chr5_7210773_7210973\n",
       "349997   chr6   67696001  1  +   67695901   67696101    chr6_67695901_67696101\n",
       "349998  chr13   32502892  1  -   32502792   32502992   chr13_32502792_32502992\n",
       "349999   chr4   12532467  1  -   12532367   12532567    chr4_12532367_12532567\n",
       "\n",
       "[350000 rows x 7 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['name']=df[0]+'_'+df['start'].astype('str')+'_'+df['end'].astype('str')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f7992753-ad14-4088-af20-fe88794dde91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "      <td>chr4_25865888_25866088</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "      <td>chr4_102915613_102915813</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "      <td>chr4_134185413_134185613</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "      <td>chr1_90925_91125</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "      <td>chr6_36053950_36054150</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479534</td>\n",
       "      <td>24479734</td>\n",
       "      <td>chr4_24479534_24479734</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210773</td>\n",
       "      <td>7210973</td>\n",
       "      <td>chr5_7210773_7210973</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67695901</td>\n",
       "      <td>67696101</td>\n",
       "      <td>chr6_67695901_67696101</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502792</td>\n",
       "      <td>32502992</td>\n",
       "      <td>chr13_32502792_32502992</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532367</td>\n",
       "      <td>12532567</td>\n",
       "      <td>chr4_12532367_12532567</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0      start        end                      name  4  5\n",
       "0        chr4   25865888   25866088    chr4_25865888_25866088  1  -\n",
       "1        chr4  102915613  102915813  chr4_102915613_102915813  1  +\n",
       "2        chr4  134185413  134185613  chr4_134185413_134185613  1  -\n",
       "3        chr1      90925      91125          chr1_90925_91125  1  -\n",
       "4        chr6   36053950   36054150    chr6_36053950_36054150  1  -\n",
       "...       ...        ...        ...                       ... .. ..\n",
       "349995   chr4   24479534   24479734    chr4_24479534_24479734  1  -\n",
       "349996   chr5    7210773    7210973      chr5_7210773_7210973  1  +\n",
       "349997   chr6   67695901   67696101    chr6_67695901_67696101  1  +\n",
       "349998  chr13   32502792   32502992   chr13_32502792_32502992  1  -\n",
       "349999   chr4   12532367   12532567    chr4_12532367_12532567  1  -\n",
       "\n",
       "[350000 rows x 6 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beddf=df[[0,'start','end','name',4,5]]\n",
    "beddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5a30fcf2-b328-478a-a90a-f54fa61aa6a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "      <td>chr4_25865888_25866088</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "      <td>chr4_102915613_102915813</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "      <td>chr4_134185413_134185613</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "      <td>chr1_90925_91125</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "      <td>chr6_36053950_36054150</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349995</th>\n",
       "      <td>chr4</td>\n",
       "      <td>24479534</td>\n",
       "      <td>24479734</td>\n",
       "      <td>chr4_24479534_24479734</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349996</th>\n",
       "      <td>chr5</td>\n",
       "      <td>7210773</td>\n",
       "      <td>7210973</td>\n",
       "      <td>chr5_7210773_7210973</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349997</th>\n",
       "      <td>chr6</td>\n",
       "      <td>67695901</td>\n",
       "      <td>67696101</td>\n",
       "      <td>chr6_67695901_67696101</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349998</th>\n",
       "      <td>chr13</td>\n",
       "      <td>32502792</td>\n",
       "      <td>32502992</td>\n",
       "      <td>chr13_32502792_32502992</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349999</th>\n",
       "      <td>chr4</td>\n",
       "      <td>12532367</td>\n",
       "      <td>12532567</td>\n",
       "      <td>chr4_12532367_12532567</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>349637 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0      start        end                      name  4  5\n",
       "0        chr4   25865888   25866088    chr4_25865888_25866088  1  -\n",
       "1        chr4  102915613  102915813  chr4_102915613_102915813  1  +\n",
       "2        chr4  134185413  134185613  chr4_134185413_134185613  1  -\n",
       "3        chr1      90925      91125          chr1_90925_91125  1  -\n",
       "4        chr6   36053950   36054150    chr6_36053950_36054150  1  -\n",
       "...       ...        ...        ...                       ... .. ..\n",
       "349995   chr4   24479534   24479734    chr4_24479534_24479734  1  -\n",
       "349996   chr5    7210773    7210973      chr5_7210773_7210973  1  +\n",
       "349997   chr6   67695901   67696101    chr6_67695901_67696101  1  +\n",
       "349998  chr13   32502792   32502992   chr13_32502792_32502992  1  -\n",
       "349999   chr4   12532367   12532567    chr4_12532367_12532567  1  -\n",
       "\n",
       "[349637 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beddf.drop_duplicates('name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89bb4ccb-b61f-4322-b5aa-cd8ee46bcefc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#beddf.to_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/unfiltered_negative.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2bc26b1-0ccf-4e19-898e-a57cfd82b224",
   "metadata": {},
   "source": [
    "## remove overlap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ee2c01-de03-40c1-9227-4ec8be2e5cf6",
   "metadata": {},
   "source": [
    "bedtools intersect -a unfiltered_negative.bed -b SCAPTURE_total_annotation.bed -v > filtered_negative.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc34305-cf03-44e8-8b4c-3bbbdaf4d7b8",
   "metadata": {},
   "source": [
    "## add TSS as complementary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "40d5e595-23a3-4a1b-93b2-c73bc8be2810",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "31500eac-c13e-4b75-8c99-0ba2d33cdb9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/ruiyanhou/nfs_share2/conda_env/envs/Py39/lib/python3.9/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424184</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57214349</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00004015123.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424185</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213879</td>\n",
       "      <td>57213964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00004015124.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424186</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213525</td>\n",
       "      <td>57213602</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00004015125.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424187</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57213203</td>\n",
       "      <td>57213357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSE00004015126.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424188</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57213125</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSE00004015127.1</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3424189 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "0             chr1  HAVANA        gene     11868     14409     .      +     .   \n",
       "1             chr1  HAVANA  transcript     11868     14409     .      +     .   \n",
       "2             chr1  HAVANA        exon     11868     12227     .      +     .   \n",
       "3             chr1  HAVANA        exon     12612     12721     .      +     .   \n",
       "4             chr1  HAVANA        exon     13220     14409     .      +     .   \n",
       "...            ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "3424184       chrY  HAVANA        exon  57214349  57214397     .      -     .   \n",
       "3424185       chrY  HAVANA        exon  57213879  57213964     .      -     .   \n",
       "3424186       chrY  HAVANA        exon  57213525  57213602     .      -     .   \n",
       "3424187       chrY  HAVANA        exon  57213203  57213357     .      -     .   \n",
       "3424188       chrY  HAVANA        exon  57212183  57213125     .      -     .   \n",
       "\n",
       "                   gene_id               gene_type  ...  \\\n",
       "0        ENSG00000290825.1                  lncRNA  ...   \n",
       "1        ENSG00000290825.1                  lncRNA  ...   \n",
       "2        ENSG00000290825.1                  lncRNA  ...   \n",
       "3        ENSG00000290825.1                  lncRNA  ...   \n",
       "4        ENSG00000290825.1                  lncRNA  ...   \n",
       "...                    ...                     ...  ...   \n",
       "3424184  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424185  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424186  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424187  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "3424188  ENSG00000292371.1  unprocessed_pseudogene  ...   \n",
       "\n",
       "        transcript_support_level     havana_transcript exon_number  \\\n",
       "0                            NaN                   NaN         NaN   \n",
       "1                              1  OTTHUMT00000362751.1         NaN   \n",
       "2                              1  OTTHUMT00000362751.1           1   \n",
       "3                              1  OTTHUMT00000362751.1           2   \n",
       "4                              1  OTTHUMT00000362751.1           3   \n",
       "...                          ...                   ...         ...   \n",
       "3424184                       NA  OTTHUMT00000058841.1           1   \n",
       "3424185                       NA  OTTHUMT00000058841.1           2   \n",
       "3424186                       NA  OTTHUMT00000058841.1           3   \n",
       "3424187                       NA  OTTHUMT00000058841.1           4   \n",
       "3424188                       NA  OTTHUMT00000058841.1           5   \n",
       "\n",
       "                   exon_id     hgnc_id           havana_gene          ont  \\\n",
       "0                      NaN         NaN                   NaN          NaN   \n",
       "1                      NaN         NaN                   NaN          NaN   \n",
       "2        ENSE00002234944.1         NaN                   NaN          NaN   \n",
       "3        ENSE00003582793.1         NaN                   NaN          NaN   \n",
       "4        ENSE00002312635.1         NaN                   NaN          NaN   \n",
       "...                    ...         ...                   ...          ...   \n",
       "3424184  ENSE00004015123.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424185  ENSE00004015124.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424186  ENSE00004015125.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424187  ENSE00004015126.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "3424188  ENSE00004015127.1  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005   \n",
       "\n",
       "        protein_id ccdsid artif_dupl  \n",
       "0              NaN    NaN        NaN  \n",
       "1              NaN    NaN        NaN  \n",
       "2              NaN    NaN        NaN  \n",
       "3              NaN    NaN        NaN  \n",
       "4              NaN    NaN        NaN  \n",
       "...            ...    ...        ...  \n",
       "3424184        NaN    NaN        NaN  \n",
       "3424185        NaN    NaN        NaN  \n",
       "3424186        NaN    NaN        NaN  \n",
       "3424187        NaN    NaN        NaN  \n",
       "3424188        NaN    NaN        NaN  \n",
       "\n",
       "[3424189 rows x 26 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gr=pr.read_gtf('/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode/gencode.v44.annotation.gtf')\n",
    "grdf=gr.df\n",
    "grdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "a666badf-fa05-44b4-8478-9fd9cb6dc9ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30266</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>OTTHUMT00000002841.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424170</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292364.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058816.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:460</td>\n",
       "      <td>OTTHUMG00000022673.2</td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424173</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292367.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000097437.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:38160</td>\n",
       "      <td>OTTHUMG00000040493.1</td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424176</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292368.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058849.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:23270</td>\n",
       "      <td>OTTHUMG00000022681.1</td>\n",
       "      <td>PGO:0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424179</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292370.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000058824.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:38513</td>\n",
       "      <td>OTTHUMG00000022676.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424183</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000292371.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>OTTHUMT00000058841.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37115</td>\n",
       "      <td>OTTHUMG00000022678.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252835 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome   Source     Feature     Start       End Score Strand  \\\n",
       "1             chr1   HAVANA  transcript     11868     14409     .      +   \n",
       "6             chr1   HAVANA  transcript     12009     13670     .      +   \n",
       "14            chr1   HAVANA  transcript     29553     31097     .      +   \n",
       "18            chr1   HAVANA  transcript     30266     31109     .      +   \n",
       "22            chr1  ENSEMBL  transcript     30365     30503     .      +   \n",
       "...            ...      ...         ...       ...       ...   ...    ...   \n",
       "3424170       chrY   HAVANA  transcript  57015104  57016096     .      -   \n",
       "3424173       chrY   HAVANA  transcript  57165511  57165845     .      -   \n",
       "3424176       chrY   HAVANA  transcript  57171889  57172769     .      -   \n",
       "3424179       chrY   HAVANA  transcript  57201142  57203357     .      -   \n",
       "3424183       chrY   HAVANA  transcript  57212183  57214397     .      -   \n",
       "\n",
       "        Frame            gene_id                           gene_type  ...  \\\n",
       "1           .  ENSG00000290825.1                              lncRNA  ...   \n",
       "6           .  ENSG00000223972.6  transcribed_unprocessed_pseudogene  ...   \n",
       "14          .  ENSG00000243485.5                              lncRNA  ...   \n",
       "18          .  ENSG00000243485.5                              lncRNA  ...   \n",
       "22          .  ENSG00000284332.1                               miRNA  ...   \n",
       "...       ...                ...                                 ...  ...   \n",
       "3424170     .  ENSG00000292364.1                processed_pseudogene  ...   \n",
       "3424173     .  ENSG00000292367.1                processed_pseudogene  ...   \n",
       "3424176     .  ENSG00000292368.1                processed_pseudogene  ...   \n",
       "3424179     .  ENSG00000292370.1                              lncRNA  ...   \n",
       "3424183     .  ENSG00000292371.1              unprocessed_pseudogene  ...   \n",
       "\n",
       "        transcript_support_level     havana_transcript exon_number exon_id  \\\n",
       "1                              1  OTTHUMT00000362751.1         NaN     NaN   \n",
       "6                             NA  OTTHUMT00000002844.2         NaN     NaN   \n",
       "14                             5  OTTHUMT00000002840.1         NaN     NaN   \n",
       "18                             5  OTTHUMT00000002841.2         NaN     NaN   \n",
       "22                            NA                   NaN         NaN     NaN   \n",
       "...                          ...                   ...         ...     ...   \n",
       "3424170                       NA  OTTHUMT00000058816.2         NaN     NaN   \n",
       "3424173                       NA  OTTHUMT00000097437.1         NaN     NaN   \n",
       "3424176                       NA  OTTHUMT00000058849.1         NaN     NaN   \n",
       "3424179                        1  OTTHUMT00000058824.3         NaN     NaN   \n",
       "3424183                       NA  OTTHUMT00000058841.1         NaN     NaN   \n",
       "\n",
       "            hgnc_id           havana_gene          ont protein_id ccdsid  \\\n",
       "1               NaN                   NaN          NaN        NaN    NaN   \n",
       "6        HGNC:37102  OTTHUMG00000000961.2  PGO:0000019        NaN    NaN   \n",
       "14       HGNC:52482  OTTHUMG00000000959.2          NaN        NaN    NaN   \n",
       "18       HGNC:52482  OTTHUMG00000000959.2          NaN        NaN    NaN   \n",
       "22       HGNC:35294                   NaN          NaN        NaN    NaN   \n",
       "...             ...                   ...          ...        ...    ...   \n",
       "3424170    HGNC:460  OTTHUMG00000022673.2  PGO:0000004        NaN    NaN   \n",
       "3424173  HGNC:38160  OTTHUMG00000040493.1  PGO:0000004        NaN    NaN   \n",
       "3424176  HGNC:23270  OTTHUMG00000022681.1  PGO:0000004        NaN    NaN   \n",
       "3424179  HGNC:38513  OTTHUMG00000022676.3          NaN        NaN    NaN   \n",
       "3424183  HGNC:37115  OTTHUMG00000022678.1  PGO:0000005        NaN    NaN   \n",
       "\n",
       "        artif_dupl  \n",
       "1              NaN  \n",
       "6              NaN  \n",
       "14             NaN  \n",
       "18             NaN  \n",
       "22             NaN  \n",
       "...            ...  \n",
       "3424170        NaN  \n",
       "3424173        NaN  \n",
       "3424176        NaN  \n",
       "3424179        NaN  \n",
       "3424183        NaN  \n",
       "\n",
       "[252835 rows x 26 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptdf=grdf[grdf['Feature']=='transcript']\n",
    "transcriptdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6c73be28-7b52-4789-b51b-3bc03167a721",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30266</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424170</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424173</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424176</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424179</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424183</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252835 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Feature     Start       End Score Strand\n",
       "1             chr1  transcript     11868     14409     .      +\n",
       "6             chr1  transcript     12009     13670     .      +\n",
       "14            chr1  transcript     29553     31097     .      +\n",
       "18            chr1  transcript     30266     31109     .      +\n",
       "22            chr1  transcript     30365     30503     .      +\n",
       "...            ...         ...       ...       ...   ...    ...\n",
       "3424170       chrY  transcript  57015104  57016096     .      -\n",
       "3424173       chrY  transcript  57165511  57165845     .      -\n",
       "3424176       chrY  transcript  57171889  57172769     .      -\n",
       "3424179       chrY  transcript  57201142  57203357     .      -\n",
       "3424183       chrY  transcript  57212183  57214397     .      -\n",
       "\n",
       "[252835 rows x 6 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptdf=transcriptdf[['Chromosome','Feature','Start','End','Score','Strand']]\n",
    "transcriptdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "115231a2-4d63-453d-80c2-326f8d42c6f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/2128785414.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcriptdf['TSS']=np.where(transcriptdf['Strand']=='+',transcriptdf['Start'],transcriptdf['End'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>11868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>12009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>29553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30266</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>30266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>30365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424170</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57016096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424173</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57165845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424176</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57172769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424179</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57203357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3424183</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57214397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252835 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome     Feature     Start       End Score Strand       TSS\n",
       "1             chr1  transcript     11868     14409     .      +     11868\n",
       "6             chr1  transcript     12009     13670     .      +     12009\n",
       "14            chr1  transcript     29553     31097     .      +     29553\n",
       "18            chr1  transcript     30266     31109     .      +     30266\n",
       "22            chr1  transcript     30365     30503     .      +     30365\n",
       "...            ...         ...       ...       ...   ...    ...       ...\n",
       "3424170       chrY  transcript  57015104  57016096     .      -  57016096\n",
       "3424173       chrY  transcript  57165511  57165845     .      -  57165845\n",
       "3424176       chrY  transcript  57171889  57172769     .      -  57172769\n",
       "3424179       chrY  transcript  57201142  57203357     .      -  57203357\n",
       "3424183       chrY  transcript  57212183  57214397     .      -  57214397\n",
       "\n",
       "[252835 rows x 7 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptdf['TSS']=np.where(transcriptdf['Strand']=='+',transcriptdf['Start'],transcriptdf['End'])\n",
    "transcriptdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dbfdd71-4c41-43f1-b5a8-5924cab78bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#transcriptdf.to_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/TSS_site_negative.bed',index=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f205c45-6204-4e50-8be9-885b490a84a2",
   "metadata": {},
   "source": [
    "# Combine all positive and negative samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "fb87cc44-a44b-4e52-9696-e1a5510552be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "5f267a40-e092-46ab-84c2-9ecce15be0b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16341</td>\n",
       "      <td>16541</td>\n",
       "      <td>chr1:16441-16442</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16443</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16452</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141373</td>\n",
       "      <td>141573</td>\n",
       "      <td>chr1:141473-141475</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>262620</td>\n",
       "      <td>262820</td>\n",
       "      <td>chr1:262720-262721</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57130187</td>\n",
       "      <td>57130387</td>\n",
       "      <td>chrY:57130287-57130289</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57194881</td>\n",
       "      <td>57195081</td>\n",
       "      <td>chrY:57194981-57194983</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57197235</td>\n",
       "      <td>57197435</td>\n",
       "      <td>chrY:57197335-57197337</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201227</td>\n",
       "      <td>57201427</td>\n",
       "      <td>chrY:57201327-57201329</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201228</td>\n",
       "      <td>57201428</td>\n",
       "      <td>chrY:57201328-57201330</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2                       3  4  5\n",
       "0       chr1     16341     16541        chr1:16441-16442  0  -\n",
       "1       chr1     16342     16542        chr1:16442-16443  0  -\n",
       "2       chr1     16342     16542        chr1:16442-16452  0  -\n",
       "3       chr1    141373    141573      chr1:141473-141475  0  -\n",
       "4       chr1    262620    262820      chr1:262720-262721  0  -\n",
       "...      ...       ...       ...                     ... .. ..\n",
       "251067  chrY  57130187  57130387  chrY:57130287-57130289  0  +\n",
       "251068  chrY  57194881  57195081  chrY:57194981-57194983  0  +\n",
       "251069  chrY  57197235  57197435  chrY:57197335-57197337  0  +\n",
       "251070  chrY  57201227  57201427  chrY:57201327-57201329  0  -\n",
       "251071  chrY  57201228  57201428  chrY:57201328-57201330  0  -\n",
       "\n",
       "[251072 rows x 6 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/stringent_annotation_upstream100_down100.bed',header=None,delimiter='\\t')\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "232e8b4a-51a0-4a28-b22b-bad31b523f1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16341</td>\n",
       "      <td>16541</td>\n",
       "      <td>chr1:16441-16442</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16443</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16342</td>\n",
       "      <td>16542</td>\n",
       "      <td>chr1:16442-16452</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>141373</td>\n",
       "      <td>141573</td>\n",
       "      <td>chr1:141473-141475</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>262620</td>\n",
       "      <td>262820</td>\n",
       "      <td>chr1:262720-262721</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57130187</td>\n",
       "      <td>57130387</td>\n",
       "      <td>chrY:57130287-57130289</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>57130287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57194881</td>\n",
       "      <td>57195081</td>\n",
       "      <td>chrY:57194981-57194983</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>57194981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57197235</td>\n",
       "      <td>57197435</td>\n",
       "      <td>chrY:57197335-57197337</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>57197335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201227</td>\n",
       "      <td>57201427</td>\n",
       "      <td>chrY:57201327-57201329</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>57201327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201228</td>\n",
       "      <td>57201428</td>\n",
       "      <td>chrY:57201328-57201330</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>57201328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2                       3  4  5       PAS\n",
       "0       chr1     16341     16541        chr1:16441-16442  0  -     16441\n",
       "1       chr1     16342     16542        chr1:16442-16443  0  -     16442\n",
       "2       chr1     16342     16542        chr1:16442-16452  0  -     16442\n",
       "3       chr1    141373    141573      chr1:141473-141475  0  -    141473\n",
       "4       chr1    262620    262820      chr1:262720-262721  0  -    262720\n",
       "...      ...       ...       ...                     ... .. ..       ...\n",
       "251067  chrY  57130187  57130387  chrY:57130287-57130289  0  +  57130287\n",
       "251068  chrY  57194881  57195081  chrY:57194981-57194983  0  +  57194981\n",
       "251069  chrY  57197235  57197435  chrY:57197335-57197337  0  +  57197335\n",
       "251070  chrY  57201227  57201427  chrY:57201327-57201329  0  -  57201327\n",
       "251071  chrY  57201228  57201428  chrY:57201328-57201330  0  -  57201328\n",
       "\n",
       "[251072 rows x 7 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf['PAS']=positivedf[3].str.split(':',expand=True)[1].str.split('-',expand=True)[0]\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b4d2d5f2-ceee-4d64-bc6f-e77e3b86f57a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57130287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57194981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57197335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0  5       PAS\n",
       "0       chr1  -     16441\n",
       "1       chr1  -     16442\n",
       "2       chr1  -     16442\n",
       "3       chr1  -    141473\n",
       "4       chr1  -    262720\n",
       "...      ... ..       ...\n",
       "251067  chrY  +  57130287\n",
       "251068  chrY  +  57194981\n",
       "251069  chrY  +  57197335\n",
       "251070  chrY  -  57201327\n",
       "251071  chrY  -  57201328\n",
       "\n",
       "[251072 rows x 3 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf=positivedf[[0,5,'PAS']]\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d28bc5e0-85c5-40b7-8754-aa1391749b81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/767875871.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  positivedf['label']='positive'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57130287</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57194981</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57197335</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201327</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201328</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0  5       PAS     label\n",
       "0       chr1  -     16441  positive\n",
       "1       chr1  -     16442  positive\n",
       "2       chr1  -     16442  positive\n",
       "3       chr1  -    141473  positive\n",
       "4       chr1  -    262720  positive\n",
       "...      ... ..       ...       ...\n",
       "251067  chrY  +  57130287  positive\n",
       "251068  chrY  +  57194981  positive\n",
       "251069  chrY  +  57197335  positive\n",
       "251070  chrY  -  57201327  positive\n",
       "251071  chrY  -  57201328  positive\n",
       "\n",
       "[251072 rows x 4 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf['label']='positive'\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "fb9708d1-7e4c-42e5-ac07-9d3883367abe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251067</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57130287</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251068</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57194981</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251069</th>\n",
       "      <td>chrY</td>\n",
       "      <td>+</td>\n",
       "      <td>57197335</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251070</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201327</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251071</th>\n",
       "      <td>chrY</td>\n",
       "      <td>-</td>\n",
       "      <td>57201328</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr strand       PAS     label\n",
       "0       chr1      -     16441  positive\n",
       "1       chr1      -     16442  positive\n",
       "2       chr1      -     16442  positive\n",
       "3       chr1      -    141473  positive\n",
       "4       chr1      -    262720  positive\n",
       "...      ...    ...       ...       ...\n",
       "251067  chrY      +  57130287  positive\n",
       "251068  chrY      +  57194981  positive\n",
       "251069  chrY      +  57197335  positive\n",
       "251070  chrY      -  57201327  positive\n",
       "251071  chrY      -  57201328  positive\n",
       "\n",
       "[251072 rows x 4 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positivedf.columns=['chr','strand','PAS','label']\n",
    "positivedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "24fd0550-d969-4821-9e96-f57164074fdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>11868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>12009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29553</td>\n",
       "      <td>31097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>29553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30266</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>30266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30365</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>30365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252830</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57015104</td>\n",
       "      <td>57016096</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57016096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252831</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57165511</td>\n",
       "      <td>57165845</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57165845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252832</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57171889</td>\n",
       "      <td>57172769</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57172769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252833</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57201142</td>\n",
       "      <td>57203357</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57203357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252834</th>\n",
       "      <td>chrY</td>\n",
       "      <td>transcript</td>\n",
       "      <td>57212183</td>\n",
       "      <td>57214397</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>57214397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252835 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Feature     Start       End Score Strand       TSS\n",
       "0            chr1  transcript     11868     14409     .      +     11868\n",
       "1            chr1  transcript     12009     13670     .      +     12009\n",
       "2            chr1  transcript     29553     31097     .      +     29553\n",
       "3            chr1  transcript     30266     31109     .      +     30266\n",
       "4            chr1  transcript     30365     30503     .      +     30365\n",
       "...           ...         ...       ...       ...   ...    ...       ...\n",
       "252830       chrY  transcript  57015104  57016096     .      -  57016096\n",
       "252831       chrY  transcript  57165511  57165845     .      -  57165845\n",
       "252832       chrY  transcript  57171889  57172769     .      -  57172769\n",
       "252833       chrY  transcript  57201142  57203357     .      -  57203357\n",
       "252834       chrY  transcript  57212183  57214397     .      -  57214397\n",
       "\n",
       "[252835 rows x 7 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/TSS_site_negative.bed',delimiter='\\t')\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "307d1b84-c23c-426d-940b-556d3e33e040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223327</th>\n",
       "      <td>chr19</td>\n",
       "      <td>transcript</td>\n",
       "      <td>10308371</td>\n",
       "      <td>10315768</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>10315768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111437</th>\n",
       "      <td>chr8</td>\n",
       "      <td>transcript</td>\n",
       "      <td>81524980</td>\n",
       "      <td>81533867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>81533867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187175</th>\n",
       "      <td>chr16</td>\n",
       "      <td>transcript</td>\n",
       "      <td>53391088</td>\n",
       "      <td>53392473</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>53391088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113258</th>\n",
       "      <td>chr8</td>\n",
       "      <td>transcript</td>\n",
       "      <td>133038366</td>\n",
       "      <td>133102628</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>133102628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110978</th>\n",
       "      <td>chr8</td>\n",
       "      <td>transcript</td>\n",
       "      <td>64574827</td>\n",
       "      <td>64581811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>64581811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136891</th>\n",
       "      <td>chr11</td>\n",
       "      <td>transcript</td>\n",
       "      <td>73289406</td>\n",
       "      <td>73298625</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>73289406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116258</th>\n",
       "      <td>chr9</td>\n",
       "      <td>transcript</td>\n",
       "      <td>94166288</td>\n",
       "      <td>94200627</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>94166288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129875</th>\n",
       "      <td>chr10</td>\n",
       "      <td>transcript</td>\n",
       "      <td>76877599</td>\n",
       "      <td>77183479</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>77183479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169042</th>\n",
       "      <td>chr14</td>\n",
       "      <td>transcript</td>\n",
       "      <td>94026377</td>\n",
       "      <td>94039411</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>94026377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234038</th>\n",
       "      <td>chr20</td>\n",
       "      <td>transcript</td>\n",
       "      <td>41182113</td>\n",
       "      <td>41300087</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>41300087</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237263 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome     Feature      Start        End Score Strand        TSS\n",
       "223327      chr19  transcript   10308371   10315768     .      -   10315768\n",
       "111437       chr8  transcript   81524980   81533867     .      -   81533867\n",
       "187175      chr16  transcript   53391088   53392473     .      +   53391088\n",
       "113258       chr8  transcript  133038366  133102628     .      -  133102628\n",
       "110978       chr8  transcript   64574827   64581811     .      -   64581811\n",
       "...           ...         ...        ...        ...   ...    ...        ...\n",
       "136891      chr11  transcript   73289406   73298625     .      +   73289406\n",
       "116258       chr9  transcript   94166288   94200627     .      +   94166288\n",
       "129875      chr10  transcript   76877599   77183479     .      -   77183479\n",
       "169042      chr14  transcript   94026377   94039411     .      +   94026377\n",
       "234038      chr20  transcript   41182113   41300087     .      -   41300087\n",
       "\n",
       "[237263 rows x 7 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=TSSdf.sample(n=237263)\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1ce0966a-6b0f-414f-b977-42f80023f306",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223327</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>10315768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111437</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>81533867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187175</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>53391088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113258</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>133102628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110978</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>64581811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136891</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>73289406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116258</th>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>94166288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129875</th>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>77183479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169042</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>94026377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234038</th>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>41300087</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237263 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome Strand        TSS\n",
       "223327      chr19      -   10315768\n",
       "111437       chr8      -   81533867\n",
       "187175      chr16      +   53391088\n",
       "113258       chr8      -  133102628\n",
       "110978       chr8      -   64581811\n",
       "...           ...    ...        ...\n",
       "136891      chr11      +   73289406\n",
       "116258       chr9      +   94166288\n",
       "129875      chr10      -   77183479\n",
       "169042      chr14      +   94026377\n",
       "234038      chr20      -   41300087\n",
       "\n",
       "[237263 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf=TSSdf[['Chromosome','Strand','TSS']]\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a3f451a2-a560-4d77-a703-262eedee7677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223327</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>10315768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111437</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>81533867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187175</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>53391088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113258</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>133102628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110978</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>64581811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136891</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>73289406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116258</th>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>94166288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129875</th>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>77183479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169042</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>94026377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234038</th>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>41300087</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237263 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand        PAS\n",
       "223327  chr19      -   10315768\n",
       "111437   chr8      -   81533867\n",
       "187175  chr16      +   53391088\n",
       "113258   chr8      -  133102628\n",
       "110978   chr8      -   64581811\n",
       "...       ...    ...        ...\n",
       "136891  chr11      +   73289406\n",
       "116258   chr9      +   94166288\n",
       "129875  chr10      -   77183479\n",
       "169042  chr14      +   94026377\n",
       "234038  chr20      -   41300087\n",
       "\n",
       "[237263 rows x 3 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf.columns=['chr','strand','PAS']\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "0d3a0cf3-620a-4b75-af6a-b4e476275c3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/1978122523.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  TSSdf['label']='negative'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223327</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>10315768</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111437</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>81533867</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187175</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>53391088</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113258</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>133102628</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110978</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>64581811</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136891</th>\n",
       "      <td>chr11</td>\n",
       "      <td>+</td>\n",
       "      <td>73289406</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116258</th>\n",
       "      <td>chr9</td>\n",
       "      <td>+</td>\n",
       "      <td>94166288</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129875</th>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>77183479</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169042</th>\n",
       "      <td>chr14</td>\n",
       "      <td>+</td>\n",
       "      <td>94026377</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234038</th>\n",
       "      <td>chr20</td>\n",
       "      <td>-</td>\n",
       "      <td>41300087</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237263 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand        PAS     label\n",
       "223327  chr19      -   10315768  negative\n",
       "111437   chr8      -   81533867  negative\n",
       "187175  chr16      +   53391088  negative\n",
       "113258   chr8      -  133102628  negative\n",
       "110978   chr8      -   64581811  negative\n",
       "...       ...    ...        ...       ...\n",
       "136891  chr11      +   73289406  negative\n",
       "116258   chr9      +   94166288  negative\n",
       "129875  chr10      -   77183479  negative\n",
       "169042  chr14      +   94026377  negative\n",
       "234038  chr20      -   41300087  negative\n",
       "\n",
       "[237263 rows x 4 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TSSdf['label']='negative'\n",
    "TSSdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "6d943254-202b-453a-80a6-b85021587e94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "      <td>chr4_25865888_25866088</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "      <td>chr4_102915613_102915813</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "      <td>chr4_134185413_134185613</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "      <td>chr1_90925_91125</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "      <td>chr6_36053950_36054150</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1141329</td>\n",
       "      <td>1141529</td>\n",
       "      <td>chr1_1141329_1141529</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1007364</td>\n",
       "      <td>1007564</td>\n",
       "      <td>chr1_1007364_1007564</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33083801</td>\n",
       "      <td>33084001</td>\n",
       "      <td>chr6_33083801_33084001</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>144843917</td>\n",
       "      <td>144844117</td>\n",
       "      <td>chr4_144843917_144844117</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>138073583</td>\n",
       "      <td>138073783</td>\n",
       "      <td>chr4_138073583_138073783</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13809 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2                         3  4  5\n",
       "0      chr4   25865888   25866088    chr4_25865888_25866088  1  -\n",
       "1      chr4  102915613  102915813  chr4_102915613_102915813  1  +\n",
       "2      chr4  134185413  134185613  chr4_134185413_134185613  1  -\n",
       "3      chr1      90925      91125          chr1_90925_91125  1  -\n",
       "4      chr6   36053950   36054150    chr6_36053950_36054150  1  -\n",
       "...     ...        ...        ...                       ... .. ..\n",
       "13804  chr1    1141329    1141529      chr1_1141329_1141529  1  +\n",
       "13805  chr1    1007364    1007564      chr1_1007364_1007564  1  +\n",
       "13806  chr6   33083801   33084001    chr6_33083801_33084001  1  -\n",
       "13807  chr4  144843917  144844117  chr4_144843917_144844117  1  +\n",
       "13808  chr4  138073583  138073783  chr4_138073583_138073783  1  -\n",
       "\n",
       "[13809 rows x 6 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf=pd.read_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/filtered_negative.bed',delimiter='\\t',header=None)\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "4b5c9917-0b35-457e-843f-542fd857f3b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>25865888</td>\n",
       "      <td>25866088</td>\n",
       "      <td>chr4_25865888_25866088</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>25865988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>102915613</td>\n",
       "      <td>102915813</td>\n",
       "      <td>chr4_102915613_102915813</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>102915713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>134185413</td>\n",
       "      <td>134185613</td>\n",
       "      <td>chr4_134185413_134185613</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>134185513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>90925</td>\n",
       "      <td>91125</td>\n",
       "      <td>chr1_90925_91125</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>91025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>36053950</td>\n",
       "      <td>36054150</td>\n",
       "      <td>chr6_36053950_36054150</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>36054050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1141329</td>\n",
       "      <td>1141529</td>\n",
       "      <td>chr1_1141329_1141529</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1007364</td>\n",
       "      <td>1007564</td>\n",
       "      <td>chr1_1007364_1007564</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>33083801</td>\n",
       "      <td>33084001</td>\n",
       "      <td>chr6_33083801_33084001</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>144843917</td>\n",
       "      <td>144844117</td>\n",
       "      <td>chr4_144843917_144844117</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>138073583</td>\n",
       "      <td>138073783</td>\n",
       "      <td>chr4_138073583_138073783</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13809 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2                         3  4  5        PAS\n",
       "0      chr4   25865888   25866088    chr4_25865888_25866088  1  -   25865988\n",
       "1      chr4  102915613  102915813  chr4_102915613_102915813  1  +  102915713\n",
       "2      chr4  134185413  134185613  chr4_134185413_134185613  1  -  134185513\n",
       "3      chr1      90925      91125          chr1_90925_91125  1  -      91025\n",
       "4      chr6   36053950   36054150    chr6_36053950_36054150  1  -   36054050\n",
       "...     ...        ...        ...                       ... .. ..        ...\n",
       "13804  chr1    1141329    1141529      chr1_1141329_1141529  1  +    1141429\n",
       "13805  chr1    1007364    1007564      chr1_1007364_1007564  1  +    1007464\n",
       "13806  chr6   33083801   33084001    chr6_33083801_33084001  1  -   33083901\n",
       "13807  chr4  144843917  144844117  chr4_144843917_144844117  1  +  144844017\n",
       "13808  chr4  138073583  138073783  chr4_138073583_138073783  1  -  138073683\n",
       "\n",
       "[13809 rows x 7 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf['PAS']=intergenicdf[1]+100\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3e507abd-bf9f-4016-8545-9624ab146550",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>25865988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>102915713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>134185513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>91025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>36054050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13809 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0  5        PAS\n",
       "0      chr4  -   25865988\n",
       "1      chr4  +  102915713\n",
       "2      chr4  -  134185513\n",
       "3      chr1  -      91025\n",
       "4      chr6  -   36054050\n",
       "...     ... ..        ...\n",
       "13804  chr1  +    1141429\n",
       "13805  chr1  +    1007464\n",
       "13806  chr6  -   33083901\n",
       "13807  chr4  +  144844017\n",
       "13808  chr4  -  138073683\n",
       "\n",
       "[13809 rows x 3 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf=intergenicdf[[0,5,'PAS']]\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4921b0f7-fdfa-46f5-8073-b52434f4b7fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_154591/3419632041.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  intergenicdf['label']='negative'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>5</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>25865988</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>102915713</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>134185513</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>91025</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>36054050</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13809 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0  5        PAS     label\n",
       "0      chr4  -   25865988  negative\n",
       "1      chr4  +  102915713  negative\n",
       "2      chr4  -  134185513  negative\n",
       "3      chr1  -      91025  negative\n",
       "4      chr6  -   36054050  negative\n",
       "...     ... ..        ...       ...\n",
       "13804  chr1  +    1141429  negative\n",
       "13805  chr1  +    1007464  negative\n",
       "13806  chr6  -   33083901  negative\n",
       "13807  chr4  +  144844017  negative\n",
       "13808  chr4  -  138073683  negative\n",
       "\n",
       "[13809 rows x 4 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf['label']='negative'\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f16b0ffc-c884-4475-8b34-9ff4113e6148",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>25865988</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>102915713</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>134185513</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>91025</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>36054050</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13809 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand        PAS     label\n",
       "0      chr4      -   25865988  negative\n",
       "1      chr4      +  102915713  negative\n",
       "2      chr4      -  134185513  negative\n",
       "3      chr1      -      91025  negative\n",
       "4      chr6      -   36054050  negative\n",
       "...     ...    ...        ...       ...\n",
       "13804  chr1      +    1141429  negative\n",
       "13805  chr1      +    1007464  negative\n",
       "13806  chr6      -   33083901  negative\n",
       "13807  chr4      +  144844017  negative\n",
       "13808  chr4      -  138073683  negative\n",
       "\n",
       "[13809 rows x 4 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenicdf.columns=['chr','strand','PAS','label']\n",
    "intergenicdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "fc5e4285-db8e-412d-99cf-4ff06f84f2f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>223327</th>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>10315768</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111437</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>81533867</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187175</th>\n",
       "      <td>chr16</td>\n",
       "      <td>+</td>\n",
       "      <td>53391088</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113258</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>133102628</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110978</th>\n",
       "      <td>chr8</td>\n",
       "      <td>-</td>\n",
       "      <td>64581811</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251072 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr strand        PAS     label\n",
       "223327  chr19      -   10315768  negative\n",
       "111437   chr8      -   81533867  negative\n",
       "187175  chr16      +   53391088  negative\n",
       "113258   chr8      -  133102628  negative\n",
       "110978   chr8      -   64581811  negative\n",
       "...       ...    ...        ...       ...\n",
       "13804    chr1      +    1141429  negative\n",
       "13805    chr1      +    1007464  negative\n",
       "13806    chr6      -   33083901  negative\n",
       "13807    chr4      +  144844017  negative\n",
       "13808    chr4      -  138073683  negative\n",
       "\n",
       "[251072 rows x 4 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negativedf=pd.concat([TSSdf,intergenicdf],axis=0)\n",
    "negativedf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "7b5d1c58-6f79-47c6-abb7-81878f3d4a0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>positive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>negative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502144 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand        PAS     label\n",
       "0      chr1      -      16441  positive\n",
       "1      chr1      -      16442  positive\n",
       "2      chr1      -      16442  positive\n",
       "3      chr1      -     141473  positive\n",
       "4      chr1      -     262720  positive\n",
       "...     ...    ...        ...       ...\n",
       "13804  chr1      +    1141429  negative\n",
       "13805  chr1      +    1007464  negative\n",
       "13806  chr6      -   33083901  negative\n",
       "13807  chr4      +  144844017  negative\n",
       "13808  chr4      -  138073683  negative\n",
       "\n",
       "[502144 rows x 4 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf=pd.concat([positivedf,negativedf],axis=0)\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "04cf4f89-9f57-4def-89c2-0f9087867c97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr1_-_16441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr1_-_141473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>positive</td>\n",
       "      <td>chr1_-_262720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr1_+_1141429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr1_+_1007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr6_-_33083901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr4_+_144844017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>negative</td>\n",
       "      <td>chr4_-_138073683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502144 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand        PAS     label              name\n",
       "0      chr1      -      16441  positive      chr1_-_16441\n",
       "1      chr1      -      16442  positive      chr1_-_16442\n",
       "2      chr1      -      16442  positive      chr1_-_16442\n",
       "3      chr1      -     141473  positive     chr1_-_141473\n",
       "4      chr1      -     262720  positive     chr1_-_262720\n",
       "...     ...    ...        ...       ...               ...\n",
       "13804  chr1      +    1141429  negative    chr1_+_1141429\n",
       "13805  chr1      +    1007464  negative    chr1_+_1007464\n",
       "13806  chr6      -   33083901  negative   chr6_-_33083901\n",
       "13807  chr4      +  144844017  negative  chr4_+_144844017\n",
       "13808  chr4      -  138073683  negative  chr4_-_138073683\n",
       "\n",
       "[502144 rows x 5 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf['name']=alldf['chr']+'_'+alldf['strand']+'_'+alldf['PAS'].astype('str')\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b040595e-c8d2-46ac-b754-31ee64dd7a8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_141473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_262720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_1141429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_1007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13806</th>\n",
       "      <td>chr6</td>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>0</td>\n",
       "      <td>chr6_-_33083901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13807</th>\n",
       "      <td>chr4</td>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>0</td>\n",
       "      <td>chr4_+_144844017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13808</th>\n",
       "      <td>chr4</td>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>0</td>\n",
       "      <td>chr4_-_138073683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502144 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr strand        PAS  label              name\n",
       "0      chr1      -      16441      1      chr1_-_16441\n",
       "1      chr1      -      16442      1      chr1_-_16442\n",
       "2      chr1      -      16442      1      chr1_-_16442\n",
       "3      chr1      -     141473      1     chr1_-_141473\n",
       "4      chr1      -     262720      1     chr1_-_262720\n",
       "...     ...    ...        ...    ...               ...\n",
       "13804  chr1      +    1141429      0    chr1_+_1141429\n",
       "13805  chr1      +    1007464      0    chr1_+_1007464\n",
       "13806  chr6      -   33083901      0   chr6_-_33083901\n",
       "13807  chr4      +  144844017      0  chr4_+_144844017\n",
       "13808  chr4      -  138073683      0  chr4_-_138073683\n",
       "\n",
       "[502144 rows x 5 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alldf['label'].replace({'positive':1,'negative':0},inplace=True)\n",
    "alldf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "55c556ec-53de-4683-89b8-c4fd5bc8e53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#alldf.to_csv('/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/final_all_sample_negative_positive_balance.tsv',sep='\\t',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a3c211-e91e-4676-9b5f-d0a0332b3f51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c01fd3e1-3219-43dd-827f-4b893d8cc4ab",
   "metadata": {},
   "source": [
    "# train human model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b4bb7a22-74db-49aa-8377-fae16044342b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from kipoiseq.transforms.functional import one_hot\n",
    "from pyfaidx import Fasta\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch\n",
    "from sklearn.model_selection import KFold\n",
    "from torch.utils.data import DataLoader\n",
    "from torch.utils.data import Dataset\n",
    "from tensorboardX import SummaryWriter\n",
    "from scipy import stats\n",
    "import torch.optim as optim\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import torch.nn as nn\n",
    "import pickle\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import h5py\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "49974fd9-82ab-4c46-a178-60f387258099",
   "metadata": {},
   "outputs": [],
   "source": [
    "def setup_seed(seed):\n",
    "     torch.manual_seed(seed)\n",
    "     torch.cuda.manual_seed_all(seed)\n",
    "     np.random.seed(seed)\n",
    "     random.seed(seed)\n",
    "     torch.backends.cudnn.deterministic = True\n",
    "# 设置随机数种子\n",
    "\n",
    "setup_seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "fcca1f6a-6d5c-468d-b358-5476e440cdcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "class scDataset(Dataset):\n",
    "    def __init__(self,ref_fastq,bed,idx,train_idx,col_name):\n",
    "        \n",
    "        #read bed file\n",
    "        genes=Fasta(ref_fastq)\n",
    "        bedfile=pd.read_csv(bed,delimiter='\\t')\n",
    "        bedfile['start']=bedfile['PAS']-100\n",
    "        bedfile['end']=bedfile['PAS']+100\n",
    "        \n",
    "        #extract sequence and do one-hot encoding \n",
    "        onehotdict={}\n",
    "        for chrom, start, end, strand, genename in tqdm(bedfile[['chr','start','end','strand','name']].values,desc='Loading Loci'):\n",
    "            if strand=='+':\n",
    "                seq_=genes.get_seq(chrom,start+1,end).seq\n",
    "            else:\n",
    "                seq_=genes.get_seq(chrom,start+1,end,rc=True).seq\n",
    "\n",
    "            onehotnp=one_hot(seq_,['A','C','T','G'])\n",
    "            onehotdict[genename]=onehotnp\n",
    "            \n",
    "        \n",
    "        #get y value \n",
    "        self.y_value=np.array(bedfile[col_name][idx])\n",
    "        #print(self.y_value)\n",
    "        \n",
    "        \n",
    "        # get corresponding X value \n",
    "        self.gn=bedfile['name'][idx].tolist()\n",
    "        #print(self.gn)\n",
    "        \n",
    "        onehotls=[]\n",
    "        for i in self.gn:\n",
    "            onehotls.append(onehotdict[i])\n",
    "        #print(onehotls)\n",
    "            \n",
    "        self.combine_array=np.stack(onehotls)\n",
    "        #print(self.combine_array)\n",
    "                \n",
    "    def __len__(self):\n",
    "        return len(self.y_value)\n",
    "    \n",
    "    \n",
    "    def __getitem__(self,index):\n",
    "        #print(self.combine_array.shape)\n",
    "        #print(index)\n",
    "        seq=np.transpose(self.combine_array[index]).astype(np.float32)\n",
    "        y_value=self.y_value[index].astype(np.float32)\n",
    "        TSS_name=self.gn[index]\n",
    "        \n",
    "        #print(TSS_name)\n",
    "\n",
    "        sample = {\"seq\": seq, \"y_true_value\": y_value,'TSS_name':TSS_name}\n",
    "        return sample \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "01eea68e-12d5-431a-8c94-3cb70d0f6a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "    \n",
    "\n",
    "# Validation\n",
    "\n",
    "def evaluate(device,net,dataloader,criterion):\n",
    "    net.eval()\n",
    "    avg_loss=0\n",
    "    TSS_name_ls=[]\n",
    "    gt_labels=[]\n",
    "    pred_labels=[]\n",
    "    y_scorels=[]\n",
    "    with torch.no_grad():\n",
    "        for data in dataloader:\n",
    "            seq=data['seq'].to(device,dtype=torch.float)\n",
    "\n",
    "            label=data['y_true_value'].to(device,dtype=torch.int64)\n",
    "            output=net(seq)\n",
    "            current_loss = criterion(output, label)\n",
    "            avg_loss+=current_loss.item() / len(dataloader) \n",
    "\n",
    "\n",
    "            gt_labels.append(label.cpu().data.numpy())\n",
    "            TSS_name=data['TSS_name']\n",
    "            TSS_name_ls.append(TSS_name[0])\n",
    "\n",
    "            preds = torch.argmax(output, 1)\n",
    "            pred_labels.append(preds.cpu().data.numpy())\n",
    "            \n",
    "            y_scorels.append(output[:,1].cpu().data.numpy())\n",
    "\n",
    "\n",
    "    return avg_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3390681d-2be6-449a-9f0b-44e59a30aff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Train\n",
    "\n",
    "def train(device,train_loader,validation_loader,net,criterion,learning_rate,lr_sched,num_epochs,logs_dir,model_dir):\n",
    "    logger = SummaryWriter(logs_dir)\n",
    "    #optimizer=optim.Adam(net.parameters(),lr=learning_rate)\n",
    "    optimizer = optim.SGD(net.parameters(), lr=0.003,momentum=0.8)\n",
    "    scheduler=optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', \n",
    "                                                     factor=0.1, patience=5)\n",
    "    best_loss=100\n",
    "    \n",
    "    for epoch in range(0,num_epochs):\n",
    "        net.train()\n",
    "        avg_train_loss=0\n",
    "        for data in train_loader:\n",
    "            seq=data['seq'].to(device)\n",
    "            # print(\"Training seq:\")\n",
    "            # print(seq)\n",
    "            #hl=data['hl'].to(device,dtype=torch.float)\n",
    "            y_true_value=data['y_true_value'].to(device,torch.int64)\n",
    "            #print(y_true_value)\n",
    "            \n",
    "\n",
    "        \n",
    "            \n",
    "            with torch.set_grad_enabled(True):\n",
    "                optimizer.zero_grad()\n",
    "                \n",
    "                #Forward pass\n",
    "                y_pred_value=net(seq)\n",
    "                #print(y_pred_value)\n",
    "                \n",
    "                current_loss = criterion(y_pred_value,y_true_value)\n",
    "                current_loss.backward()\n",
    "                optimizer.step()\n",
    "                avg_train_loss+=current_loss.item()/len(train_loader)\n",
    "                logger.add_scalar('train_loss_epoch', avg_train_loss, epoch + 1)\n",
    "                \n",
    "                \n",
    "        print(\"The average train loss is %f\" %avg_train_loss)\n",
    "        \n",
    "        \n",
    "        #get validation loss \n",
    "        validation_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls=evaluate(device,net,validation_loader,criterion)\n",
    "        \n",
    "        # print(gt_labels)\n",
    "        # print(y_scorels)\n",
    "        \n",
    "        auc=roc_auc_score(gt_labels, y_scorels)\n",
    "        print('Validation auc: %f'%auc)\n",
    "        \n",
    "        print(\"The average validation loss is %f\" %validation_loss)\n",
    "\n",
    "        logger.add_scalar('valid_loss_epoch', validation_loss, epoch + 1)\n",
    "        \n",
    "        \n",
    "        # store best model\n",
    "        if validation_loss<best_loss:\n",
    "            best_loss=validation_loss\n",
    "            state={'epoch':epoch+1,'state_dict':net.state_dict(),'optimizer':optimizer.state_dict(),'loss':criterion,'scheduler':scheduler.state_dict()}\n",
    "            torch.save(state,model_dir+'/model_last.pth.tar')\n",
    "            \n",
    "            \n",
    "        # LR scheduler on plateau (based on validation loss)\n",
    "        if lr_sched:\n",
    "            scheduler.step(validation_loss)\n",
    "            \n",
    "        print(\"Finish training.\")\n",
    "            \n",
    "\n",
    "          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "aa88959f-f3a1-4b19-955b-199dfa9a64bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "  \n",
    "\n",
    "# Test\n",
    "\n",
    "def test(testdataloader,device,net,model_dir,save_file,fold_no):\n",
    "    \n",
    "    #get model\n",
    "    check_point=torch.load(model_dir+'/model_last.pth.tar')\n",
    "    net.load_state_dict(check_point['state_dict'])\n",
    "    criterion = check_point['loss']\n",
    "    \n",
    "    avg_loss,gt_labels,pred_labels,y_scorels,TSS_name_ls=evaluate(device,net,testdataloader,criterion)\n",
    "    #y_true_value=np.array([i[0]for i in y_true_value])\n",
    "    auc_value=roc_auc_score(gt_labels, y_scorels)\n",
    "    # print(\"True y value: \\n\")\n",
    "    # print(y_true_value)\n",
    "    \n",
    "    # print(\"Pred y value: \\n\")\n",
    "    # print(y_pred_value)\n",
    "    \n",
    "    # Save predictions\n",
    "    if save_file is not None:\n",
    "        pickle.dump({'y_true': gt_labels, 'y_pred': pred_labels,'y_score':y_scorels,'TSS':TSS_name_ls}, open(save_file, 'wb'))\n",
    "\n",
    "    print(\"The test auc is %f\" %auc_value)\n",
    "    # print(\"Test loss is %f\"%test_loss)\n",
    "    # print(corr_value)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "07356ddb-a9ec-4ef9-b72e-de60424b6870",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model\n",
    "\n",
    "class Net(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Net, self).__init__()\n",
    "        self.conv = nn.Sequential(\n",
    "            nn.Conv1d(4, 128, 8),\n",
    "             nn.ReLU(),\n",
    "            nn.Conv1d(128,64,4),\n",
    "            nn.ReLU(),\n",
    "            nn.Conv1d(64,32,2),\n",
    "             nn.BatchNorm1d(32),\n",
    "              nn.MaxPool2d(2),\n",
    "            nn.Dropout(0.4),\n",
    "             nn.Flatten(),\n",
    "            nn.Linear(1504, 32),\n",
    "            nn.ReLU(),\n",
    "        )\n",
    "        self.fc = nn.Sequential(\n",
    "            nn.Linear(32, 2),\n",
    "            nn.Sigmoid()\n",
    "        )\n",
    "        \n",
    "    def forward(self, x):\n",
    "        x = self.conv(x)\n",
    "        #print(x.shape)\n",
    "        #combinex=torch.cat((x,extrafeature),dim=1)\n",
    "        #x = x.view(-1, 3040)\n",
    "        x = self.fc(x)\n",
    "        # x = nn.functional.normalize(x)\n",
    "        return x\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "442075c3-c770-4c95-859b-cdefdd4ba57a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Main\n",
    "\n",
    "#device=torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')\n",
    "device=1\n",
    "\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "9343df0b-7e9e-473e-9fa6-4f2ebf1ddb2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KFold(n_splits=10, random_state=666, shuffle=True)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kf=KFold(n_splits=10,random_state=666,shuffle=True)\n",
    "kf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5b71c405-89bd-4ee4-aafb-b63836066ae1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_fastq='/mnt/ruiyanhou/nfs_share2/annotation/annotation_from_Gencode/GRCh38.primary_assembly.genome.fa'\n",
    "bed='/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/final_all_sample_negative_positive_balance.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "c8ee595b-25be-438c-b9db-ac038693e5ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>PAS</th>\n",
       "      <th>label</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>16441</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>16442</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_16442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>141473</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_141473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>-</td>\n",
       "      <td>262720</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_-_262720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>1141429</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_1141429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>+</td>\n",
       "      <td>1007464</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1_+_1007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6</th>\n",
       "      <td>-</td>\n",
       "      <td>33083901</td>\n",
       "      <td>0</td>\n",
       "      <td>chr6_-_33083901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4</th>\n",
       "      <td>+</td>\n",
       "      <td>144844017</td>\n",
       "      <td>0</td>\n",
       "      <td>chr4_+_144844017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4</th>\n",
       "      <td>-</td>\n",
       "      <td>138073683</td>\n",
       "      <td>0</td>\n",
       "      <td>chr4_-_138073683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>502144 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     strand        PAS  label              name\n",
       "chr                                            \n",
       "chr1      -      16441      1      chr1_-_16441\n",
       "chr1      -      16442      1      chr1_-_16442\n",
       "chr1      -      16442      1      chr1_-_16442\n",
       "chr1      -     141473      1     chr1_-_141473\n",
       "chr1      -     262720      1     chr1_-_262720\n",
       "...     ...        ...    ...               ...\n",
       "chr1      +    1141429      0    chr1_+_1141429\n",
       "chr1      +    1007464      0    chr1_+_1007464\n",
       "chr6      -   33083901      0   chr6_-_33083901\n",
       "chr4      +  144844017      0  chr4_+_144844017\n",
       "chr4      -  138073683      0  chr4_-_138073683\n",
       "\n",
       "[502144 rows x 4 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beddf=pd.read_csv(bed,delimiter='\\t',index_col=0)\n",
    "beddf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51681acb-a95f-46fe-8e49-648620bb40e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fold_no, (train_index, test_index) in enumerate(kf.split(beddf)):\n",
    "    print(\"this is fold number: %i\" %fold_no)\n",
    "    \n",
    "    print(len(train_index))\n",
    "    print(len(test_index))\n",
    "    \n",
    "\n",
    "    np.random.seed(666)\n",
    "    \n",
    "    idx_rest = np.random.permutation(len(beddf)-len(test_index))\n",
    "    \n",
    "    idx_val = train_index[idx_rest[:int(0.2*len(idx_rest))]]\n",
    "    idx_train = train_index[idx_rest[int(0.2*len(idx_rest)):]]\n",
    "    \n",
    "    print(\"train_id\")\n",
    "    print(idx_train[:10])\n",
    "    \n",
    "    print(\"validation_id\")\n",
    "    print(idx_val[:10])\n",
    "    \n",
    "    print(\"test_id\")\n",
    "    print(test_index[:10])\n",
    "    \n",
    "\n",
    "    train_set=scDataset(ref_fastq,bed,idx_train,idx_train,'label')\n",
    "    validation_set=scDataset(ref_fastq,bed,idx_val,idx_train,'label')\n",
    "    test_set=scDataset(ref_fastq,bed,test_index,idx_train,'label')\n",
    "    \n",
    "    train_loader = DataLoader(train_set, batch_size=128, shuffle=True)\n",
    "    val_loader = DataLoader(validation_set, batch_size=1, shuffle=False)\n",
    "    test_loader = DataLoader(test_set, batch_size=1, shuffle=False)\n",
    "    \n",
    "\n",
    "    model_dir='/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/only_annotation/model_save_final/model_dir'\n",
    "    log_dir='/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/only_annotation/model_save_final/log_dir'\n",
    "    save_dir='/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/only_annotation/model_save_final/TSS_true_pred_y_'+str(fold_no)+'.pkl'\n",
    "    \n",
    "    \n",
    "    criterion  = nn.CrossEntropyLoss().to(device)\n",
    "    net = Net().to(device)\n",
    "    init_lr =0.0005\n",
    "    lr_sched = True\n",
    "    num_epoch=100\n",
    "    \n",
    "    train(device,train_loader,val_loader,net,criterion,init_lr,lr_sched,num_epoch,log_dir,model_dir)\n",
    "    \n",
    "    test(test_loader,device,net,model_dir,save_dir,fold_no)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f8dd982-6d39-41e4-b467-13a7692b5552",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1aa4ce0e-2253-42ad-adce-d8bb59d9e22b",
   "metadata": {},
   "source": [
    "# Plot picture"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "72583723-dfc1-416d-b130-e059b335f3a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from hilearn.plot import ROC_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "15b5d1ed-517e-4567-a2a7-8366c0bc7c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_truels=[]\n",
    "y_scorels=[]\n",
    "for i in range(0,10):\n",
    "    inputfile='/mnt/ruiyanhou/nfs_share2/three_primer/eMSC/remove_false_positive/only_annotation/model_save_final/TSS_true_pred_y_'+str(i)+'.pkl'\n",
    "    with open(inputfile,'rb') as f:\n",
    "        mydict=pickle.load(f)\n",
    "\n",
    "    y_truenp=np.array([i[0] for i in mydict['y_true']])\n",
    "    y_scorenp=np.array([i[0] for i in mydict['y_score']])\n",
    "\n",
    "    y_truels.append(y_truenp)\n",
    "    y_scorels.append(y_scorenp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "56bd0d32-6489-46e9-9e52-a96e74467171",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(502144,)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_truels\n",
    "\n",
    "y_truels[0].shape\n",
    "\n",
    "total_true_y_value=np.concatenate(y_truels)\n",
    "total_true_y_value.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1823cd8b-5cbd-4adb-9471-bb3266928e3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([8.0851104e-09, 5.3111058e-02, 3.5155020e-04, ..., 7.3227250e-07,\n",
       "       2.3056762e-07, 1.2106824e-09], dtype=float32)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_y_score=np.concatenate(y_scorels)\n",
    "total_y_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "48f3f8ee-0080-4320-82b8-b585c4ad0e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_y_truels=[]\n",
    "mouse_y_scorels=[]\n",
    "for i in range(0,10):\n",
    "    mouse_inputfile='/mnt/ruiyanhou/nfs_share2/three_primer/mouse_model_redo/model_save/PAS_true_pred_y_'+str(i)+'.pkl'\n",
    "    with open(mouse_inputfile,'rb') as f:\n",
    "        mouse_mydict=pickle.load(f)\n",
    "\n",
    "    mouse_y_truenp=np.array([i[0] for i in mouse_mydict['y_true']])\n",
    "    mouse_y_scorenp=np.array([i[0] for i in mouse_mydict['y_score']])\n",
    "\n",
    "    mouse_y_truels.append(mouse_y_truenp)\n",
    "    mouse_y_scorels.append(mouse_y_scorenp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "d8bd0c69-4c87-424f-a09b-1d8f2f1c880b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0]),\n",
       " array([1, 1, 1, ..., 0, 0, 0])]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_truels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "48e8d593-d1da-4c0f-baec-a4b739af70e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(50215,)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_truels[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "ea4b066c-56e3-440d-9e1a-45cb421fa863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(213787,)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_total_true_y_value=np.concatenate(mouse_y_truels)\n",
    "mouse_total_true_y_value.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "e1b05463-d43f-4228-8a23-45b1d2cf850b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.0000000e+00, 6.4599764e-04, 9.9997413e-01, ..., 3.2314460e-04,\n",
       "       5.8540609e-03, 6.9042753e-14], dtype=float32)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_total_y_score=np.concatenate(mouse_y_scorels)\n",
    "mouse_total_y_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "b054426d-606e-4a1a-ae1a-10c186b3ef55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhgAAAHACAYAAADgARHPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAChRElEQVR4nOzdd1iTVxsH4F8SSNhLZIgo7r0B96aCOMBd96hb66p11dXWUa2zrV/d4qwD96gLt6Ag1r0FFyKKCGGTcb4/QiKRgIkmvASe+7pyJZx3PYnIeXLeM3iMMQZCCCGEED3icx0AIYQQQooeSjAIIYQQoneUYBBCCCFE7yjBIIQQQojeUYJBCCGEEL2jBIMQQgghekcJBiGEEEL0jhIMQgghhOidCdcBFDS5XI7Xr1/D2toaPB6P63AIIYQQo8EYQ3JyMkqVKgU+P/82imKXYLx+/Rru7u5ch0EIIYQYrZcvX6J06dL57lPsEgxra2sAig/HxsaG42gIIYQQ4yEWi+Hu7q6qS/NT7BIM5W0RGxsbSjAIIYSQL6BNFwPq5EkIIYQQvaMEgxBCCCF6RwkGIYQQQvSOEgxCCCGE6B0lGIQQQgjRO0owCCGEEKJ3lGAQQgghRO8owSCEEEKI3lGCQQghhBC9owSDEEIIIXrHaYJx4cIFdOrUCaVKlQKPx8OBAwc+e8y5c+dQv359iEQiVKxYEUFBQQaPkxBCCCG64TTBSE1NRZ06dbBq1Sqt9o+OjkaHDh3QunVr3LhxAxMmTMDQoUNx4sQJA0dKCCGEEF1wuthZ+/bt0b59e633X716NcqVK4elS5cCAKpVq4ZLly5h+fLl8PX1NVSYhBBCCNGRUa2mGhYWBh8fH7UyX19fTJgwIc9jMjMzkZmZqfpZLBYbKjxCSCHDGCBjgDz7IZPneJ3zWa5e9tnyT87HGMCQ/Vr5zBjkcgY5Y4rtcoDJ5ZAxAAyQyxXlcsYU++fcV3U8VGVyeXYZso9nDHKpHHKZHEwqUz0zqQxyqRxMLsu+ruJ45IgLyvMz5eekfl3l+/n4M1OcJ7sM4CH7baj2BXiq2BjvYzkDT1GGHA/GU+2DT8uRsyzHdZBjG1Pfhk+2Icc5cj6rXitjzn7TLPuNKJ/zPpb38WCe4g0wHk+1E1PbT/FK+Vp1bHa8ms/9SZnykJzXyHEa1fWUK5tmf/bKa2SK30Bk44K/AwSoUqcUCppRJRhv3ryBs7OzWpmzszPEYjHS09Nhbm6e65iFCxfi559/LqgQCdE7mRyQyIEsGSCR5Xgtz/5ZBmR98lqa/dCqsvyScmiorPMqz6dyV5VDcbzGijzfcvbxHPLs14yn2I6cf+i5wANyXV/ARSD69elb4vIjJnlKuxGM9zsGokS/bchIb8BJDEaVYHyJ6dOnY9KkSaqfxWIx3N3dOYyIFAZypl5ha6qk1bZp2E+aX0Wv4Titk4RPjlV+yySa8PJ4rT2+XAY+k4MvlyuemRw8JodA/vE1Xy6HQPk6e5vytWIfBj6TAwyKn8FylDHws3/mMZa9TQ6ect9PynjZx/OZsuzjcfzs76yKayi+2PJ5PPD4ygf/47OAp9pHmeoov+iqlWt4Rq5y3sfz4WN7gdq5s8s/XoOpbQMvxz4a41KW81T7gAfF58jjfdwOxXbwso9j2efiKX8DeODxPrYU5Hw/qt+S7AN4Ar7iwecD2c88AV+xLcev1MdzfPwdU75WO292geLz0XyOnAfwVUUfy3if7JPz3DmvyYP69ZU7yRnDjl3/w5Z//kLrFv6Y3KE8ylV2BBeMKsFwcXFBXFycWllcXBxsbGw0tl4AgEgkgkgkKojwiB5I5UBqFpCcBaTkeE6TABlSxXO6FEiXKB5pytfZzzm351eJy4y40jbhA6Z8BlPlMxhMeXKYQgZTJocJk8JELoNALgNfJgVfJgNPlv1aKgVPqnjmSyVqryGRgC+RQiCXfaw85TLwsyvKTyvhXNvlMvAYg0DttUxViQrkml5/UnnnU5Grv5apKnBVDAIeBCZ8CIQmEJgKFA+RECYiEwiEJuCbiWBqZgqBuami3EIEgYUZ+GYiQGQKvkgInpkQPJEQPGGO1yIheCJT8EQiRZmJCcA3+VhJqWr67OpCVdtlb+Pz1WtAHk9RUeR6QFHRqdXsvI/7E6KFiIgIbN/9N+bNm4cZM2Zw+rtjVAlG48aNcezYMbWyU6dOoXHjxhxFRDSRyIDEDOBDhuJZ+fpDBpCYDiRmAh/SAXGmInnImUikS7mJWcADTAWAkK94NhUApnxAmP2sLFNtz35W267c/9NjeXKYyKUwkWTCRJIFU0kWBFmZMMnMhElmBkwy02GSkQ5BRjpM0tNhkpYGQVoqBKmpMElNhSAlBfyUZJikJIOXnAqkpoGlpqvfbDYEPh88cxF45iLwzUTgKR/mIvBEQvDNhB9/NlPuIwTPzEK1L988ex9Rjn2FpoqKlK+olHnKijlnmeKreXblDPB4/I/bs7+xK37mg29hpjiviVH9OSNEr969ewdHR0d4eXnhwYMHqFixItchcZtgpKSk4MmTJ6qfo6OjcePGDTg4OKBMmTKYPn06YmJisGXLFgDAyJEj8ddff2HKlCkYMmQIzpw5g927d+Po0aNcvYViQ86A92nA2zTgbSoQl/LxtfLxLg1IylAkC19LJACshIqHpRCwNAUsTAEzE8WzuQlgbvrxda5tJoDI5POJgqI1ABDkM2CbZUkgF6dAlpQCuTgF8qTsx4dkRVlSCuTiVMjFyaptshz7sbT0r/9AlLFoKONZmINvaQ6epeI592uLXOWqZwsz8M0VFTTfwlxRUSu/1Zua0DdnQozA1atXERgYiKlTp2LChAmFIrkAOE4wrl27htatW6t+VvaVGDhwIIKCghAbG4sXL16otpcrVw5Hjx7FxIkTsXLlSpQuXRrr16+nIap6kpQJPPsARCd+fDxPBN6kAvFpitsX2uIBsDUD7ESAvTlgZwbYmymela9tzRQJhLXwYzJhnZ1QCA3QF06ekQn5BzFkCUmQJSRBniBGZkIi0hOyyz4kQf5BrEgQEpMhT0qGPDkVLD3z8yfXBp+vsZLPKwn4+NpC8WylIVGwMPvYAkAIKXa2bduGoUOHwtPTE3369OE6HDU8xgzdzlq4iMVi2NraIikpCTY2NlyHU+BkcuBxAvD0AxCdI5l4lgi8/8wXbR4ARwugpCXgZAE4WwFOOV47WigSB3szwEaUf6uAPrAsCaRvEyCLi4cs7r3i9bsPkGcnEKpE4oMYsvdJX92SwLe2BN/WSvGwyX7YWkGQ/awotwbfRrGfwNYafBsr8KwswLeyUNw+oBYBQogeyOVyzJgxA4sWLcKgQYOwevXqAulvqEsdSjcti7hMKXArDgh/DVyNAa7H5n8Lw8kSKGcHeNgB5bOfS1kryh0tFLcUDE2ekgZp3HvIsh/SuPeQvVU+J6jK5AlJup9cIIDAwQZ8B1sI7G3VXvMdbBTPtlYQ2NtkJxGW4Ntag29tAZ6gCAwxJIQUCYwxPHjwAEuWLMGkSZMK5ZcXSjCKGIlMkUiEvQIiXgM33gCZMvV9rIRAJQegvL16IuFhp9hmCIwxyBOSciUJOZ+Vr3VqaTA1gYmTAwTOJSBwLgGTkg6KhKGEMmlQJBECBzvwHRRJQ2H8j0gIIdqIjo7G69ev0bRpU+zbtw/8QnyLlBKMIoAx4PobYP8D4OhjIOGT+tnRHPByA7xLKZ6rOxrm9oXsgxiS6FeQRL2CJDoGkuhXkD6PhTQmDtK3CYBE+yEiPEtzmGQnDQKnEqrXqrLs13x7G0oYCCHFwoULF9CtWzeUK1cOV69eLdTJBUAJhlGTyYHjT4G/rwG3334sdzQHWnoAXqUAbzdFC4W+6mDGGKSv4pB19wmyHj6D5MkLZD19CcmTF5B/+Pw07HwH29yJg5ND7sTBykI/ARNCSBGwYcMGjBo1Ck2bNkVwcLBRfLGiBKOQypACxx4DJ54q5pGwMwN8KwD+lRRTAey7D6yJBKISFfubmQD+FYHAqkBTd/30lWByOSRPXyIj4g4yI+8hMzupYClpeR4jcHGEaTk3mJYrrXj2cIOJuzMEzo4wcXIAT2j69YERQkgxsnDhQsyYMQMjR47EH3/8AVNT4/g7SglGIXQqCvjhpGLYKJ+nmIOCz1O0Vkw/o0gmEjMU+9qKgEF1gEF1AQfNk5lqTRb/ARnX7yMz8h4yrt9D5n/3IU9Kyb2jiQDCKh4QVisP04plIKxQBqYVy8C0nBv4ll8ZBCGEEDUdO3aEnZ0dRo0axXUoOqEEo5A5FQUMO/zxZ+U6FMrnDKniYSsCxnoDfWp+ecdMSXQM0i9dR3rof8i4dhfSZ69z7cMzE0JUuwrMvGtCWLsyRNUqwLR8aWqJIIQQA3r8+DF+/vlnrFu3DrVq1UKtWrW4DklnlGAUIhlSRcsFoHnGxpwYgAG1Fa0Z2mISKTLCbyP1VCjSToRC8uRFrn1MK5WFWf1qEDWoDrP61SGsXgE8U/o1IYSQgnL69Gn07NkTTk5OiI+PN9oFOqnmKESOPVbcFtGGOBM49gToWjX//WTvE5F6Kgxpp8KQfjYc8uTUjxtNTWDWoAbMm9WDWcPaENWrCoGt9Ze/AUIIIV+MMYb//e9/GD9+PL755hv8888/sLOz4zqsL0YJRiFy4unHPhefw+cBJ/JIMBhjyLj8H5I27kfq8Utqw0P5JWxh0aYhLH2bwaKNN/jWlnp8B4QQQr7UlStXMHbsWEyaNAmLFy+GwMgn96MEoxBJzNAuuQAU+yk7eqrKMjKRsu80ktbuQdbdp6pyYY2KsPRtCot2jSGqV43WriCEkEIkNTUVlpaWaNy4McLDw+Hl5cV1SHpBCUYhYmcG8AFos6YYn6fYHwCYVIrkPSfxYdEGSGMUE2LwLMxg3as9bAZ2hqhG4VhZjxBCiLq7d++ic+fOmD59OoYOHVpkkguAEoxCpV15xVBUbcgZ4FsRSA+9gXdTl0HyIBoAIHAtCdvh3WHTrxMEdtSfghBCCqujR4+id+/e8PDwgI+PD9fh6B0lGIWIRMvl0HlMDsvMVNT9aS5eXwgHoJgh0+77PrD9rhv45oZfUY8QQsiXYYxh6dKlmDJlCjp16oRt27bB2rrofSGkBKOQiP4A/Hzh4888MDDkngqWxxRZyLSD88AeK5ILm4EBcPhpOAT2xW/5eUIIMTZSqRSHDh3CtGnTMG/evEK/psiXogSjEMiSAd8fB9IkQKPSwKDqEkw+lIEUM2vw5DIwvkD1bJmRgmmH5qPJ41AAAM/KAiXmfQ++GbVaEEJIYRYXF4e3b9+iVq1aOH36NIRCAy1fXUhQglEIbPhPsViZnRmw0hewPBaC4OW/43y1VrhUpTnE5jawSRej2cOLaHn/HISyLNWxLCUNqYfPwbqHL3dvgBBCSL5u3LiBzp07w9nZGeHh4UU+uQAoweDcmxTgD8WdDsxuAbhYAW+OXYKQSfHNnZP45s7J/E/A5yP16EVKMAghpJDat28f+vfvj2rVqmH//v1GsRKqPhTNGz9GZPkVxa2R+q5Al+xJs2QfkgC5lj0+5XLItFgmnRBCSMFbuXIlunXrho4dO+LChQsoXbo01yEVGEowOBSXAuy9r3g9o5libgsAENjbAtp2+uHzqXMnIYQUUp6envj111+xc+dOWFhYcB1OgaIEg0ObbiiGpnqVUjyULP2b6dSCYdmhuUHiI4QQoruYmBhMnDgRUqkUTZs2xcyZM4vNbZGcKMHgSLoE2H5H8XpEA/VtogY1tDsJjwe+rTUsO7XSa2yEEEK+jHKq7+DgYLx69YrrcDhFCQZHDj1SrIjqbgO0LfexnMlkeDdp8ceCvLLe7HKnVT/REFVCCCkE/vnnH7Rs2RJly5ZFREQEPDw8uA6JU5RgcGT/A8Vzn5of+14AQOKqncgIvQGehTkcf/8BfBsrxQZln4zsZ76NFVy2LoSlb9MCjJoQQogmly9fRp8+fdCzZ0+cPXsWLi4uXIfEORqmyoG3qcDVGMXrzlU+lmfefIiEhesAAI4Lx8OmTwdYf9seqYfPIfXoRcg+iCGwt4Flh+aw7NSKWi4IIYRjUqkUJiYmaNKkCY4ePYr27dsXy/4WmlCCwYHjTxSLldV1BkpnDwBhWRLEjZ0PSGWw7NgS1r39AQB8MxGse/jSPBeEEFLIPH/+HAEBAZg5cya6d+8Of39/rkMqVOgWCQeOPlY8d6j8sSxx9W5IHkSD72iHkkt/pAyYEEIKscuXL8PLywtisRhVq1blOpxCiRKMApbz9oh/RcWz5EUsPizZBABwnDsGAgdbjqIjhBDyOZs2bULr1q1RrVo1hIeHo2bNmlyHVChRglHAzj0DGIA62bdHGGOIn74CLD0TZk3qwqon3QohhJDCKisrCytWrMDAgQNx6tQpODo6ch1SoUV9MArYxReK5xZlFc/pZ8KRdjIUMDVByd9/oFsjhBBSCCUlJSEhIQHlypXDxYsXYW1tTX+vP4NaMAoQY0DoS8XrZu4Ak0oRP+tPAIDtd10hrOzBXXCEEEI0evLkCRo1aoRvv/0WjDHY2NhQcqEFSjAK0KP3QHw6YGYC1HMB0k6GQvL4OfgOtrCfPIjr8AghhHzizJkz8Pb2hlwux5YtWyix0AElGAUoNHvWWE9XQGQCJG0+BACw6dsRAltrDiMjhBDyqQ0bNqBdu3bw8vLClStXUKVKlc8fRFQowShAytsjTd0ByfPXSD8bDgCw6d+Jw6gIIYRo4urqirFjx+Lo0aOwt7fnOhyjQwlGAWHs4/DUpu6AeOthgDGYt/KCaTk3boMjhBACAEhISMDChQvBGIO/vz9WrFgBExMaD/ElKMEoIM8SgaRMQCQAqtlLkbzjGADAZkBnbgMjhBACALh//z68vb2xdOlSPHv2jOtwjB4lGAXk1lvFczVHQHbtNmTvEsC3t4GlXzNuAyOEEIJ///0XjRo1gpmZGSIiIlCuXLnPH0TyRQlGAbkVp3iu7QykHr8EALBs1wQ8U2p6I4QQLl26dAkdO3ZEixYtEBoaSsmFnlCCUUDuZLdg1HJiSP33IgDA0r85hxERQkjxxhgDADRp0gQbN27EgQMHYGNjw3FURQclGAWAMeDBe8XrSuKXkD6PBc9MCPOWXtwGRgghxdTbt2/Rpk0bnD59Gnw+HwMHDoRAIOA6rCKF2ucLwPt0IDED4AFwDj2PdADmrbzBtzTnOjRCCCl2bt26hc6dOyMjIwOWlpZch1NkUQtGAXiU3XpRxhZgIaEAFP0vCCGEFKwDBw6gSZMmsLe3R3h4OBo3bsx1SEUWJRgF4HGC4rmirQwZ/90HAJi3aMBhRIQQUvxkZmZi8uTJ8PPzw6VLl1CmTBmuQyrS6BZJAVAmGOXS4wCpDCbuLjAtW4rboAghpJhIT0+HWCyGs7MzLl68CGdnZ/D59P3a0CjBKADKBKPMiwcAAPMmdbkLhhBCipHY2FgEBgZCJBLh/PnzcHV15TqkYoMSjALwJLsPRqnrVwAAZs3qcxgNIYQUD5GRkQgICAAAHDx4kFZCLWDURmRgCemKJdoBwDVMMf+FOSUYhBBiUHv27EHz5s3h5uaGiIgINGhA/d4KGiUYBvYk+/ZIKWEWzDNSICjlBNPSztwGRQghRVxGRga6du2Kc+fO0W0RjlCCYWDK/hcVWCIAQFi5LHfBEEJIEZaamooNGzYAAPr374+tW7fC3JzmG+IKJRgG9jJJ8Vwq4RUAQFi9AofREEJI0fTixQs0a9YM48ePV62ESn0uuEUJhoHFJCueHZ89AgCYedfiMBpCCCl6wsLC4O3tjcTERISFhcHDw4PrkAgowTA4VYLxWDHBlln9ahxGQwghRUtoaChatWqFypUrIzw8HLVq0Ze4woISDAN7nZ1gOCXGgu9gC4GLI7cBEUJIEeLp6Yn58+fj9OnTKFmyJNfhkBwowTAgiQyIS1W8dk6Kg6hGBbonSAghX0ksFqNHjx6IiIiAUCjE5MmTIRQKuQ6LfIISDAN6kwrIGWDKZLBL/QBhNergSQghXyMqKgqNGzfGyZMnkZSUxHU4JB+UYBiQ8vaIc0YC+GAQVi3HbUCEEGLEzp07By8vL2RlZeHq1avw8fHhOiSSD0owDEjZwdPpQywAQFjFg7tgCCHEiGVmZqJ///6oV68erl69iqpVq3IdEvkMWovEgGLEiueS7xRzYJhW9uAuGEIIMUJSqRRpaWmwsbHBmTNn4OHhAVNTU67DIlqgBMOAVLdIkuIgcC4BgZ01twERQogR+fDhA3r27AmBQIB///0XlSpV4jokogPOb5GsWrUKHh4eMDMzQ8OGDREeHp7v/itWrECVKlVgbm4Od3d3TJw4ERkZGQUUrW5iVAnGG7o9QgghOnj48CEaNmyI69evY+rUqTQCzwhxmmDs2rULkyZNwpw5c3D9+nXUqVMHvr6+ePv2rcb9d+zYgWnTpmHOnDm4f/8+NmzYgF27dmHGjBkFHLl2VH0wxHEQVqEOnoQQoo2TJ0+iYcOGMDExQXh4OFq3bs11SOQLcJpgLFu2DMOGDcPgwYNRvXp1rF69GhYWFti4caPG/UNDQ9G0aVP06dMHHh4eaNeuHXr37v3ZVg+uKG+RlEx6C1NqwSCEEK3cu3cPTZs2RVhYGCpUoOH9xoqzBCMrKwuRkZFqw4z4fD58fHwQFham8ZgmTZogMjJSlVBERUXh2LFj8Pf3z/M6mZmZEIvFao+CkCZRPADAMSUeQurgSQghecrKysKhQ4cAAOPHj8fhw4dha2vLcVTka3DWyTM+Ph4ymQzOzs5q5c7Oznjw4IHGY/r06YP4+Hg0a9YMjDFIpVKMHDky31skCxcuxM8//6zX2LURn6Z4FkkyYJ6VTn0wCCEkD/Hx8ejWrRuuXr2KR48eoUyZMtTnogj4qhaMzMxMfcWhlXPnzmHBggX43//+h+vXr2Pfvn04evQofv311zyPmT59OpKSklSPly9fFkis77ITDLvUD+BbWUDgQJk4IYR86s6dO/Dy8sKDBw8QEhKCMmXKcB0S0ROdWjD+/fdf7Ny5ExcvXsTLly8hl8thaWmJevXqoV27dhg8eDBKlSql1bkcHR0hEAgQFxenVh4XFwcXFxeNx8yaNQv9+/fH0KFDAQC1atVCamoqhg8fjp9++gl8fu58SSQSQSQS6fI29ULZgmGf+gEmrrQADyGEfOrKlSv45ptvUKFCBZw7dw5ly5blOiSiR1q1YOzfvx+VK1fGkCFDYGJigqlTp2Lfvn04ceIE1q9fj5YtW+L06dMoX748Ro4ciXfv3n32nEKhEA0aNEBISIiqTC6XIyQkBI0bN9Z4TFpaWq4kQiAQAAAYY9q8lQLzXpVgJMDElVZQJYSQT9WsWROjRo3CpUuXKLkogrRqwVi8eDGWL1+O9u3ba2wl6NmzJwAgJiYGf/75J7Zt24aJEyd+9ryTJk3CwIED4enpCW9vb6xYsQKpqakYPHgwAGDAgAFwc3PDwoULAQCdOnXCsmXLUK9ePTRs2BBPnjzBrFmz0KlTJ1WiUVi8y9GCQUu0E0KIQkZGBiZOnIgJEyagSpUqWLx4MdchEQPRKsHIa1THp9zc3PDbb79pffFevXrh3bt3mD17Nt68eYO6devi+PHjqo6fL168UEtoZs6cCR6Ph5kzZyImJgYlS5ZEp06dMH/+fK2vWVDUbpFUoASDEELevHmDwMBA3Lx5E35+fqhSpQrXIRED4jEd7y2cPXvWqCc9EYvFsLW1RVJSEmxsbAx2nVFHgWNPgLEnVmB4l9KwHdbdYNcihJDC7vr16wgICIBMJsPBgwfh5eXFdUjkC+hSh+o8isTPzw8VKlTAvHnzCmxEhjF6n654tk/9AAF18iSEFGMZGRno2LEjXFxcEBERQclFMaFzghETE4OxY8ciODgY5cuXh6+vL3bv3o2srCxDxGe04nN28qQ+GISQYkgulyMjIwNmZmY4dOgQLly4ADc3N67DIgVE5wTD0dEREydOxI0bN3D16lVUrlwZo0ePRqlSpTBu3DjcvHnTEHEancQMxZ0nmzQxjSIhhBQ7aWlp+PbbbzFgwAAAgKenJ8zNzTmOihSkr5poq379+pg+fTrGjh2LlJQUbNy4EQ0aNEDz5s1x9+5dfcVolCRSRYJhIpdC4FSC42gIIaTgvHz5Es2aNcOxY8fQu3dvrsMhHPmiBEMikSA4OBj+/v4oW7YsTpw4gb/++gtxcXF48uQJypYtix49eug7VqMilSueRbaW4JlyNiM7IYQUqCtXrsDLywvv37/H5cuX0aVLF65DIhzRueb7/vvv8c8//4Axhv79+2Px4sWoWbOmarulpSWWLFmi9YyeRZUywTB1NNxIFUIIKWyOHTuGihUrYt++fXBycuI6HMIhnVsw7t27hz///BOvX7/GihUr1JILJUdHR5w9e1YvARorKVMs1GNW0o7bQAghxMDkcjkuX74MAJg7dy5CQkIouSC6Jxhz5sxBjx49cq3vIZVKceHCBQCAiYkJWrZsqZ8IjZCcAXIoEgwRJRiEkCIsOTkZgYGBaN26NV6+fAk+n8/J+k+k8NE5wWjdujUSEhJylSclJRn1BFz6JJN/fC10tucuEEIIMaDo6Gg0adIE58+fx4EDB+Du7s51SKQQ0bkPBmMMPB4vV/n79+9haWmpl6CMnTRHgmHmRAkGIaToiYyMhJ+fH2xtbREWFobq1atzHRIpZLROMLp27QoA4PF4GDRokFoTmEwmw61bt9CkSRP9R2iEciYYQhcaokoIKXo8PDzQsWNHLFmyBCVK0N85kpvWCYatrS0ARQuGtbW12oQpQqEQjRo1wrBhw/QfoRFSa8Fwpf94hJCiQSqV4pdffsHw4cNRunRpbNq0ieuQSCGmdYKh/EXy8PDA5MmT6XZIPiQZWQCEAAAhzeJJCCkCEhMT8e233+L06dOoWbMmevbsyXVIpJDTuQ/GnDlzDBFHkZLxNhGAE0xkEghK2HIdDiGEfJVHjx6hc+fOePv2LU6cOIG2bdtyHRIxAlolGPXr10dISAjs7e1Rr149jZ08la5fv6634IxVVlwCACcImDzfz4oQQgq79PR0tGrVCjY2Nrh69SoqVarEdUjESGiVYAQEBKg6dQYEBFCl+RmZ8WIAgAmTf2ZPQggpnBhjkMvlMDc3x9atW9GgQQPY2dlxHRYxIjzGGOM6iIIkFotha2uLpKQk2NgYZhrv/zacRmCKD2ykabj9g4VBrkEIIYYikUgwbtw4MMawevVqrsMhhYgudajOE20NHToU586d+9LYioWsxDQAgAmPWjAIIcbl/fv3aNeuHTZs2AAvLy+uwyFGTOcE4927d/Dz84O7uzt+/PFH3Lx50xBxGbWs5FQAgAndSSKEGJF79+7B29sbd+7cQUhICL777juuQyJGTOcE4+DBg4iNjcWsWbMQERGB+vXro0aNGliwYAGePXtmgBCNj0Sc3YKh86dLCCHcWbt2LSwtLREREYHmzZtzHQ4xcl9UBdrb22P48OE4d+4cnj9/jkGDBmHr1q2oWLGivuMzSlnJGQAAEwE1YRBCCjfGGO7fvw8AWLx4MS5fvgwPDw9ugyJFwld9x5ZIJLh27RquXr2KZ8+ewdnZWV9xGTVJcjoASjAIIYVbZmYmBg8ejPr16yMmJgZCoRDW1tZch0WKiC9KMM6ePYthw4bB2dkZgwYNgo2NDY4cOYJXr17pOz6jJEnLBEAJBiGk8IqLi0ObNm2wc+dObNiwAW5ublyHRIoYnWfydHNzQ0JCAvz8/LB27Vp06tRJbeEzAkgyJQAAE+qEQQgphO7cuYMOHTogKysLFy5cgLe3N9chkSJI5wRj7ty56NGjB024kg9JphQAYErDSAghhZCtrS1q1aqF1atXo3Tp0lyHQ4oonb9iDxs2jJKLfDCJFNLs5VSpBYMQUlgwxvDHH3/g/fv3cHd3x5EjRyi5IAalVQtG165dERQUBBsbG3Tt2jXfffft26eXwIyVPCUNMr7iYzUxFXAcDSGEAGlpaRgyZAh27doFe3t79O/fn+uQSDGgVYJha2urWn/ExsaG1iLJhzw5FTK+IrGgTp6EEK7FxMQgICAA9+/fR3BwMLp168Z1SKSY0CrB2LRpk+p1UFCQoWIpEuTJqZAKslsw6A4JIYRDaWlpaNy4MQDg0qVLqFevHscRkeJE5yqwTZs2SExMzFUuFovRpk0bfcRk1OTJaR9bMCjBIIRwhDEGCwsLLF26FOHh4ZRckAKn8yiSc+fOISsrK1d5RkYGLl68qJegjJnaLRJKMAghBUwul2P27Nng8Xj49ddf0aNHD65DIsWU1gnGrVu3VK/v3buHN2/eqH6WyWQ4fvw4TdQCgKWkQcqnWySEkIKXkpKC/v374+DBg1i0aBHX4ZBiTusEo27duuDxeODxeBpvhZibm+PPP//Ua3DGSJ7ysQWD+ngSQgrK8+fP0blzZ0RFReHQoUPo2LEj1yGRYk7rBCM6OhqMMZQvXx7h4eEoWbKkaptQKISTkxMEAhqWKU/+2IJBo1QJIQVl/vz5SE5ORlhYGGrWrMl1OIRon2CULVsWgOL+HsmbPDkVcmrBIIQUkNevX6NUqVJYvnw50tPT4ejoyHVIhADQMsE4dOgQ2rdvD1NTUxw6dCjffTt37qyXwIyVYpiqJQDAlPpgEEIMRCaTYcqUKVi/fj3u37+PUqVKwdLSkuuwCFHRKsEIDAzEmzdv4OTkhMDAwDz34/F4kMlk+orNKCmGqdoAAASUYBBCDCApKQm9e/fGiRMnsHz5cri6unIdEiG5aJVg5LwtQrdI8idPTv3YB4MSDEKInj19+hSdOnXC69ev8e+//6Jdu3Zch0SIRjrPg6FJYmIiLYCWTbEWSXYfDEowCCF6JpFIYG1tjatXr6JKlSpch0NInnSuAhctWoRdu3apfu7RowccHBzg5uaGmzdv6jU4YyRPSYM0O8GgFgxCiL78888/SElJQdWqVXHlyhVKLkihp3MVuHr1ari7uwMATp06hdOnT+P48eNo3749fvzxR70HaGxyzuRJLRiEkK8lkUgwZswY9OnTB8HBwQBAC04So6DzLZI3b96oEowjR46gZ8+eaNeuHTw8PNCwYUO9B2hsWI5hqjSTJyHkayQkJKBHjx64cOEC1qxZg0GDBnEdEiFa07kKtLe3x8uXLwEAx48fh4+PDwDFwjrFfQQJkD3RFq2mSgj5SqmpqWjUqBFu3ryJ06dPY/jw4VyHRIhOdG7B6Nq1K/r06YNKlSrh/fv3aN++PQDgv//+Q8WKFfUeoDFhjNFiZ4QQvbC0tMSECRPg5+eH8uXLcx0OITrTOcFYvnw5PDw88PLlSyxevBhWVlYAgNjYWIwePVrvARoTlpoOMEaLnRFCvghjDCtWrIBAIMC4ceOK/d9UYtx0TjBMTU0xefLkXOUTJ07US0DGTJ6SBgCQZd8ioU6ehBBtZWZmYvTo0di4cSOmT5/OdTiEfLUvmgfj8ePHOHv2LN6+fZtr4q3Zs2frJTBjpEowhCIANEyVEKKdt2/folu3bggPD8eWLVvQv39/rkMi5KvpnGCsW7cOo0aNgqOjI1xcXNSGS/F4vOKdYCSnKp6FQgC02BkhRDtTpkzB48ePcf78eTRq1IjrcAjRC50TjHnz5mH+/PmYOnWqIeIxasoEQ2qqSDBouXZCSH7EYjFsbGywbNky/Prrr6opAAgpCnRuxP/w4QN69OhhiFiMHsu+RSI3MQVALRiEEM0YY1iwYAGqVq2KuLg4ODg4UHJBihydE4wePXrg5MmThojF6MkzswAA0uwEg1owCCGfSk9PR9++ffHTTz9h+PDhKFmyJNchEWIQOt8iqVixImbNmoUrV66gVq1aMDU1Vds+btw4vQVnbFimBAA+ThVOLRiEkBxev36NwMBA3LlzB7t27ULPnj25DokQg9E5wVi7di2srKxw/vx5nD9/Xm0bj8cr1gkGJNkJhoCWayeE5Pb69Wt8+PABFy9eRIMGDbgOhxCD0jnBiI6ONkQcRUKuFgxKMAghAE6ePIlWrVrB09MT9+/fh4nJF80QQIhR+eIqMCsrCw8fPoRUKtVnPEaNZaknGDSTJyHFm1wux+zZs+Hr64utW7cCACUXpNjQuQpMS0vDd999BwsLC9SoUQMvXrwAAHz//ff47bff9B6gMVEmGFIeJRiEFHepqano0aMH5s2bh4ULF2LIkCFch0RIgdK5Cpw+fTpu3ryJc+fOwczMTFXu4+ODXbt26TU4Y/NpCwbdIiGkeEpJSUGzZs1w4sQJ7N+/H9OmTVOblJCQ4kDntroDBw5g165daNSokdp/mBo1auDp06d6Dc7YMOUw1ewWDOrkSUjxZGVlhcDAQHTp0gW1a9fmOhxCOKFzgvHu3Ts4OTnlKk9NTS32GTqTKPqjyHiKzIJaMAgpXjZv3gwTExP07dsXc+bM4TocQjilcxXo6emJo0ePqn5WJhXr169H48aNdQ5g1apV8PDwgJmZGRo2bIjw8PB8909MTMSYMWPg6uoKkUiEypUr49ixYzpf1xCUo0ioBYOQ4kUmk2HKlCkYNGgQQkNDuQ6HkEJB5xaMBQsWoH379rh37x6kUilWrlyJe/fuITQ0NNe8GJ+za9cuTJo0CatXr0bDhg2xYsUK+Pr64uHDhxpbSbKysvDNN9/AyckJwcHBcHNzw/Pnz2FnZ6fr2zAIlqW4RSIDtWAQUlyIxWL06dMH//77L5YvX47x48dzHRIhhYLOVWCzZs1w48YNSKVS1KpVCydPnoSTkxPCwsJ0njhm2bJlGDZsGAYPHozq1atj9erVsLCwwMaNGzXuv3HjRiQkJODAgQNo2rQpPDw80LJlS9SpU0fXt2EQLEtxi0SafYuEWjAIKfrGjBmDixcv4ujRo5gwYUKxv1VMiBKPMca4uHBWVhYsLCwQHByMwMBAVfnAgQORmJiIgwcP5jrG398fDg4OsLCwwMGDB1GyZEn06dMHU6dOhUCgeeGPzMxMZGZmqn4Wi8Vwd3dHUlISbGxs9Pqe4obNQcqBM/CfeRYZMMGlwYC7fi9BCCkksrKyIBQKERMTA7FYjGrVqnEdEiEGJxaLYWtrq1UdqvV3bKlUqlZRA0BcXBx+/vlnTJkyBZcuXdIpyPj4eMhkMjg7O6uVOzs7482bNxqPiYqKQnBwMGQyGY4dO4ZZs2Zh6dKlmDdvXp7XWbhwIWxtbVUPQ65YqJoHA9SCQUhRtnbtWtSuXRvv37+Hm5sbJReEaKB1FThs2DC1dUaSk5Ph5eWFVatW4cSJE2jdurXBO1vK5XI4OTlh7dq1aNCgAXr16oWffvoJq1evzvOY6dOnIykpSfV4+fKlweJjWRIwfEwwaLEzQooWqVSKcePGYcSIEfDx8dF7KyghRYnWnTwvX76Mv/76S/Xzli1bIJPJ8PjxY9ja2mLq1Kn4/fff4e/vr9X5HB0dIRAIEBcXp1YeFxcHFxcXjce4urrC1NRU7XZItWrV8ObNG1Vz5adEIhFEIpFWMX0tliWBnPcxZ6Pl2gkpOj58+IBevXrh7Nmz+PvvvzFy5EiuQyKkUNO6BSMmJgaVKlVS/RwSEoJu3brB1tYWgKLvxN27d7W+sFAoRIMGDRASEqIqk8vlCAkJyXO4a9OmTfHkyRPI5XJV2aNHj+Dq6qoxuShoLFMCGf9jzkYtGIQUHTdv3sTNmzdx8uRJSi4I0YLWCYaZmRnS09NVP1+5cgUNGzZU256SkqLTxSdNmoR169Zh8+bNuH//PkaNGoXU1FQMHjwYADBgwABMnz5dtf+oUaOQkJCA8ePH49GjRzh69CgWLFiAMWPG6HRdQ2ESiWqacIDWIiGkKLh27RrkcjlatWqFqKgotG7dmuuQCDEKWleBdevWVa0GePHiRcTFxaFNmzaq7U+fPkWpUqV0univXr2wZMkSzJ49G3Xr1sWNGzdw/PhxVcfPFy9eIDY2VrW/u7s7Tpw4gYiICNSuXRvjxo3D+PHjMW3aNJ2uaygsUwKpgBIMQooCxhhWrFiBhg0bYsuWLQAAS0tLjqMixHhoPUz1/PnzaN++PVxdXREbG4vevXtjw4YNqu2jR49GamoqNm/ebLBg9UGXITa6etGkH97FJKHrpMMAgGfjABoST4jxycrKwpgxY7B+/XpMnjwZv/32W55D4QkpTnSpQ7Xu5NmyZUtERkbi5MmTcHFxQY8ePdS2161bF97e3l8WcRHBJBJIlSup8ii5IMQYJScno0OHDrh69So2bdqEQYMGcR0SIUZJp6nCq1Wrlud47+HDh+slIGOm6OSpSDDo9gghxsnKygq1atXCwoUL0bRpU67DIcRoaVUNXrlyResTpqWl6TSapChhWVmQChQ5GyUYhBiXw4cP49ixY+DxeFi1ahUlF4R8Ja2qwf79+8PX1xd79uxBamqqxn3u3buHGTNmoEKFCoiMjNRrkEYjSwpZ9kqqtNAZIcaBMYZFixYhICAAu3bt4jocQooMrW6R3Lt3D3///TdmzpyJPn36oHLlyihVqhTMzMzw4cMHPHjwACkpKejSpQtOnjyJWrVqGTruQolJJJCZKT5SmiackMIvIyMDw4YNw7Zt2zBz5kz8/PPPXIdESJGhVYJhamqKcePGYdy4cbh27RouXbqE58+fIz09HXXq1MHEiRPRunVrODg4GDreQo1JZao+GNSCQUjhN2zYMAQHB+Off/7Bt99+y3U4hBQpOnXyBABPT094enoaIhajxhgDciQY1IJBSOEll8vB5/Mxc+ZMjBs3Dl5eXlyHREiRQ9WgvshkAABp9lTh1IJBSOG0Z88eNGvWDCkpKahSpQolF4QYCFWDesKkigSDWjAIKZzkcjl+/vln9OzZE2XLlqWJswgxMJ1vkZA8ZCcYymGq1IJBSOGRlpaGQYMGYc+ePZg3bx5mzJgBHs2ER4hBUYKhJ0ym3oJhQn+7CCk0zp8/j2PHjmHfvn3o0qUL1+EQUix8VYKRkZEBMzMzfcVi3FS3SGiiLUIKiydPnqBChQpo3749oqKi4OTkxHVIhBQbOleDcrkcv/76K9zc3GBlZYWoqCgAwKxZs9QWPytuVH0w6BYJIYXCtm3bULNmTezYsQMAKLkgpIDpXA3OmzcPQUFBWLx4MYRCoaq8Zs2aWL9+vV6DMypSqeLJVPGZUCdPQrghl8sxffp09O/fH71790b37t25DomQYknnanDLli1Yu3Yt+vbtq9YLu06dOnjw4IFegzMmyhYMuakpAGrBIIQLqamp6NKlCxYtWoQlS5Zg48aNEIlEXIdFSLGkcx+MmJgYVKxYMVe5XC6HRCLRS1DGSHWLxIRaMAjhikgkgomJCY4cOQJ/f3+uwyGkWNM5wahevTouXryIsmXLqpUHBwejXr16egvM6ChHkVALBiEF7sKFCzAxMUGTJk2wd+9ersMhhOALEozZs2dj4MCBiImJgVwux759+/Dw4UNs2bIFR44cMUSMRuHTTp7UgkFIwdiwYQNGjRqFrl27okmTJlyHQwjJpnM1GBAQgMOHD+P06dOwtLTE7Nmzcf/+fRw+fBjffPONIWI0DsqJtrI7eVILBiGGJZVKMXHiRAwdOhRDhw7F1q1buQ6JEJLDF82D0bx5c5w6dUrfsRg1lj2KRE4tGIQUiGHDhmHr1q1YtWoVRo8ezXU4hJBP6FwNli9fHu/fv89VnpiYiPLly+slKKOkbMEwoT4YhBSEMWPG4MSJE5RcEFJI6VwNPnv2DLLsDo05ZWZmIiYmRi9BGSPVMNXsFgyaKpwQ/Tt16hQ6d+6MrKwseHp6om3btlyHRAjJg9a3SA4dOqR6feLECdja2qp+lslkCAkJgYeHh16DMybKtUiULRgmtFAjIXrDGMOqVaswYcIE+Pj4ICMjQ22iP0JI4aN1ghEYGAgA4PF4GDhwoNo2U1NTeHh4YOnSpXoNzqio5sHIvkVCLRiE6IVEIsH333+PNWvWYOLEiVi8eDFMTGidRkIKO63/l8rlcgBAuXLlEBERAUdHR4MFZYxyDVOlFgxC9OLIkSPYuHEjNmzYgCFDhnAdDiFESzp/DYiOjjZEHMYvexSJarEzasEg5Ku8ffsWTk5O6NKlC+7du6dxBmFCSOH1Re2MqampOH/+PF68eIGsrCy1bePGjdNLYMZG2YIhpWGqhHy1I0eOoE+fPggKCkLXrl0puSDECOmcYPz333/w9/dHWloaUlNT4eDggPj4eFhYWMDJyanYJhig5doJ+WqMMSxZsgRTp05Fp06divfkfYQYOZ2rwYkTJ6JTp0748OEDzM3NceXKFTx//hwNGjTAkiVLDBGjUVCOIqGpwgn5MpmZmRg8eDCmTJmCadOmYf/+/bC2tuY6LELIF9K5Grxx4wZ++OEH8Pl8CAQCZGZmwt3dHYsXL8aMGTMMEaNRUHXy5Ct6d1ILBiG6kcvliIqKwrZt27BgwQLw+fSfiBBjpvMtElNTU9V/fCcnJ7x48QLVqlWDra0tXr58qfcAjYYqwaAWDEJ0cePGDZiYmKBmzZo4d+4cJRaEFBE6/0+uV68eIiIiAAAtW7bE7NmzsX37dkyYMAE1a9bUe4DG4tNbJNSCQcjn7du3D02bNsWcOXMAgJILQooQnf83L1iwAK6urgCA+fPnw97eHqNGjcK7d++wZs0avQdoNCSKYarS7FskJvR3kpA8McYwb948dOvWDR07dqSVUAkpgnS+ReLp6al67eTkhOPHj+s1IGP1aR8MSjAIyduYMWPw999/45dffsHMmTPB49HEMYQUNXqrBq9fv46OHTvq63TGR0YJBiHa6tatG/bs2YNZs2ZRckFIEaVTNXjixAlMnjwZM2bMQFRUFADgwYMHCAwMhJeXl2o68eJINdEWL3s1VUowCFETHh6O4cOHQyaToW3btujevTvXIRFCDEjranDDhg1o3749goKCsGjRIjRq1Ajbtm1D48aN4eLigjt37uDYsWOGjLVwU90iUXyk1MmTkI/++ecftGzZErdv30ZycjLX4RBCCoDW1eDKlSuxaNEixMfHY/fu3YiPj8f//vc/3L59G6tXr0a1atUMGWehp1qunYapEqIil8sxc+ZM9OnTBz169MDZs2dhZ2fHdViEkAKgdTX49OlT9OjRAwDQtWtXmJiY4Pfff0fp0qUNFpwxYdmjSGS87BYMuq1MCPbs2YMFCxZg0aJF2Lx5M8zMzLgOiRBSQLQeRZKeng4LCwsAAI/Hg0gkUg1XJVB18pTyFJ08abl2UpylpqbC0tISPXv2RPny5eHl5cV1SISQAqbTMNX169fDysoKACCVShEUFARHR0e1fYrrYmeqYarUgkGKuUuXLqF79+7YtGkT2rdvT8kFIcWU1glGmTJlsG7dOtXPLi4uuSbH4fF4xTbBwCcJBvXBIMXRpk2bMGLECDRp0oQSC0KKOa0TjGfPnhkwDOP3cZgqLXZGih+ZTIYpU6Zg2bJlGDZsGP766y8IhUKuwyKEcIiqQX2RUQsGKb5SU1Nx8uRJ/PHHH1izZg0lF4QQ3acKJ5opR5FIeTQPBik+njx5AhMTE3h4eCAyMpISC0KIClWDeqJaTTX7I6WZPElRd+bMGXh7e+OHH34AAEouCCFqqBrUF2UfDEowSDHwv//9D+3atYOXlxfWr1/PdTiEkEKIqkE9UQ1TpQSDFHGTJ0/GmDFjMGbMGBw9ehT29vZch0QIKYS+qBp8+vQpZs6cid69e+Pt27cAgH///Rd3797Va3BGRUYtGKR4aNy4MdasWYOVK1fCxIS6cRFCNNO5Gjx//jxq1aqFq1evYt++fUhJSQEA3Lx5E3PmzNF7gMZCNUwVihm2KMEgRcn9+/fx008/gTGGbt26Yfjw4VyHRAgp5HSuBqdNm4Z58+bh1KlTap262rRpgytXrug1OKNCCQYpov799180atQIBw8eRGJiItfhEEKMhM7V4O3bt9GlS5dc5U5OToiPj9dLUMaISbMXOwMNUyVFA2MMy5cvR8eOHdGiRQuEhoZSfwtCiNZ0rgbt7OwQGxubq/y///6Dm5ubXoIyRkwqgxw8yLNbMGiiLWLsdu3ahUmTJuHHH3/EgQMHYGNjw3VIhBAjonMPrW+//RZTp07Fnj17wOPxIJfLcfnyZUyePBkDBgwwRIzGQSqDjP9xCVVqwSDGSiqVwsTEBN27d4e9vT18fX25DokQYoR0rgYXLFiAqlWrwt3dHSkpKahevTpatGiBJk2aYObMmYaI0SgwmXqCQS0YxBjdunUL1atXx6VLl2BiYkLJBSHki+ncgiEUCrFu3TrMmjULd+7cQUpKCurVq4dKlSoZIj7jIZVBKvj4cdJy7cTYHDhwAP369UOlSpVQpkwZrsMhhBg5nROMS5cuoVmzZihTpgz9EcqBfXKLxFSQz86EFCKMMSxcuBA//fQTunXrhs2bN8PS0pLrsAghRk7nhvw2bdqgXLlymDFjBu7du2eImIwSk0oh4yvyNR4APrVgECPx4cMHrF69GrNnz8bu3bspuSCE6IXOCcbr16/xww8/4Pz586hZsybq1q2L33//Ha9evTJEfMYjRwsGzYFBjEFsbCzevn0LBwcH3LlzBz///DP4fPrlJYToh85/TRwdHTF27FhcvnwZT58+RY8ePbB582Z4eHigTZs2XxTEqlWr4OHhATMzMzRs2BDh4eFaHbdz507weDwEBgZ+0XX1ieXog0EJBinsIiMj4eXlhbFjxwIADUElhOjdV1WF5cqVw7Rp0/Dbb7+hVq1aOH/+vM7nUI61nzNnDq5fv446derA19dXtcZJXp49e4bJkyejefPmXxq+fslkkPGoBYMUfrt370bz5s3h5uaGlStXch0OIaSI+uKq8PLlyxg9ejRcXV3Rp08f1KxZE0ePHtX5PMuWLcOwYcMwePBgVK9eHatXr4aFhQU2btyY5zEymQx9+/bFzz//jPLly3/pW9ArasEgxuDXX39Fr1690LVrV5w/fx6urq5ch0QIKaJ0rgqnT5+OcuXKoU2bNnjx4gVWrlyJN2/eYOvWrfDz89PpXFlZWYiMjISPj8/HgPh8+Pj4ICwsLM/jfvnlFzg5OeG777777DUyMzMhFovVHgZBfTCIEXB2dsbChQuxdetWmJmZcR0OIaQI03mY6oULF/Djjz+iZ8+ecHR0/KqLx8fHQyaTwdnZWa3c2dkZDx480HjMpUuXsGHDBty4cUOrayxcuBA///zzV8WpDSaVQiaiBIMUPi9evMCBAwcwbtw4WgWVEFJgdE4wLl++bIg4tJKcnIz+/ftj3bp1Wic306dPx6RJk1Q/i8ViuLu76z84qQwyc8XHSdOEk8IiNDQUXbp0gYWFBQYMGAA7OzuuQyKEFBNaJRiHDh1C+/btYWpqikOHDuW7b+fOnbW+uKOjIwQCAeLi4tTK4+Li4OLikmv/p0+f4tmzZ+jUqZOqTC6XAwBMTEzw8OFDVKhQQe0YkUgEkUikdUxfKudU4TRNOCkMNm/ejOHDh6Nhw4bYu3cvJReEkAKlVYIRGBiIN2/ewMnJKd8hoTweDzKZTOuLC4VCNGjQACEhIarzyuVyhISEqIbP5VS1alXcvn1brWzmzJlITk7GypUrDdMyoSVFJ09FgkEtGIRru3btwqBBg/Ddd9/hf//7H4RCIdchEUKKGa0SDGUrwaev9WHSpEkYOHAgPD094e3tjRUrViA1NRWDBw8GAAwYMABubm5YuHAhzMzMULNmTbXjld/KPi0vcFJqwSDcY4yBx+MhICAA27ZtQ58+fcDj0bSyhJCCp3NVuGXLFmRmZuYqz8rKwpYtW3QOoFevXliyZAlmz56NunXr4saNGzh+/Liq4+eLFy8QGxur83kLnFwOKZ/6YBDuREVFoXHjxrh16xbMzMzQt29fSi4IIZzhMcaYLgcIBALExsbCyclJrfz9+/dwcnLS6RYJF8RiMWxtbZGUlKTX2QuflmyO0EpNMLPXItR1Bg5+q7dTE/JZ586dQ/fu3WFvb4/Dhw+jatWqXIdECCmCdKlDdf6urWyC/dSrV69ga2ur6+mKBGWOpmzBoGGqpCCtXbsW33zzDerWrYurV69SckEIKRS0HqZar1498Hg88Hg8tG3bFiYmHw+VyWSIjo7WeaKtIiO7X4qcJtoiBSw+Ph7Tp0/HiBEjsHz5cpiamnIdEiGEANAhwVCO8rhx4wZ8fX1hZWWl2iYUCuHh4YFu3brpPUCjoGzBoKnCSQH58OEDBAIBHB0dcfv2bZQqVYrrkAghRI3WCcacOXMAAB4eHujVqxdNM5xTdi8W5SgS6uRJDOnBgwfo3LkzvL29sW3bNkouCCGFks5V4cCBAym5+NQnfTBomCoxlBMnTqBRo0YwMTEpkCnwCSHkS2nVguHg4IBHjx7B0dER9vb2+Q59S0hI0FtwRiM7waAWDGJIf/zxByZOnAg/Pz/s2LGj2HaqJoQYB60SjOXLl8Pa2lr1msbWq2Ny9QSDWjCIIaSkpGDixIlYtGgRBNmzxhJCSGGlVYIxcOBA1etBgwYZKhbj9cktEmrBIPoSHx+PU6dOoXfv3pg+fTol94QQo6FzVXj9+nW19UAOHjyIwMBAzJgxA1lZWXoNzmgwasEg+nfnzh14eXlh4sSJSEpKouSCEGJUdK4KR4wYgUePHgFQTE3cq1cvWFhYYM+ePZgyZYreAzQKygRDQC0YRD8OHz6Mxo0bw9bWFlevXqX+FoQQo6NzVfjo0SPUrVsXALBnzx60bNkSO3bsQFBQEPbu3avv+IyKsgXDhL5okq+wd+9eBAQE4JtvvsGlS5dQtmxZrkMihBCdfdFU4coVVU+fPg1/f38AgLu7O+Lj4/UbnZFg2Z+HVJlgUP878hV8fHywdOlSBAcHq01oRwghxkTnBMPT0xPz5s3D1q1bcf78eXTo0AEAEB0drVoBtdhR9cHInsmTWjCIjmJjY9GxY0dERUXB1tYWEydOBJ9P99oIIcZL65k8lVasWIG+ffviwIED+Omnn1CxYkUAQHBwMJo0aaL3AI1C9kye1IJBvsT169cREBAAmUyGpKQkrsMhhBC90DnBqF27ttooEqXff/+9+I7N/2QUCbVgEG3t2bMHAwcORI0aNXDgwAG4ublxHRIhhOiFzgmGUmRkJO7fvw8AqF69OurXr6+3oIxOdh8MGS12RnTw7t07DBkyBAEBAdi4cSPMzc25DokQQvRG5wTj7du36NWrF86fPw87OzsAQGJiIlq3bo2dO3eiZMmS+o6x8KOpwokO0tLSwOfzUbJkSURERKBKlSo0xwUhpMjRuSr8/vvvkZKSgrt37yIhIQEJCQm4c+cOxGIxxo0bZ4gYCz32yWqqNNEWycurV6/QvHlz1f+VqlWrUnJBCCmSdG7BOH78OE6fPo1q1aqpyqpXr45Vq1ahXbt2eg3OaNBU4UQLV69eRWBgIIRCIcaMGcN1OIQQYlA6V4VyuRympqa5yk1NTVXzYxQ7NFU4+Yxt27ahZcuWqFChAiIiIlCnTh2uQyKEEIPSuSps06YNxo8fj9evX6vKYmJiMHHiRLRt21avwRkNZQuGQJF4UQsG+dT169fRp08fhISEwMnJietwCCHE4HS+RfLXX3+hc+fO8PDwgLu7OwDg5cuXqFmzJrZt26b3AI0CtWAQDZKTkxEWFoZ27dphyZIl4PF41N+CEFJs6JxguLu74/r16wgJCVENU61WrRp8fHz0HpzRoMXOyCeio6PRuXNnvHnzBlFRUbC2tuY6JEIIKVA6JRi7du3CoUOHkJWVhbZt2+L77783VFzG5dOJtijBKNYuXLiAbt26wdbWFufPn6fkghBSLGldFf7999/o3bs3rl27hsePH2PMmDH48ccfDRmb0VAOU5XSRFvF3sGDB9G2bVvUqlULV69eRfXq1bkOiRBCOKF1VfjXX39hzpw5ePjwIW7cuIHNmzfjf//7nyFjMx7KmTypBaPYa9iwIaZMmYITJ06gRIkSXIdDCCGc0boqjIqKwsCBA1U/9+nTB1KpFLGxsQYJzKh80geDEoziJTExEUOGDEFcXBxcXFwwf/58jUO5CSGkONG6KszMzISlpeXHA/l8CIVCpKenGyQwo/LJTJ602Fnx8ejRIzRq1Aj79+9HVFQU1+EQQkihoVMnz1mzZsHCwkL1c1ZWFubPnw9bW1tV2bJly/QXnbFQtmDwaLn24uT06dPo0aMHnJ2dER4ejkqVKnEdEiGEFBpaJxgtWrTAw4cP1cqaNGmi9q2t2I7xV020lT1MtZh+DMXJ27dv0blzZ7Ro0QI7d+5ULfxHCCFEQesE49y5cwYMw7gxpt7J05RaMIosiUQCPp8PJycnhISEwMvLCyYmOk8nQwghRR51R9QH5TBVPrVgFGXv379Hu3btMG3aNABA48aNKbkghJA8UIKhDzRVeJF39+5deHt7486dOwgICOA6HEIIKfSoKtSHTxIMmiq8aDl69CgaN24MS0tLREREoFmzZlyHRAghhR5VhfpAU4UXaXv27EHr1q1x+fJleHh4cB0OIYQYBbqBrAdMnj2KhE8TbRUVmZmZuHnzJry9vbF27VqYmJiAz6d/WEII0dYX/cW8ePEi+vXrh8aNGyMmJgYAsHXrVly6dEmvwRkNasEoUuLi4tC6dWv4+/sjJSUFQqGQkgtCCNGRzn819+7dC19fX5ibm+O///5DZmYmACApKQkLFizQe4BGQTkPBiUYRu/GjRvw8vJCdHQ0jh49CisrK65DIoQQo6RzVThv3jysXr0a69atU1tvoWnTprh+/bpegzMan04VTgmGUTpx4gSaNm0KJycnREREoGHDhlyHRAghRkvnqvDhw4do0aJFrnJbW1skJibqIybjwxgYABn1wTBqlSpVQr9+/XDhwgWULl2a63AIIcSo6VwVuri44MmTJ7nKL126hPLly+slKKMjl0PO+zh9JyUYxiMtLQ1TpkxBUlISypcvjzVr1qitt0MIIeTL6FwVDhs2DOPHj8fVq1fB4/Hw+vVrbN++HZMnT8aoUaMMEaNRUN4eASjBMBYxMTFo0aIF/vrrL9y4cYPrcAghpEjReZjqtGnTIJfL0bZtW6SlpaFFixYQiUSYPHkyvv/+e0PEWOgxxiAVUIJhTMLDwxEYGAgTExNcvnwZ9erV4zokAIq1TmQyGddhEEKKCYFAoNafUp90TjB4PB5++ukn/Pjjj3jy5AlSUlJQvXr14t3bnjFqwTAib968QevWrVG7dm3s378fLi4uXIcEsViM+Ph41agsQggpKCKRCI6OjrCxsdHreb94oi2hUIjq1avrMxbjJZerJtkCaLGzwkoul4PH48HFxQW7d+9G27ZtYWZmxnVYEIvFiImJgZWVFRwdHWFqagoej36JCCGGxRiDRCJBUlKSak4rfSYZOicYrVu3zveP35kzZ74qIKPE1IeoUt1Q+KSkpKB///7w8vLCjBkz0KFDB65DUomPj4eVlRVKly5NiQUhpECZm5vD2toar169Qnx8PLcJRt26ddV+lkgkuHHjBu7cuYOBAwfqKy7jwhikAlqqvbB6/vw5OnfujKioKHz33Xdch6NGIpEgMzMTjo6OlFwQQjjB4/Fga2uLmJgYSCQSvfXJ0DnBWL58ucbyuXPnIiUl5asDMko5+mBQ/4vC5fLly+jSpQusrKwQFhaGmjVrch2SGmWHTkN1siKEEG0o/wbJZDK9/T3SW3XYr18/bNy4UV+nMyqMMZpkq5BaunQpqlevjvDw8EKXXORErReEEC4Z4m+Q3lZTDQsLKxQd5jghpxaMwkQmk+Hp06eoXLkytmzZAqFQCKFQyHVYhBBSrOicYHTt2lXtZ8YYYmNjce3aNcyaNUtvgRkVxmip9kIiKSkJvXv3RmRkJKKioor38GlCCOGQztWhra2t2sPBwQGtWrXCsWPHMGfOHEPEWPhRH4xC4cmTJ2jUqBFCQ0OxdetWWFpach0SIYWOv78/eDwerK2tkZaWlud+QUFB4PF4aNWqVb7n8/DwAI/Hw7lz5zRul0gkWLduHdq3b49SpUpBJBLB1tYW9evXxw8//ID79+9/xbv5vKCgIHh7e8PKygoODg7w9/dHaGjoF53r1atXGDlyJMqUKQORSIRSpUph0KBBiI6O1rh/q1atwOPx8n3w+ZorjX///RfffPMN7OzsYGFhgVq1amHx4sWQSCRfFDsXdGrBkMlkGDx4MGrVqgV7e3tDxWSUKMHg1oULFxAYGIiSJUvi6tWrqFKlCtchEVLoxMXF4eTJkwAUQ7cPHDiAPn36GOx6Dx48QEBAAB49egShUAhvb2+0bNkSqampuHHjBpYtW4YVK1Zg48aNBhmFOGHCBKxcuRLm5uZo164dMjIycOrUKZw8eRLBwcEIDAzU+lx37txB69atER8fDw8PD3Ts2BFPnz7F5s2bsX//fly4cAF16tRRO8bPzw8eHh4azxcZGYk7d+6gefPmubYtWrQI06ZNA5/PR8OGDeHk5ISwsDBMnToVp06dwrFjx4yjYzjTkUgkYlFRUboeVmgkJSUxACwpKUlv50y7GMn21h/DyqxgrPVmvZ2W6ODmzZusS5cuLCEhgetQdJKens7u3bvH0tPTuQ6FFAPLly9nAJirqysDwPz8/PLcd9OmTQwAa9myZb7nLFu2LAPAzp49q1b+8uVL5ujoyACwQYMGsfj4+FzHhoSEsHr16rE5c+Z8wbvJ36lTpxgAVqJECfbo0SNVeWhoKBMKhczOzo59+PBBq3PJ5XJWq1YtBoANGTKESSQS1bY//viDAWDVq1dnUqlU6/i8vb0ZALZu3Tq18vDwcMbj8ZipqSk7fvy4qjwpKYm1bt2aAWALFizQ+jra0vZvkS51qM4JRoMGDdjp06d1PazQMESCkXo+gu32HMfKrGDsm616Oy35jKysLLZ48WKWlpbGdShfjBIMUpDq16/PALCDBw8yKysrJhAIWGxsrMZ9vzbB8Pf3VyUX+cnIyGARERG6vA2ttG/fngFgy5cvz7Vt3LhxDABbsmSJVue6ePEiA8AcHBxYcnJyru1NmjRhANiBAwe0Ot+jR48YAGZmZsYSExPVtn333XcMABs2bFiu4x4+fMh4PB5zdHTUKZnRhiESDJ0b9OfNm4fJkyfjyJEjiI2NhVgsVnsUSwyqYaoCukVSIBISEuDn54cZM2Z88f1UUjg9e/ZMde8/NTUVkyZNgru7O8zNzVG/fn0cPnxYte+ePXvQsGFDWFpawtnZGePGjUN6errG8758+RIjRoxA2bJlIRKJ4OTkhK5duyIiIiLXvufOnQOPx8OgQYM0nmvQoEEa+x08f/4co0aNQuXKlWFhYQEHBwfUqFEDI0aMwMOHDzXGNHbsWFSoUAFmZmZwcHBAx44dDfI7ff/+fVy/fh2lSpVCx44d0bVrV8hkMvzzzz96v9a9e/dw7NgxmJubY9myZfnuKxKJ4Onpqdfrp6enq2aV7t69e67tyrKcv0v5iYyMBAA0aNBAY8fx1q1bAwAOHjyo1fm2bdsGAOjUqRNsbW01XktT35fKlSujVKlSiI+Px+XLl7W6Fpe0rg5/+eUXpKamwt/fHzdv3kTnzp1RunRp2Nvbw97eHnZ2dsW3X0aOTp6mlGAY3P379+Ht7Y2bN2/i9OnTaNu2LdchEQPIyspC27ZtsX37djRq1AiNGjXCzZs30aVLF5w+fRrLly9Hnz59YG1tDV9fX8hkMvz5558YOnRornPdvn0b9evXx9q1a2Fubo6uXbuiUqVK2L9/P5o0aYI9e/Z8dbwvX75E/fr1sXr1agCKzpQtW7aESCTCunXrEBYWprZ/WFgY6tSpg1WrVsHU1BQdOnRAzZo1ceLECbRo0QK7du3KdQ1lYjN37lyd49u6dSsAoHfv3uDz+ejXrx+Aj5WdPh07dgyAog8CF/XCw4cPkZmZiZIlS6J06dK5ttevXx8AcOvWLa3Ol5qaCgB5vpcSJUoAAG7evKnV+bZv3w4Aqn8DQ16LS1p38vz5558xcuRInD171pDxGKccw1SpBcOwYmNj0ahRI7i7uyMiIgLlypXjOiRiIGFhYWjTpg2ioqJUI4KCgoIwePBgjBo1Cu/fv0dYWJjq2+/r169Rr1497NixA7/++ivKly8PQDGUvm/fvoiPj8eUKVPw22+/qSYV2rt3L3r27IkhQ4agWbNmcHV1/eJ4169fj4SEBIwdOxZ//vmn2rYXL16o9f4Xi8Xo1q0bxGIxtm3bhr59+6q2Xbt2De3atcPQoUPRpk0blCxZ8otjUmKM5arU2rZtC1dXV1y/fh337t3T6+KVN27cAPCxItfFs2fPvuj/NWNM9frFixcAoDG5AABLS0vY2dnhw4cPSE5OhrW1db7nVv4bPH/+XON25SiSvLbnFBYWhqdPn6JEiRJo3769xms9fvxY47kYY6pyba7FNa0TDOU/XsuWLQ0WjNHKOUyVJmQ0CMYYeDweXF1dsXLlSnTt2lXvSwsXNowxsLQMrsPQCc/CTG8zAvL5fPz9999qw40HDBiAH3/8EU+ePMHMmTPVmtZLlSqFvn37Yvny5bhw4YIqwTh37hxu376NMmXKYN68eWrxdevWDYGBgdi3bx82btyIn3766YvjfffuHQDAx8cn17YyZcqo/bxx40bExsbihx9+UEsuAMDT0xOzZs3CpEmTsG3bNkycOFG1zdXVFVWqVIGjo6NOsZ0/fx4vXrxAjRo1VOtJ8fl89O7dG8uWLcPWrVuxcOFCnc6Zn/fv3wPAFyVHVlZWXz2iRLlshYWFRZ77WFpaIjExUasEo0WLFgCAiIiIXMlYWloadu/eDQBITk7+bGzKlqRvv/1W40iQFi1aIDQ0FJs3b8bIkSPVtu3duxdJSUlaX4trOg1TNdR0xqtWrcLvv/+ON2/eoE6dOvjzzz/h7e2tcd9169Zhy5YtuHPnDgDFPbEFCxbkuX9BYHIGqYCGqRpKZmYmRo0ahUaNGmH48OF53hcvalhaBqI92nEdhk7KPTsJnqW5Xs7l4eGBypUrq5Xx+XyULVsW8fHxaNcu92ejTCpiY2NVZRcvXgQA9OzZU+Mf9P79+2Pfvn2q/b5UgwYNAAAzZsyAQCCAj49PnrMbK4eKfjpxoZJy6GJ4eLha+cKFC78oEVDeBunfv79aef/+/bFs2TJs374dCxYsKBRT1js6OiIoKIjrMNRUqVIFXbp0wf79+9G5c2esXbsWXl5eiIqKwoQJE1QJVV5zWihJJBJVMvLpv4XS6NGjsWrVKly5cgUDBgzArFmzULJkSZw8eRKjRo2CiYkJpFLpZ69VGOgUYeXKleHg4JDvQ1e7du3CpEmTMGfOHFy/fh116tSBr68v3r59q3H/c+fOoXfv3jh79izCwsLg7u6Odu3aqday5wStRWIwb9++Vd2Hz+/bCCl63NzcNJYrO9lp2q7clpmZqSp7/fo1AOQ5H4Gy/Gv/hgwaNAg9e/bEvXv30KlTJ9jb26NFixZYsGAB3rx5o7bvs2fPAABNmzbVOPmSl5cXACA+Pv6rYgKAjIwMBAcHg8fj5Zrzom7duqhRowZevnyJ8+fPq23TNtlQtm7n3F/ZT0DZqlPQlL8H+U0kpuzr8LnWC6UNGzagRYsWePr0Kdq2bQsbGxvUrVsXkZGRWLx4MYC8+00o/fvvv3j//j0qVaqEhg0batzH3d0d+/btg52dHbZu3YrKlSvD3t4evXr1QoUKFTBkyBCtrlUY6NSC8fPPP+fq8fq1li1bhmHDhmHw4MEAgNWrV+Po0aPYuHEjpk2blmt/5X1EpfXr12Pv3r0ICQnBgAED9Bqb1mgmT4O4desWOnXqhMzMTJw7dw6NGzfmOqQCxbMwQ7lnJ7kOQyc8C/2tR/S5b2j6+gb3Jd/a5XJ5rjKBQIBdu3Zh2rRpOHjwIM6cOYOrV6/i4sWL+O2333D8+HE0adJE7fju3bvnO+Ns1apVdY7tU4cOHUJSUhKEQiF69+6da7syidm6davayAVlQp9fJZ1ze873UbduXWzfvh3Xr1/XOd74+HhMnjxZ5+Nytnoob0m9evVK476pqalITEyEvb291gmGvb09zp07h3///Rfnzp1DUlISKlSogL59+6pmI61Ro0a+51C2JGnq3JmTj48PoqKisHPnTty5cwcCgQBNmjRBt27dVHXl565VGOiUYHz77bdwcnLS28WzsrIQGRmJ6dOnq8r4fD58fHxy9bjOS1paGiQSSZ6tJ5mZmWrfZgwzlJYSDEOYMmUKHBwccPDgwVz3sIsDHo+nt9sNxVmpUqUA5N0pTtmakLNFRLk4nvJe/qdevnyZ5/Xq1auHevXqYe7cuRCLxZg7dy6WL1+OCRMmqG55lC5dGg8fPsS0adNUt1YMRXnPPysrK9+hjcHBwVi1apXqto6yg2Re02ADir+nygQlZ4dKf39//Pjjjzh+/Dg+fPig07ftlJQUbN68Wev9lXImGFWqVIFIJMK7d+8QExOTq7VLmfjUrl1bp2vweDz4+/vD399frXzDhg0ANA8tVRKLxaphsZ9LMABFQjNq1Khc5WFhYeDz+ap+IYWZ1tWhIe7NxcfHQyaTwdnZWa3c2dk5V5NiXqZOnYpSpUpp7FgFKO5Z5lw7xd3d/avjzoXRVOH6wrIXzwMU2f6lS5eKZXJB9EfZn2HPnj2QyWS5tiu/Veacslk5muTRo0e59k9ISND6m7mNjQ0WLlwIHo+n6jcGAN988w0AYP/+/Vq+iy8THx+PEydOQCgUIiEhQdFxWMOjVatWEIvFavM41KtXD1ZWVoiPj8fVq1c1nl9ZYVasWBEuLi6q8urVq8Pf3x/p6en44Ycf8o0xKysL165dU/3s4eGRZ5z5PXIyNzdHmzZtAEDjEOTg4GAAinkovlZaWho2bNgAoVCYb+fU4OBgZGRkoGnTpqq+Qro6evQooqKi4OfnZ5i6TM+0rg4//QcsDH777Tfs3LkT+/fvz7Mz1fTp05GUlKR65PfN44vJ5bSaqh6kp6ejX79+8Pb2RlpaGhwdHWnBMvLVWrVqhVq1auHZs2eYPXu22t+y/fv3Y9++fbCyslLd2waAcuXKoUyZMrh9+7ZapZuamorhw4drbAndunWrWhKh9O+//4IxplYhjBgxAk5OTli8eDHWrl2b65aLVCrFiRMncp1v+vTpqFq1Kv766y+t3vvOnTshkUjg6+ubbyuC8tZJzjkxzMzMMGLECADAyJEjc33pU7bAAIo1Pz61Zs0aODo6YtOmTRgyZIiqI2ROFy5cQJMmTXDkyBGt3o8uJk2aBEAxOeTjx49V5WFhYVizZg3s7Ozw3XffqR0THh6OqlWrapxb59GjR7n+3RMSEtCrVy+8ePECM2bMyHNYLJB3R1tNIiMjc9W5oaGhGDx4MMzMzD47eVmhoe00ooaQmZnJBAIB279/v1r5gAEDWOfOnfM99vfff2e2trY6TzFriKnCU46eZws6/MnKrGBswvHP709ye/36NfP29mbm5uZs165dXIdTYGiq8Nyio6PznaK6ZcuWDACLjo7OtU05vfWna1vcunWLlShRggFg1apVY71792ZNmzZlAJiJiYnG37kNGzYwAEwgELDWrVuzTp06MWdnZ1apUiUWEBCQa3psZVmFChVYYGAg6927N2vUqBHj8XiMz+ez3bt3q50/LCxMtVaHu7s7a9++PevTpw9r06YNs7OzYwBy/W0cOHCgxveXF+V6Fzt27Mh3v/fv3zNTU1NmYmLC3r59qypPS0tTfd4WFhasXbt2rG/fvqxVq1bM1NSUAWDffvstk8vlGs97//59VqlSJQaACYVC1rx5c9a7d2/WuXNn1RTjAoGAbd5smEWcxo8fr4o9ICCAtW/fnpmYmGisdxhj7OzZswwAK1u2bK5tc+bMYWZmZqx58+bs22+/Ze3bt2dWVlaq6dBlMlmecbx8+ZLx+XwmFAq1Wi+pbNmyrHTp0szPz4/17t2beXp6Mh6Px8zNzdmRI0d0+Qi0VijWItE3b29vNnbsWNXPMpmMubm5sYULF+Z5zKJFi5iNjQ0LCwvT+XqGSDCSj5xnv3T+m5VZwdgPJ/V22mLj2rVrzM3Njbm5ubFr165xHU6BogQjN0MkGIwx9vz5czZs2DDm7u7OTE1NmaOjIwsMDGRXr17NM5ZNmzaxmjVrMqFQyJydndnQoUNZfHy8qqLPmWCcP3+ejRkzhtWtW5eVKFGCmZmZsfLly7Nvv/02zy9CsbGxbMqUKaxGjRrMwsKCWVhYsAoVKrCAgAAWFBSUa90LXRKMhw8fqirXlJSUz+7foUMHBoD98ccfauVZWVns77//Zs2bN2f29vbMxMSEOTo6snbt2rF//vknz+RCKTMzk61Zs4b5+voyZ2dnZmpqyqytrVm9evXY5MmT2cOHDz8b29fYtGkTa9CgAbOwsGB2dnbMz8+PXb58WeO++SUY58+fZwEBAax06dJMKBQyBwcH1q5dO7Zv377PxrBo0SIGgHXp0kWrmBcuXMgaNmzIHBwcmFAoZGXLlmXDhw9nT58+1er4L1EkE4ydO3cykUjEgoKC2L1799jw4cOZnZ0de/PmDWOMsf79+7Np06ap9v/tt9+YUChkwcHBLDY2VvXQtACNJgZJMA6fY3MC17IyKxibZrzrwHHm4sWLrHnz5uz169dch1LgKMEghBQGhkgwdBpFYgi9evXCu3fvMHv2bLx58wZ169bF8ePHVR0/X7x4oTYc7e+//0ZWVlauBWzmzJnzRfPz64VcToud6Ugul2Pr1q3o27cvmjVrhvPnzxeKSX4IIYToB+cJBgCMHTsWY8eO1bjt09UKlUPKChUaRaKT1NRUDBgwAPv374ebmxt8fHwouSCEkCKmUCQYRi/HYmeUYOTvxYsXCAgIwOPHj7F///48hxcTQggxbpRg6AMtdqaVmJgYeHt7w8zMDKGhoTpPckMIIcR40PdtPWBgkAqyWzAEHAdTiJUqVQqTJk1CREQEJReEEFLEUYKhD3JqwciLTCbD1KlTVYstTZky5YuWcCaEEGJcKMHQB1rsTCOxWIyAgAAsWbIEcXFxXIdDCCGkAFEfDH2g5dpziYqKQqdOnRATE4Njx47B19eX65AIIYQUIEow9IFaMHIZOnQosrKycOXKFb0sOU0IIcS4UIKhDwyQZicYxX2iLbFYDBsbGwQFBcHKygoODg5ch0QIIYQDxbw61A8ml6taMEyL6ScqlUrx/fffw9vbG+np6ShTpgwlF4QQUoxRC4Y+sI/DVItjC8aHDx/Qs2dPnDt3Dn/++SfMzc25DokQQgjHKMHQhxxThRe3FowHDx6gc+fOeP/+PU6ePInWrVtzHRIhhJBCgBIMfcgxiqS4tWA8ffoUIpEI4eHhqFChAtfhEEIIKSSKWXVoIMVsFAljDEePHgVjDB06dMB///1HyQXRm2fPnoHH46FVq1Zch1KkhYeHg8fjgcfj4Zdffsl3Xw8PD/B4vFyLT+YUFBT02X+3e/fu4fvvv0fNmjVha2sLkUgENzc3dO7cGVu2bEFWVtYXvpvPe/XqFQYPHoxSpUrBzMwMlStXxpw5c5CRkfFF59u+fTuaNm0Ka2trWFlZwcvLC+vWrQNjLM9jIiIi0LNnT5QqVQqmpqaws7ND8+bNsWnTpnyPU8rKykL16tXB4/FgYlL42weKQXVYAHL0wSjqCUZWVhaGDx+Ojh074vTp0wBgFL/ohBB1W7duVb3evn27Qa/FGMOsWbNQu3Zt/PXXX0hOTkbr1q3RtWtXlCtXDsePH8fAgQNRrVo1g1z/yZMnqFevHoKCglCiRAkEBARAJpPhl19+gY+PDzIzM3U636hRo9CvXz9cu3YNDRo0QMuWLfHkyRMMHz4cgwcP1njM3r170bhxY+zZsweurq7o2rUr6tevjytXrmDIkCHo16/fZ6+7YMECPHjwQKdYOcWKmaSkJAaAJSUl6e2ciUEHmO/kO6zMCsZOPtXbaQudd+/esRYtWjChUMg2bdrEdThFQnp6Ort37x5LT0/nOpRCIzo6mgFgLVu25DqUIisrK4s5OjoyAMzFxYUBYFeuXMlz/7JlyzIA7OzZs3nus2nTpjz/3aZNm8YAMGdnZ3b06NFc2xMSEtiMGTOYqanpl7ydz2ratCkDwMaNG6cqk0gkrEuXLgwAmzNnjtbnCg4OZgCYvb09u3btmqr89evXrGbNmgwA27Fjh9oxEomEOTk5MQBs+/btatvu3bvHHBwcGAB25syZPK977949JhQK2fDhwxkAJhAItI5ZG9r+LdKlDi3i37cLSDG4RRITEwMvLy88ePAAZ86cwaBBg7gOiRDyhY4fP474+Hg0bdoUo0ePBqDeoqFP4eHhWLRoEczNzXH27Fn4+/vn2sfe3h7z58/H2bNnDXL9y5cvw8nJCYsXL1aVm5iY4O+//4apqSn++OMPSKVSrc73999/AwAmT56MBg0aqMpdXV2xbNkyAFC7DqDoDP/27VtUqVIFffr0UdtWrVo1VetFRESExmsyxjB8+HDY2dnht99+0yrOwqCIVocFLOctkiK62JmLiwu6deuG8PBwNG3alOtwSDGRnp6OadOmoWzZshCJRKhYsSIWLVqU63715/ptzJ07FzweD0FBQWrlyr4FALBq1SrUrFkT5ubmKFeuHBYvXqy6zvXr19GpUyc4ODjAysoKAQEBeP78ea7rxMbGYvHixWjZsiXc3NwgFArh4uKCrl275ll55Ixh/fr1qF27NszNzeHi4oIRI0YgMTFRh09MO9u2bQMA9OvXT1W57dq1CxKJRO/XWrp0KRhjGDdu3GdvgRjib8vRo0cBAJ06dYJIJFLb5uzsjObNm+PDhw+4dOmSVueLjIwEAI2/ay1btgSfz8eNGzfw4sULVfmn181LiRIlNJavWbMGly5dwtKlS2Fvb6/VuQoDSjD0Iccw1aK0XDtjDIsWLcKZM2cgEAiwZMkSlC1bluuwSDGRlZWFdu3aYd26dfD09ETr1q0RExODadOmYdasWXq91sSJE/Hjjz+ibNmy8PHxwfv37zF16lTMnTsXly9fRvPmzfH69Wt88803cHV1xaFDh9C2bVukp6ernefgwYOYOnUq4uLiULt2bXTp0gWlSpXC/v370bRpU5w8eTLPGKZMmYIxY8bA1dUV7du3B2MMa9euRefOnfNMqJSJiS6SkpJw6NAhCIVC9OzZE+XKlUOTJk0QHx+P48eP63y+/MjlctU5P/3mXlBu3rwJAKhfv77G7cryW7duaXW+1NRUANBY0QuFQlhZWaldFwDKly+PChUq4OHDh9ixY4faMffv38e2bdtgb2+PLl265DpnbGwspk2bhrZt22rVT6MwoQRDH+Tyj4udFZEWjIyMDAwYMADTpk1DeHg41+EUS4wBaRLjemjREV5rYWFhEAgEiI6Oxt69e3H8+HFcvHgRAoEAy5cvR0pKit6utXv3bty+fRtHjx7F4cOHER4eDpFIhCVLlqBfv35YunQpIiMjsWvXLty9exdt2rTB06dPsXPnTrXzNG3aFHfu3MGDBw/w77//YteuXbh+/TqOHz8OHo+H0aNH5zlaYOvWrbh16xZOnDiBffv24e7du6hYsSIuXryo11sHwcHByMjIQPv27VWz7SorLn3fJomKioJYLIZIJEKNGjV0Pn7QoEGqRErbx9y5c9XOoWxJKF26tMZrKMs1tUhpUrJkyTz3T0hIgFgszrVdIBBg8+bNsLOzQ9++fdGgQQN8++23aNOmDWrXro3SpUsjJCRE4+zHY8eORUZGBv73v/9pFV9hQt3/9SFnH4wi0ILx5s0bBAYG4ubNm9ixYwd69+7NdUjFUroUqGZkf1PujwYsTPVzLj6fjzVr1sDGxkZV5unpifbt2+PIkSO4du2a3oay/vLLL2pDratWrQp/f3/s378fpUuXxsiRI1XbhEIhxo8fjzNnzuD8+fNqowZq1aql8fy+vr7o0aMHtm/fjjt37mjc79dff0WVKlVUPzs6OmLkyJGYPHkyLly4gDZt2qi2mZqaqu2rC2USkfPbcM+ePTF+/HgcPnwYSUlJsLW1/aJzf+r9+/cAFN/2BQLd/zg2a9ZM52Pq1q2r9rMyEbWwsNC4v6WlJQAgOTlZq/O3aNECO3fuRFBQEPz8/NS2bdy4UfX60/M1bdoU58+fR5cuXXD9+nVcv34dgOL36ZtvvkH58uVzXevgwYPYt28f5syZg8qVK2sVX2FCCYYeMAZIs//zFIUWjF69euHFixe4cOECvLy8uA6HFFNly5bVWIkq/9DGxsbq7Vrt2rXLVab8g5/fNk0xZGZm4vjx4wgPD8e7d+9Uczvcvn0bAPD48WONCYam6+T1Xt3c3L5ouKLy/7WdnR06deqkKi9RogT8/f1x8OBB7NmzB0OHDtX53IYwdOjQQhOL0uTJkxEcHIxdu3ahTJkyGDt2LCwsLLB3717Mnj0bJiYmkEql4PPVbxD8888/GDx4MBo1aoR//vkHNWrUwOvXr7FkyRIsXboUZ8+eRWhoqKq/RnJyMsaOHYvKlStj+vTpXLzVr0YJhj4wBhnP+OfByMrKglAoxOrVq2FjYwM3NzeuQyrWzE0ULQLGxFyPf1HyatK2trYGAJ3nLsiPpt915b30/LZ9GsPt27fRuXNnPHv2LM9r5fVNWdP71fd73b59Oxhj6N69e66Oh/369cPBgwexbdu2XJW6Nn09lLd+cu6r7LT44cMHyGSyL2rF+FrKf6u0tDSN25V9KpSf9ec0aNAAmzZtwrBhw/D777/j999/V23r0KEDTE1NceDAAbU+Go8fP8bAgQPh5OSEI0eOqGKqVKkS1qxZg9evX+PIkSPYuHEjRo0aBQCYMWMGXr16hdOnT2vdSbSwoQRDHxhTtWAY41ThjDH8+uuvOHr0KM6fP2+wyW6Ibng8/d1uMEaffgP8GnK5/IuvpW0cjDH07NkTz549w8iRIzFy5EiUL18eVlZW4PF4mDFjBhYuXJhnHwx9vt+8KG+PnDt3LtftB2VLy4ULF/D8+XO1Dt3K2wt5VdI5tylvOQCKlh4bGxuIxWLcvXsXtWvX1ine9evXaz26QykwMBCBgYGqn8uUKYP//vsPr1690ri/slyXDuz9+vVD69atsXv3bjx69AhmZmZo27YtOnTogObNmwOAWp+TnTt3QiKRwM/PT5Vc5NSzZ08cOXIEFy5cUCUYhw8fhpmZGX799Vf8+uuvuY6RyWSqW4QrVqzIdWuoMKAEQx+Y3GjnwUhLS8PgwYOxe/du/Prrr0abKZPiTSgUAkCeHT9fvnxp8BgePHiABw8ewNPTUzVXQk5RUVEGjyE/kZGRuH//PgDFzJZPnjzRuB9jDNu3b8eMGTNUZaVLl8a9e/fyfQ/KbTlbYvh8Pvz8/LB7927s2LFD5wTj0qVL2Lx5s07HeHh4qCUYderUwcGDB1V9Hj6lLNc1Njc3N0ycOFGtLD09HTdu3IC1tbXaqBVlEpNX3xZl+YcPH9TKMzIycP78+TxjUG4zxFBmfTCy6rCQYvg4isSIPtFXr16hefPmOHLkCPbt24eZM2d+0bA3Qrjm6OgIExMTREdH55owSSKR5PtHWl+UlYOmWx0fPnzAqVOnDB5DfpRzX0yePBmMMY0P5Vojyn2VWrRoAeDjnBKfksvlqm3Kb/BKkyZNAo/Hwx9//KFKcPISGhqq9nNQUFCeseb1+HQUSYcOHQAoWgQ+vdUUFxeHixcvwt7eXi9zcGzcuBGpqano378/zM3NVeUuLi4AgGvXrmk8TjlHioeHh6rs2bNneb5HQDEyRflzYV23x4iqw0LMSGfyvHDhAuLj43H58mWN468JMRZCoRCNGzdGQkICVq1apSqXSqX44YcfEB0dbfAYKlasCD6fjzNnzuDx48eq8oyMDIwcORIJCQl6u1ZMTAyqVq2KqlWrarW/TCbDP//8AwD5jgpr3rw53NzccP/+fdWEUgAwZMgQWFlZ4fjx41izZk2uc//00094+PAhSpcujW7duqltb9iwIaZMmYL09HS0adMGx44dy3XdpKQkzJkzB61bt9bq/ejC29sbTZs2xdu3bzF16lRVuVQqxejRoyGRSDBu3DiYmqrfjxwwYACqVq2K/fv35zqnpkTh4MGDmDJlChwdHfHzzz+rbQsICACg+Jv7aevWlStXsHz5cgBA9+7dv+xNFlJ0i0QPmJEtdhYREQEvLy/06dMHAQEBavdMCTFWs2fPhq+vLyZMmIBdu3bBxcUFkZGRSEtLw8CBA3VuateVk5MTvvvuO6xbtw516tRBmzZtYG5ujosXL0Imk2HQoEG5ZhL9UhKJBA8fPtR6/5MnTyIuLg6VK1fOc8IpQHFLo1evXli2bBm2bt2qmgrb1dUVW7ZsQe/evTFy5EgsX74cdevWhUwmQ3h4OF68eAE7Ozvs3r0bZmZmuc67cOFCmJiYYOHChejQoQPKli2LevXqwdzcHK9evcLVq1eRlZWFSpUq6f5haGHTpk1o3LgxVq5ciTNnzqB69eqIiIhAVFQUmjRponGUxosXL/Dw4UMkJSXl2ubl5YUKFSqgWrVqsLS0xJ07d3D37l2UKFEC//77LxwdHdX2r1+/PiZPnowlS5Zg9OjRWLVqFapXr47Xr18jLCwMcrkcw4cPh4+Pj0HeP1eMoDos/OQyBsZTfJSFOcGQy+WYNm0avL29cebMGQCg5IIUGT4+Pjh06BC8vLxw/fp1nD9/Ho0aNUJERIRa07Mh/f3331i6dCnKlSuHkJAQXLx4ET4+Prh27Rqns+AqO3dqM6eNcp9//vlH7XaTcv6G7777DhKJBAcOHMDRo0dhYWGB8ePH49atW2jcuLHGc/J4PMybNw+3bt3CmDFjYGFhgZCQEAQHB+Pp06fw9fXFtm3bcPfuXT2829wqVaqE//77D4MGDcK7d++wf/9+8Pl8zJo1CyEhITr3PZs4cSJsbGxw6dIlHDhwAFlZWZg0aRLu3bsHT09Pjcf8/vvv2LdvH9q1a4c3b95g//79uHfvHlq2bIkdO3bkahkqCngsry7NRZRYLIatrS2SkpLUJvD5Gm9X7oAXFNPg3h4J2BTCfpLJycno27cvjh49it9//x0TJ06k/haFQEZGBqKjo1GuXDmN3/wIIaQgaPu3SJc6lG6R6IGU8YDsurowtmDExsaiXbt2ePHiBQ4fPqxxNUNCCCFEnyjB0AMpQ6FOMEqUKIH69etj165dqF69OtfhEEIIKQYowdADKfuYVRSmBGPDhg2oU6cOPD09Dd7BjRBCCMmpEFWHxkua3YuFz+TgF4JuDVKpFBMmTMDQoUNx8OBBrsMhhBBSDFELhh5ImSKrEEAOrnO2xMRE9OrVCyEhIVi1ahVGjzayxSwIIYQUCZRg6IEsO8EwYfmvd2BojDEEBATg9u3bOHHiBNq2bctpPIQQQoovSjD0QIqcLRjckMvl4PP5WLx4MRwcHAw2YQ0hhBCiDeqDoQeqWyQcTCnCGMNff/2FDh06QCqVomHDhpRcEEII4RwlGHoglWffIingFgyJRIJRo0bh+++/pyXWCSGEFCp0i0QPlLdICjLBeP/+Pbp3747Lly9j/fr1+O677wrs2oQQQsjnUIKhB7LsOyMF2clz7969uHPnDkJCQnItj0wIIYRwjW6R6IE0+2MsiE6eT548AQAMGzYMd+/epeSCEEJIoUQJhh6ohqkaMMFgjGHJkiWoWrUqLly4AB6PBycnJ4NdjxBCCPkalGDowcdhqoYZRZKZmYnBgwfjxx9/xJQpU9CsWTODXIcQohsej1dgS8FzoXr16uDxeKhQoUK++82dOxc8Hg+DBg3Kdz8ejwcej4dnz55p3J6amoply5ahdevWcHZ2hlAohL29PRo3bozZs2fjxYsXX/hOPk8mk2H58uWoVasWzM3NUbJkSfTs2RP379/X+VytWrVSvVdNj+PHj+d5bFBQELy9vWFlZQUHBwf4+/sjNDQ03+tdvnwZ/v7+cHBwgJWVFby9vbFlyxad49Y36oOhB4ZswXj37h0CAwMRGRmJbdu2oW/fvnq/BiGEfCoyMlJVuUZFRSE0NBRNmjQx2PVCQ0PRrVs3vHnzBhYWFmjUqBGcnZ2RlJSEiIgIXLlyBYsXL8aRI0fg4+Oj12vL5XL06NED+/fvh52dHTp06ID4+HgEBwfj6NGjOHv2LLy9vXU+b7du3WBlZZWr3M3NTeP+EyZMwMqVK2Fubo527dohIyMDp06dwsmTJxEcHIzAwMBcx+zduxe9evWCXC5HixYt4OjoiJCQEAwcOBC3bt3CkiVLdI5bb1gxk5SUxACwpKQkvZ0zeN4RVmYFY/5LYvV2TqXk5GTm6+vLrly5ovdzE+6lp6eze/fusfT0dK5DIV8AACtbtizXYRjE+PHjGQDm6urKALCRI0fmue+cOXMYADZw4MB8zwmAAWDR0dFq5f/99x8zMzNjANjUqVNZSkqK2naZTMb27t3LKlSowDZt2vSF7yhv69atYwBYpUqV2Js3b1TlwcHBDACrWLEik0gkWp+vZcuWGt9nfk6dOsUAsBIlSrBHjx6pykNDQ5lQKGR2dnbsw4cPase8f/+e2djYMABs7969qvI3b96wihUrMgDs7NmzWl1f279FutShdItED2TZH6M+WzD279+PR48ewcrKCsePH0fDhg31dm5CCMmPTCbDP//8AwDYsmULeDwedu/ejaysLL1fizGG/v37IyMjA3PnzsVvv/0GS0tLtX34fD66du2KyMhIeHp66j2GZcuWAQAWL14MZ2dnVXm3bt3QuXNnPHnyxOALRypjmDlzptpkiY0bN8bIkSORmJiIDRs2qB2zfv16iMViBAQEoGvXrqpyZ2dnLF68GACwdOlSg8adH0ow9ODjPBhf3weDMYZ58+aha9euWL9+/VefjxBj8+zZM/B4PLRq1QpisRiTJk1CuXLlYGpqigkTJgBQLOr3559/wtfXF2XLloVIJEKJEiXg5+eHU6dOaTyv8r74s2fPcODAATRq1AiWlpZwcHBA79698erVK43HJSQkYOzYsShVqhTMzMxQvXp1rFy5EuwzM/du3boVzZo1g42NDSwsLFC7dm0sXLgQGRkZufYdNGgQeDwezp07h9OnT6NFixawtraGk5MThg0bhqSkJADA27dvMWLECLi5ucHMzAze3t44d+6c9h+ulk6ePIm3b9+iYcOG8PHxQfPmzZGQkICjR4/q/VrHjx/HnTt3ULp0afz000/57mtra4uaNWvq9frR0dG4f/8+zM3N0aFDh1zbu3fvDgA4fPiwXq+bU3p6Os6cOaN2PW1iUP57aDqmQ4cOMDMzw+nTpzX+zhUESjD04ONqql+XYKSnp6NPnz6YNWsWfvnlFyxatEgf4RFilNLT09GyZUsEBQWhbt266Ny5M+zt7QEAV65cwbhx4/Do0SNUqVIFXbp0QZUqVXDy5En4+vpi48aNeZ73f//7H7p37w5zc3P4+/vDysoKO3fuRJs2bZCenq6274cPH9CsWTOsWrVKtZigm5sbJk+ejHHjxuV5jREjRmDAgAGIjIxE8+bN0aFDB8TGxmLGjBlo06YN0tLSNB63f/9++Pn5gTEGPz8/iEQirF+/HgEBAYiPj0fjxo1x4sQJNG/eHHXr1kVERAT8/Pxw+/btXOdSJlRBQUFafNrqtm7dCgDo16+f2vO2bdt0PtfnKCvJHj16wMSk4LsF3rx5EwBQs2ZNmJqa5tpev359AMCtW7d0PveGDRswevRojB07Fn/88UeenVQfPnyIzMxMlCxZEqVLl9Y6BmXsyu05CYVC1KxZExkZGXj06JHOseuFVjdnihBD9MHY8MtJVmYFY72XPP/ic8jlctamTRtmYWHBgoOD9RYbKdyoD0Zu0dHRqnv1jRs3znXfmTHGoqKiWFhYWK7y69evMzs7O2ZjY8OSk5PVtinvi1tYWLDQ0FBVeWpqKmvSpAkDwDZs2KB2zMiRIxkA5ufnx1JTU1XlV69eZVZWVhr7YCjv25cqVUrtXnpiYiJr1qwZA8B++OEHtWMGDhzIADA+n8+OHDmiKheLxaxmzZoMAKtevTrr168fy8rKUm2fOXMmA8AGDBiQ67NQvl9d+yyIxWJmbm7OTExM2Lt37xhjjH348IGJRCImEolYQkJCrmO+pg9G06ZNGQC2detWneJkjLGzZ8+qzqnt49N/r5UrVzIArEuXLhqvkZiYyAAwBwcHreNSfvafPkxNTdkvv/ySa/+DBw8yAKxevXp5ntPOzo4BYGKxmDH2sS7Lrz4LDAxkANihQ4c+G7Mh+mDQKBI9kOlhmCqPx8PUqVNRsmRJ1KtXT1+hEWLU/vjjD9jZ2eUqL1euHMqVK5ervF69ehgzZgzmz5+Ps2fPolOnTrn2mThxIho3bqz62cLCApMmTUJoaCguXLiAIUOGAFAMmdy8eTP4fD7++usvWFhYqI7x9vbGmDFjNLYy/vHHHwCAOXPmqN1Lt7W1xapVq1C3bl2sWbMG8+bNg5mZmdqxffr0UWumt7a2xrBhwzB+/Hi8evUKly5dUvuWPXnyZMyfPx/nz5/PFUeZMmVQpUoV2Nra5tqWn7179yI9PR0dOnSAo6MjAKhGVuzbtw+7d+/GiBEjdDpnft6/fw8AKFmypM7Huri4YODAgTodo3xPSikpKQCg9u+bk7I/SHJystbXaNGiBYYOHYomTZrA1dUVL1++RHBwMObNm4fZs2fDxsYG48eP1zoGZRyJiYlITk6GtbW16hh9x65PlGDowdcMU92xYwdOnDiBoKAgtGvXTt+hkSIgNjYWsbGxamX29vYoV64cMjIycO/evVzHKJtMHz58iNTUVLVtHh4ecHBwwLt37/Dy5Uu1bdbW1qhUqRJkMpmq+TWnWrVqwdTUFE+fPlX1C1Byc3NT6yD3tVxdXfPt0CeTyRASEoLQ0FDExsYiMzMTAPD48WO1509p+n9WuXJlAFD7nCMjI5Geng5vb2+N80D07t07V4IhkUhw5coVANA4pLx27dqoXbs2bt68iRs3bqBRo0afja18+fIAAE9PT9UtIiVbW1s4ODjk+v0A8MXzIChvg/Tv31+tvH///ti3bx+2bt2q1wTja1StWvWLbgEZ2i+//KL2c+XKlTFjxgx4enrC19cXc+fOxfDhw2Fubs5RhAWDEgw9kEA5ikT7Fgy5XI5Zs2ZhwYIF6N+/PyQSCYRCoaFCJEZszZo1+Pnnn9XK+vbti23btuHVq1do0KBBrmNYdgfEQYMGqSo8pa1bt6Jfv37YvXs3xo4dq7atXbt2OHHiBFJTUzWe9+3btyhZsiQmTpyYq8PZ0qVLMWnSpC96j5qUKVMmz22vXr1Cx44dNSZBSnl9a9N0j9va2hoAVEkKALx+/RoAULZsWY3n0TTB1vv375GVlQVHR8dcIyFyHnfz5k3ExMTk2qZpfgTlPAp5zZ1gZWWlagX4WjExMTh79ixsbGzQuXNntW3KiZxCQ0MRHR2t1oLE4/E+e26Wo1Nszv1LlCgBQDHnDxeUn29e/WKUCbryd+RrtGvXDp6enrh27RquXr2KVq1aaRWDpjhyzq+RlpYGGxsbg8b+JSjB0ANVCwZPuxaMlJQU9OvXD4cOHcLixYsxefJkrf6DkuJpxIgRuf7YK7/Jli5dGpGRkXkeGxQUpLEFAwB69uypdqsA+PiHyNLSUuN5lbcrli9fjrlz56pty6sC/FKf3j7IaejQobh58ya6deuGKVOmoEqVKrC2tgafz8fatWsxYsSIPEd58Pnc9m3P7/96frEVRNzbt2+HXK74O/bNN9/k2i6Xy8EYw7Zt2zBr1ixVubKJPr8KMue2nMlX3bp1cfnyZVy/fl3VmVRbDx48wG+//abTMY6OjmqTTykT2bxGESnL80o0dVWpUiVcu3ZNrdXpczGkpqYiMTER9vb2qv+jNjY2sLW1RVJSEl69eoXq1asbPHZdUYKhB1KecrEz7VowNmzYgJCQEBw6dAgdO3Y0ZGikCHB1dYWrq6vGbWZmZhp7kCtVqVIlz20lS5bM8763QCDI97yfmzrakFJTU3Hq1Ck4Oztj165dEAgEatujoqL0ch3lZ/78+XON2zWVlyhRAkKhEPHx8UhNTdXYiqGcJlvfCZk+KG+PiMViXL58Od/9ciYYylah/D575TYLCwu1Wz0dOnTAqlWrsGfPHixevFinkSRv3rzB5s2btd4fUFS2OROMOnXqAADu3LkDiUSSayTJ9evXAShub+nDhw8fAKgnWVWqVIFIJMK7d+8QExOT63cjrxjq1KmDCxcu4Pr167kSDIlEgjt37sDMzEx1C7Cg0TBVPZBBuz4Yb9++BQB8//33uHnzJiUXhHyBpKQkyOVyuLq65kouJBIJ9u/fr5frNGjQAObm5oiMjNRYce7cuTNXmampqapfhabtd+7cwc2bN2FlZYW6devqJU59uXnzJm7fvg1nZ2dIpVIwxjQ+ypUrh0ePHuHq1auqY5s1awYej4cbN26obi196tChQwCAJk2aqP27+fn5oUaNGnj16hXmz5+fb4xisRh3795V/dyqVas848zr8ek6KOXKlUO1atWQnp6ucZ6P4OBgANDYYVhX7969w8WLFwGoDy01NzdHmzZtAAB79uzROgZlh2Dl9pyOHDmCjIwM+Pj45NsaaEiUYOjBx+Xa827B2LhxIzw8PHDlyhXw+XxVxy1CiG6cnJxga2uLO3fuqH3LlslkmDp1qt7G/FtZWaF///6QyWT4/vvv1ebIuHbtGv766y+Nx33//fcAFAuA5UxMkpOTMXbsWDDGMGLECIP/0R8wYACqVq2qdcKlnPuiZ8+euRK3nL799lsA6nNiuLu7o2vXrpDJZBg8eDDEYrHaMWFhYaqZJZWTpSnxeDxs27YNZmZmmDt3LqZPn57rth5jDIcOHYKnpyciIiK0ej+6UPYdmjJliuqLIADs27cPhw4dQsWKFREQEKB2zP79+1G1alUMGDBArTw0NBQHDhyATCZTK3/27Bm6dOmC1NRUdO7cOVdfIGUM8+bNU+ugHBYWhjVr1sDOzg7fffed2jFDhw6FjY0NDh48iH379qnK3759iylTpgAAfvjhB50+C7367EDWIsYQ82DMnXuZlVnB2NRl93Jtk0qlbNKkSQwAGzZsGMvMzNTbdYnxo3kwclPOg9GyZcs895k/fz4DwAQCAfvmm29Yr169mIeHBzM3N2djxoxhANicOXPUjslvfYi8rvn+/XtWpUoV1ZocvXr1Yu3atWMmJiaq62hai2T48OEMADM3N2cdOnRgPXr0YCVLlmQAWKNGjdTm1GDs4zwYmtaNUM71kNccE2XLlmWa/pTrMg+GVCpVrTmSc44QTW7dusUAMEdHR7U5Od6+fctq1KjBADBbW1vWoUMH1qdPH9a4cWPG4/EYADZlypQ8z3vp0iXm7Oysmqukbdu2rE+fPqxDhw6qcjMzM3b69OnPvh9dyWQy1qVLFwaA2dvbs+7du7NWrVoxHo/HzM3NNa4FtWnTJo2/M8pyFxcX5u/vz/r06cOaNm2qWmulRo0aLC4uTmMcyvVfLCwsWEBAAGvfvj0zMTFhAoGA7d+/X+MxwcHBjM/nMx6Px1q3bs26d++umjNj0qRJWn8GhpgHgxIMPZj5cxgrs4KxGcvuqpUnJiYyPz8/xufz2R9//MHkcrnerkmKBkowctMmwWCMsc2bN7N69eoxCwsLVqJECRYQEMBu3ryp+gOvjwSDMcbi4+PZqFGjmIuLCxOJRKxq1apsyZIlTC6X57vY2ZYtW1iTJk2YlZUVMzMzYzVq1GDz589naWlpufblOsE4ceIEA8A8PDw+uy9jTJVIfDqBU0pKClu0aBHz9vZmNjY2zMTEhLm4uLCAgAB2/Pjxz543OTmZLVmyhLVs2ZKVLFmSmZiYMDs7O9awYUM2Z84c9vLlS63i+xJSqZQtXbqU1ahRg5mZmbESJUqw7t27s7t372rcP68E4969e2zUqFGsfv36qvdga2vLGjVqxJYuXarx3//T8zZo0IBZWFgwOzs75ufnxy5fvpzvMZcuXWJ+fn7Mzs6OWVhYME9PTxYUFKTT+zdEgsFj7DMT6hcxYrFY1fNW07CeLzHjl6vYbtsQA9kd/DLh4zz58fHx8PX1xcKFC2mOC6JRRkaGasgfV/dJCSFE279FutShhaIPxqpVq+Dh4QEzMzM0bNgQ4eHh+e6/Z88eVK1aFWZmZqhVqxaOHTtWQJFqJoXifqVyHoxz587h1atXcHR0REREBCUXhBBCih3OE4xdu3Zh0qRJmDNnDq5fv446derA19dXraNNTqGhoejduze+++47/PfffwgMDERgYCDu3LlTwJF/pJoqnMfw999/w8fHR7VELtdj7gkhhBAucF77LVu2DMOGDcPgwYNRvXp1rF69GhYWFnmuhrhy5Ur4+fnhxx9/RLVq1fDrr7+ifv36efboLggy8MBkEhwNno/Ro0djzJgx+P333zmLhxBCCOEapwlGVlYWIiMj4ePjoyrj8/nw8fFBWFiYxmPCwsLU9gcAX1/fPPfPzMyEWCxWe+ibhPHxbkMXXA3dhzVr1mDlypWcLDtMCCGEFBacJhjx8fGQyWS5FkhydnbGmzdvNB7z5s0bnfZfuHAhbG1tVQ93d3f9BJ+DjM+HpWd/jB7zN4YPH6738xNCCCHGpsh/zZ4+fbraAkxisVjvScb3/g7oHlsXFSo66PW8hBBCiLHiNMFwdHSEQCBAXFycWnlcXBxcXFw0HuPi4qLT/iKRCCKRSD8B56FGg9KoYdArEEIIIcaF01skQqEQDRo0QEhIiKpMLpcjJCQk1yqPSo0bN1bbHwBOnTqV5/6EGINiNh0NIaSQMcTfIM5vkUyaNAkDBw6Ep6cnvL29sWLFCqSmpmLw4MEAFPPpu7m5YeHChQCA8ePHo2XLlli6dCk6dOiAnTt34tq1a1i7di2Xb4OQL6Jc80EikcDc3JzjaAghxZVEIgGAfNeh0RXnCUavXr3w7t07zJ49G2/evEHdunVx/PhxVUfOFy9eqM0l0aRJE+zYsQMzZ87EjBkzUKlSJRw4cAA1a9bM6xKEFFqmpqYQiURISkqCtbU1eDwe1yERQooZxhiSkpIgEolyLVf/NWiqcEI4JhaLERMTAysrK9ja2sLU1JQSDUKIwTHGIJFIkJSUhJSUFLi5uX22XtSlDuW8BYOQ4k75nzQ+Ph4xMTEcR0MIKW5EIpFWyYWuKMEgpBCwsbGBjY0NJBIJZDIZ1+EQQooJgUCg19siOVGCQUghYmpqarD/7IQQUpA4X4uEEEIIIUUPJRiEEEII0TtKMAghhBCid5RgEEIIIUTvKMEghBBCiN5RgkEIIYQQvaMEgxBCCCF6V+zmwVDOjC4WizmOhBBCCDEuyrpTm1VGil2CkZycDABwd3fnOBJCCCHEOCUnJ8PW1jbffYrdYmdyuRyvX7/W68qVYrEY7u7uePnyJS2gpif0meoffab6RZ+n/tFnql+G+DwZY0hOTkapUqXUVjrXpNi1YPD5fJQuXdog51auJ0H0hz5T/aPPVL/o89Q/+kz1S9+f5+daLpSokychhBBC9I4SDEIIIYToHSUYeiASiTBnzhyIRCKuQyky6DPVP/pM9Ys+T/2jz1S/uP48i10nT0IIIYQYHrVgEEIIIUTvKMEghBBCiN5RgkEIIYQQvaMEgxBCCCF6RwmGllatWgUPDw+YmZmhYcOGCA8Pz3f/PXv2oGrVqjAzM0OtWrVw7NixAorUeOjyma5btw7NmzeHvb097O3t4ePj89l/g+JG199RpZ07d4LH4yEwMNCwARohXT/TxMREjBkzBq6urhCJRKhcuTL9389B189zxYoVqFKlCszNzeHu7o6JEyciIyOjgKIt/C5cuIBOnTqhVKlS4PF4OHDgwGePOXfuHOrXrw+RSISKFSsiKCjIcAEy8lk7d+5kQqGQbdy4kd29e5cNGzaM2dnZsbi4OI37X758mQkEArZ48WJ27949NnPmTGZqaspu375dwJEXXrp+pn369GGrVq1i//33H7t//z4bNGgQs7W1Za9evSrgyAsnXT9PpejoaObm5saaN2/OAgICCiZYI6HrZ5qZmck8PT2Zv78/u3TpEouOjmbnzp1jN27cKODICyddP8/t27czkUjEtm/fzqKjo9mJEyeYq6srmzhxYgFHXngdO3aM/fTTT2zfvn3/b+/eg5o60z+Af0kwIUJQERACChUKdVF00eKCnXEVdmDaeldsyyJ4bRVst04VxtbiZWFpB21Xa6toG9qOAroWa8WiFaWLtB1ZSvACDcilbLtBV6UqF4HAs3/44/wIJEAw3PT5zGSGnPOe9zzvw4G85z3vySEAlJ6e3mX58vJyGj58OG3YsIGKiopoz549JBaLKTMzs0/i4w5GD/j6+lJkZKTwvqWlhRQKBf3tb3/TWz4kJISee+45nWXTp0+nl19+uU/jHEqMzWlHWq2W5HI5ffrpp30V4pDSm3xqtVry9/engwcPUnh4OHcwOjA2px999BGNHz+empqa+ivEIcXYfEZGRtLs2bN1lm3YsIFmzJjRp3EOVT3pYGzatIm8vLx0li1dupSCgoL6JCa+RNKNpqYm5OfnIzAwUFgmEokQGBiI77//Xu8233//vU55AAgKCjJY/nHTm5x2VF9fj+bmZtjY2PRVmENGb/O5fft22NvbY+XKlf0R5pDSm5yeOHECfn5+iIyMxJgxYzBx4kTEx8ejpaWlv8IetHqTT39/f+Tn5wuXUcrLy3Hq1Ck8++yz/RLzo6i/P5seu4edGevmzZtoaWnBmDFjdJaPGTMGP/30k95tqqur9Zavrq7usziHkt7ktKPo6GgoFIpOfyyPo97k88KFC/j444+hUqn6IcKhpzc5LS8vx7lz5xAaGopTp07h2rVrWLduHZqbmxEbG9sfYQ9avcnnSy+9hJs3b+KZZ54BEUGr1eKVV17B5s2b+yPkR5Khz6a7d++ioaEBMpnMpPvjEQw25CQkJCA1NRXp6emwsLAY6HCGnHv37iEsLAwHDhyAra3tQIfzyGhtbYW9vT2SkpIwdepULF26FG+++Sb27ds30KENSdnZ2YiPj8eHH36IH3/8EV988QUyMjKwY8eOgQ6N9RCPYHTD1tYWYrEY169f11l+/fp1ODg46N3GwcHBqPKPm97ktE1iYiISEhJw9uxZeHt792WYQ4ax+SwrK0NlZSXmzJkjLGttbQUAmJubQ61Ww83NrW+DHuR6c4w6Ojpi2LBhEIvFwrIJEyaguroaTU1NkEgkfRrzYNabfG7ZsgVhYWFYtWoVAGDSpEmoq6vDmjVr8Oabb0Ik4vNjYxn6bLK2tjb56AXAIxjdkkgkmDp1KrKysoRlra2tyMrKgp+fn95t/Pz8dMoDwDfffGOw/OOmNzkFgHfffRc7duxAZmYmpk2b1h+hDgnG5vOpp57C5cuXoVKphNfcuXMxa9YsqFQqjB07tj/DH5R6c4zOmDED165dEzprAFBSUgJHR8fHunMB9C6f9fX1nToRbZ034kdo9Uq/fzb1ydTRR0xqaipJpVJKTk6moqIiWrNmDY0cOZKqq6uJiCgsLIxiYmKE8rm5uWRubk6JiYlUXFxMsbGxfJtqB8bmNCEhgSQSCf3jH/8gjUYjvO7duzdQTRhUjM1nR3wXSWfG5rSqqorkcjlFRUWRWq2mkydPkr29Pf31r38dqCYMKsbmMzY2luRyOaWkpFB5eTmdOXOG3NzcKCQkZKCaMOjcu3ePCgoKqKCggADQrl27qKCggH7++WciIoqJiaGwsDChfNttqhs3bqTi4mLau3cv36Y6GOzZs4fGjRtHEomEfH196YcffhDWzZw5k8LDw3XKHzlyhDw8PEgikZCXlxdlZGT0c8SDnzE5dXFxIQCdXrGxsf0f+CBl7DHaHncw9DM2p9999x1Nnz6dpFIpjR8/nuLi4kir1fZz1IOXMflsbm6mrVu3kpubG1lYWNDYsWNp3bp1VFNT0/+BD1Lnz5/X+3+xLY/h4eE0c+bMTttMmTKFJBIJjR8/npRKZZ/Fx49rZ4wxxpjJ8RwMxhhjjJkcdzAYY4wxZnLcwWCMMcaYyXEHgzHGGGMmxx0MxhhjjJkcdzAYY4wxZnLcwWCMMcaYyXEHg7H/k5ycjJEjRw50GL1mZmaG48ePd1kmIiIC8+fP75d4BpstW7ZgzZo1Ax2GSWRnZ8PMzAy//fabsOz48eNwd3eHWCzGX/7yF6OPZ1dXV7z//vsPFVdRURGcnZ1RV1f3UPWwR0SffYUXYwMgPDxc7zfblZaWdrutUqmkESNG9FlsSqVSiMfMzIycnJwoIiKCrl+/bpL6NRoN3b9/n4iIKioqCAAVFBTolPntt9/6/JsQY2NjhXaKRCJydnam1atX061bt4yqx5TfLqrRaEgul1NlZaWw7Ntvv6Xnn3+eHB0dCQClp6f3qC6VSkVz5swhOzs7kkql5OLiQiEhISb7PfZEY2MjaTQaam1tFZbZ29tTdHQ0/frrr3T37l2qr683KqYbN25QXV2d8N6YnLS3aNEi2r59u9HbsUcPj2CwR05wcDA0Go3O64knnhjosAAA1tbW0Gg0+OWXX3DgwAF8/fXXCAsLM0ndDg4OkEqlXZYZMWJEv4zSeHl5QaPRoKqqCkqlEpmZmVi7dm2f79eQgwcPwt/fHy4uLsKyuro6TJ48GXv37u1xPf/9738REBAAGxsbnD59GsXFxVAqlVAoFP161i6RSODg4AAzMzMAQG1tLW7cuIGgoCAoFArI5XLIZDLY29v3uE47OzsMHz78oWNbvnw5PvroI2i12oeuiw1xA93DYcyUujrr3blzJ02cOJGGDx9Ozs7OtHbtWp2HpXUcwVCpVPTHP/6RrKysSC6Xk4+PD+Xl5Qnrc3Jy6JlnniELCwtydnam9evXU21trcHY9I2QxMXFkUgkovr6emppaaFt27aRk5MTSSQSmjx5Mn399ddC2cbGRoqMjCQHBweSSqU0btw4io+PF9aj3RknOozgtD2PoH1+9u/fT46OjtTS0qIT09y5c2n58uXC++PHj9Pvf/97kkql9MQTT9DWrVupubnZYDtjY2Np8uTJOss2bNhAo0aNEt5rtVpasWIFubq6koWFBXl4eND777+vU0fHNpw/f56IHjxUbMmSJTRixAgaNWoUzZ07lyoqKgzGQ0Tk5eVFH3zwgcH16OHZenp6Opmbm3fZ/rbnQ5w8eZImTZpEUqmUpk+f3ulhh90dP/fv36dNmzaRs7MzSSQScnNzo4MHD+rso6amRu/zKM6fP6/3eDtx4gRNmzaNpFIpjR49mubPny+sc3Fxoffee0/4uX19Li4uVFFRQWZmZjp/A0RE7733Ho0bN044jhobG0kqldLZs2e7zSd7tPEIBntsiEQi7N69G1evXsWnn36Kc+fOYdOmTQbLh4aGwtnZGXl5ecjPz0dMTAyGDRsGACgrK0NwcDAWLVqES5cuIS0tDRcuXEBUVJRRMclkMrS2tkKr1eLvf/87du7cicTERFy6dAlBQUGYO3cuSktLAQC7d+/GiRMncOTIEajVahw6dAiurq5667148SIA4OzZs9BoNPjiiy86lVmyZAlu3bqF8+fPC8tu376NzMxMhIaGAgBycnKwbNkyvPbaaygqKsL+/fuRnJyMuLi4HrexsrISp0+f1nlkeWtrK5ydnXH06FEUFRXh7bffxubNm3HkyBEAwBtvvIGQkBCd0Sh/f380NzcjKCgIcrkcOTk5yM3NhZWVFYKDg9HU1KR3/7dv30ZRURGmTZvW45gNcXBwgFarRXp6erePDN+4cSN27tyJvLw82NnZYc6cOWhubgbQs+Nn2bJlSElJwe7du1FcXIz9+/fDysqq0378/f2hVqsBAMeOHRNy1VFGRgYWLFiAZ599FgUFBcjKyoKvr6/e2PPy8gAASqUSGo0GeXl5cHV1RWBgIJRKpU5ZpVKJiIgI4dHqEokEU6ZMQU5OTpf5YY+Bge7hMGZK4eHhJBaLydLSUngtXrxYb9mjR4/S6NGjhfcdz/jkcjklJyfr3XblypW0Zs0anWU5OTkkEomooaFB7zYd6y8pKSEPDw+aNm0aEREpFAqKi4vT2ebpp5+mdevWERHR+vXrafbs2TrX3dtDu7NwQ3MwOo7wzJs3j1asWCG8379/PykUCuFsNCAgQGeUhIjo888/J0dHR70xED0YfRCJRGRpaUkWFhbCWfCuXbsMbkNEFBkZSYsWLTIYa9u+PT09dXLQ2NhIMpmMTp8+rbfetkdZV1VVGdw3jJhvsHnzZjI3NycbGxsKDg6md999V3jkONH/jy6kpqYKy27dukUymYzS0tKIqPvjR61WEwD65ptv9MbQfgSDiKimpkZnlIeo8/Hm5+dHoaGhBtvVfgSDSH9O0tLSaNSoUcJcn/z8fDIzM+s0grRgwQKKiIgwuC/2eOARDPbImTVrFlQqlfDavXs3gAdn8wEBAXBycoJcLkdYWBhu3bqF+vp6vfVs2LABq1atQmBgIBISElBWViasKywsRHJyMqysrIRXUFAQWltbUVFRYTC2O3fuwMrKCsOHD4enpyfGjBmDQ4cO4e7du/jPf/6DGTNm6JSfMWMGiouLATy4A0SlUsHT0xOvvvoqzpw587CpQmhoKI4dO4bGxkYAwKFDh/DCCy8IZ6OFhYXYvn27TjtXr14NjUZjMG8A4OnpCZVKhby8PERHRyMoKAjr16/XKbN3715MnToVdnZ2sLKyQlJSEqqqqrqMt7CwENeuXYNcLhfisbGxwf3793V+P+01NDQAACwsLHqcFwCIj4/XaXdbbHFxcaiursa+ffvg5eWFffv24amnnsLly5d1tvfz8xN+trGxgaenp/C77O74UalUEIvFmDlzplExd0WlUiEgIOCh6pg/fz7EYjHS09MBPLjzatasWZ1G0mQyWZfHB3s8cAeDPXIsLS3h7u4uvBwdHVFZWYnnn38e3t7eOHbsGPLz84XJfYaG1rdu3YqrV6/iueeew7lz5/C73/1O+MdaW1uLl19+WacjU1hYiNLSUri5uRmMTS6XQ6VS4cqVK6irq8M///lPeHh49KhdPj4+qKiowI4dO9DQ0ICQkBAsXrzYyOzomjNnDogIGRkZ+Pe//42cnBzh8khbO7dt26bTzsuXL6O0tLTLD2yJRAJ3d3dMnDgRCQkJEIvF2LZtm7A+NTUVb7zxBlauXIkzZ85ApVJh+fLlBn8X7eOZOnWqTjwqlQolJSV46aWX9G5ja2sLAKipqTEmNXjllVd09qFQKIR1o0ePxpIlS5CYmIji4mIoFAokJib2uO7ujh+ZTGZUrD1hijolEgmWLVsGpVKJpqYmHD58GCtWrOhU7vbt27Czs3vo/bGhzXygA2CsP+Tn56O1tRU7d+4Uzs7brvd3xcPDAx4eHnj99dfx4osvQqlUYsGCBfDx8UFRURHc3d2NikMkEundxtraGgqFArm5uTpnrbm5uTrXya2trbF06VIsXboUixcvRnBwMG7fvg0bGxud+trmO7S0tHQZj4WFBRYuXIhDhw7h2rVr8PT0hI+Pj7Dex8cHarXa6HZ29NZbb2H27NlYu3at0E5/f3+sW7dOKNNxBEIikXSK38fHB2lpabC3t4e1tXWP9u3m5gZra2sUFRX1uDMHPBh16JhXfSQSCdzc3DrdRfLDDz9g3LhxAB50bkpKSjBhwgShHV0dP5MmTUJrayu+/fZbBAYG9jjmrnh7eyMrKwvLly/vUflhw4bpPX5WrVqFiRMn4sMPP4RWq8XChQs7lbly5cpDd37Z0McjGOyx4O7ujubmZuzZswfl5eX4/PPPsW/fPoPlGxoaEBUVhezsbPz888/Izc1FXl6e8AERHR2N7777DlFRUVCpVCgtLcWXX35p9CTP9jZu3Ih33nkHaWlpUKvViImJgUqlwmuvvQYA2LVrF1JSUvDTTz+hpKQER48ehYODg97bTu3t7SGTyZCZmYnr16/jzp07BvcbGhqKjIwMfPLJJzqjFwDw9ttv47PPPsO2bdtw9epVFBcXIzU1FW+99ZZRbfPz84O3tzfi4+MBAE8++ST+9a9/4fTp0ygpKcGWLVuEiYVtXF1dcenSJajVaty8eRPNzc0IDQ2Fra0t5s2bh5ycHFRUVCA7OxuvvvoqfvnlF737FolECAwMxIULF3SW19bWCqMHAIRLE11dpjl58iT+/Oc/4+TJkygpKYFarUZiYiJOnTqFefPm6ZTdvn07srKycOXKFURERMDW1lb4krPujh9XV1eEh4djxYoVOH78uNDOnnSKDYmNjUVKSgpiY2NRXFyMy5cv45133jFY3tXVFVlZWaiurtYZ/ZkwYQL+8Ic/IDo6Gi+++GKnkZHKykr8+uuvJusYsSFsoCeBMGZKXd2mumvXLnJ0dCSZTEZBQUH02Wef6UyUaz8prrGxkV544QUaO3YsSSQSUigUFBUVpTOB8+LFi/SnP/2JrKysyNLSkry9vTtN0myvuy/yamlpoa1bt5KTkxMNGzas022qSUlJNGXKFLK0tCRra2sKCAigH3/8UViPDpPyDhw4QGPHjiWRSKT3NtX2+237sqmysrJOcWVmZpK/vz/JZDKytrYmX19fSkpKMtgOfbepEhGlpKSQVCqlqqoqun//PkVERNCIESNo5MiRtHbtWoqJidHZ7saNG0J+0W4Co0ajoWXLlpGtrS1JpVIaP348rV69mu7cuWMwplOnTpGTk5POLbn6bu8EQOHh4QbrKSsro9WrV5OHhwfJZDIaOXIkPf3006RUKjvV+9VXX5GXlxdJJBLy9fWlwsJCnbq6O34aGhro9ddfJ0dHR5JIJOTu7k6ffPKJzj6MmeRJRHTs2DGaMmUKSSQSsrW1pYULFwrrOk7yPHHiBLm7u5O5uTm5uLjo1PPxxx8TALp48WKnHMXHx1NQUJDBHLLHhxlRN/daMcbYEEdEmD59unCpqy9lZ2dj1qxZqKmpGdJfPd+VHTt24OjRo7h06ZLO8qamJjz55JM4fPhwpwnL7PHDl0gYY488MzMzJCUl8bdLPqTa2lpcuXIFH3zwQae7ggCgqqoKmzdv5s4FAwDwCAZjjJnQozyCERERgZSUFMyfPx+HDx+GWCwe6JDYIMYdDMYYY4yZHF8iYYwxxpjJcQeDMcYYYybHHQzGGGOMmRx3MBhjjDFmctzBYIwxxpjJcQeDMcYYYybHHQzGGGOMmRx3MBhjjDFmctzBYIwxxpjJ/Q+zuTE6kVK54gAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 600x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,5))\n",
    "plt.rcParams['font.size'] = 16\n",
    "\n",
    "ROC_plot(mouse_total_true_y_value, mouse_total_y_score, threshold=0.5, color='crimson',base_line=False,legend_label='mouse')\n",
    "ROC_plot(total_true_y_value, total_y_score, threshold=0.5, color='dodgerblue',legend_label='human')\n",
    "plt.legend(fontsize=15)\n",
    "#plt.savefig('/mnt/ruiyanhou/nfs_share2/three_primer/figure/deep_learning_mouse_human.pdf',dpi=300,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bd84ef0-dc24-45c6-a72c-5a1337c38aef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PY39",
   "language": "python",
   "name": "py39"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
